---
title: "CountsPlotting"
author: "John E Froberg"
date: "Sys.Date()"
output:
  pdf_document: default
  html_document: default
---

```{r setup}
knitr::opts_chunk$set(echo=FALSE, message=FALSE)

library('devtools')
library('tidyverse')
library('riboWaltz')
library("rmarkdown")
library("patchwork")
library("pheatmap")
library("RColorBrewer")
library("ggplotify")
library("topGO")
library("ggstance")
library("DESeq2")
library("biomaRt")
library("UpSetR")
library("GGally")
library("clusterProfiler")
library("org.Mm.eg.db")
library("enrichplot")

# #variables that are normally passed in from the Snakemake run:
experimentTypeFile="experimentTypesFull_SubFall2021.csv"
lengthCountsFile="dedup_lengthDistroTidy.txt"
CDS="dedupBams/featureCounts_CDS_summary.txt"
three_utr="dedupBams/featureCounts_three_prime_utr_summary.txt"
five_utr="dedupBams/featureCounts_five_prime_utr_summary.txt"
gtf="/n/macklis_lab/users/jfroberg/gtfs/Mus_musculus.GRCm38.95_chrNamed_headFix.gtf"
bamFiles="dedup_RPbams"
outDir="interactivePlots"
dir.create(outDir)
RiboCodeFile="RiboCode_ORFs_out.txt"


JF_theme <- theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "black"), plot.title = element_text(size = 7, face = "bold"),legend.title=element_text(size=7), legend.text=element_text(size=7), axis.text = element_text(size=7), axis.title.x = element_text(size=7),axis.title.y=element_text(size=7))
```


```{r getGeneNames_from_biomaRt}
mouse_ensembl <- useMart("ensembl", dataset = "mmusculus_gene_ensembl")
mouse_gene_symbols <- getBM(attributes=c("ensembl_gene_id","external_gene_name"), mart=mouse_ensembl)
mouse_gene_symbols_wGeneid <- mouse_gene_symbols %>% mutate(Geneid=ensembl_gene_id)
ensmusg2ensmust <- getBM(attributes=c("ensembl_gene_id","external_gene_name", "ensembl_transcript_id"), mart=mouse_ensembl) %>% mutate(Geneid=ensembl_gene_id)

ensembl_to_entrez <- org.Mm.egENSEMBL2EG %>%
  as.data.frame()
ensembl_to_allGO <- as.data.frame(org.Mm.egGO2ALLEGS) %>%
  left_join(ensembl_to_entrez, by = "gene_id")


# mouse_gene_symbols_resExp <- getBM(attributes=c("ensembl_gene_id","external_gene_name"), filters="ensembl_gene_id", values=resExp_df$Geneid, mart=mouse_ensembl)
# mouse_gene_symbols_resExp$Geneid <- mouse_gene_symbols_resExp$ensembl_gene_id
```


###Load in the metadata from the experiment types file. 
###Compute "descriptive names" by dropping the "samps" and "date" fields and pasting together.
###Save descriptiveNames as Upper Case "Samples" in experimentTypes.
```{r loadMetadata, echo=FALSE}
experimentTypes=read.csv(experimentTypeFile)
 descriptiveNames <- experimentTypes %>%
  unite(col="descriptiveNames", -samps, -date,-exp, sep=".")
 descriptiveNames <- dplyr::select(descriptiveNames, samps, descriptiveNames)
 experimentTypes$Samples <- descriptiveNames$descriptiveNames
 experimentTypes
```

###Convert length counts to relative frequencies, add in the metadata
```{r countsToFreqwMeta}
lengthCounts <- read.csv(lengthCountsFile)
totalCounts <- lengthCounts %>%
  group_by(samps,anno) %>%
  summarize(sum=sum(count))

lengthDistro <- merge(lengthCounts, totalCounts, by=c("samps", "anno"))
lengthDistro <- lengthDistro %>%
  mutate(freq=count/sum)

#metadata <- read.csv("metadata.csv")
lengthDistro <- merge(lengthDistro, experimentTypes, by="samps")

```
###Find the length with max freq for all samples
```{r maxFreq}
maxFreq <- lengthDistro %>%
  group_by(samps, anno) %>%
  summarize(max=max(freq))

max <- merge(maxFreq, lengthDistro, by=c("samps","anno"))
maxLengthDistro <- filter(max, freq==max)
maxLengthCDS_RPs <- filter(maxLengthDistro, exp=="RP" & anno=="CDS")
```


###Plot Freq vs length for the CDS with various groupings of the samples
```{r freqPlotting}
CDS_riboProf <- ggplot(filter(lengthDistro, exp=="RP",anno=="CDS"), aes(x=length, y=freq, color=Samples))+geom_line(size=1)+JF_theme+xlim(20,40)+xlab("length (nt)")+theme(axis.title.x = element_text(vjust=0))

CDS_riboProf_byDate <- ggplot(filter(lengthDistro, exp=="RP",anno=="CDS"), aes(x=length, y=freq, color=Samples))+geom_line(size=1)+facet_wrap(~date)+xlim(20,40)+JF_theme

CDS_riboProf_byDate_noxlim <- ggplot(filter(lengthDistro, exp=="RP",anno=="CDS"), aes(x=length, y=freq, color=descriptiveNames))+geom_line(size=1)+facet_wrap(~date)+ylim(0,0.5)+JF_theme

if ( "AF" %in% levels(lengthDistro$exp) ) {
CDS_AF_byDate <-ggplot(filter(lengthDistro, exp=="AF",anno=="CDS"), aes(x=length, y=freq, color=Samples))+geom_line(size=1)+facet_wrap(~date)+ylim(0,0.5)+JF_theme
CDS_AF_byDate
}


allAnno_RiboProf_bySample <- ggplot(filter(lengthDistro, exp=="RP", anno %in% c("CDS","three_prime_utr","snoRNA")), aes(x=length, y=freq, color=anno))+geom_line(size=1)+facet_wrap(~Samples, nrow=ceiling(nlevels(factor(lengthDistro$Samples))/6))+xlim(20,40)+JF_theme+xlab("length (nt)")

CDS_riboProf
CDS_riboProf_byDate
allAnno_RiboProf_bySample
```



###Load in the featureCounts. Use only CDS reads for RiboProf, 3'+5'UTR+CDS reads for AlkFrag libraries
```{r loadFeatureCounts, echo=FALSE}
loadFC <- function(experimentType, CDS, five_utr, three_utr) {
sampNames=experimentType$samps
CDS <- read.table(file=CDS, header= T, sep="\t", stringsAsFactors = F)
Five_prime <- read.table(file=five_utr, header= T, sep="\t", stringsAsFactors = F)
Three_prime <- read.table(file=three_utr, header= T, sep="\t", stringsAsFactors = F)
filteredCDS <- CDS %>% 
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)
filtered5p <- Five_prime %>%
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)
filtered3p <- Three_prime %>%
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)

#move to new Rmd
maxSampCol <- 6+length(sampNames)
mRNA_counts <- filteredCDS[,7:maxSampCol]+filtered5p[,7:maxSampCol]+filtered3p[,7:maxSampCol]
CDS_counts <- filteredCDS[,7:maxSampCol]
#

colnames(mRNA_counts) <- sampNames
colnames(CDS_counts) <- sampNames
mRNA_counts <- mRNA_counts %>%
  mutate(Geneid=filteredCDS$Geneid, Length=filteredCDS$Length)
CDS_counts <- CDS_counts %>%
  mutate(Geneid=filteredCDS$Geneid, Length=filteredCDS$Length)
mRNA_counts_tidy <- gather(mRNA_counts,"samps","counts", -Geneid, -Length)
CDS_counts_tidy <- gather(CDS_counts,"samps","counts", -Geneid, -Length)
mRNA_counts_tidy <- merge(mRNA_counts_tidy, experimentType, by="samps")
CDS_counts_tidy <- merge(CDS_counts_tidy, experimentType, by="samps")

raw_counts_tidy <- rbind(filter(mRNA_counts_tidy, exp=="AF"),filter(CDS_counts_tidy, exp=="RP"))
}


featureCounts_tidy <- loadFC(experimentTypes, CDS, five_utr, three_utr)
total_mRNA_counts <- featureCounts_tidy %>%
  group_by(Samples) %>%
  summarize(mRNA_counts=sum(counts))
total_mRNA_counts
total_mRNA_counts_plot <- ggplot(total_mRNA_counts, aes(x=Samples, y=mRNA_counts, fill=Samples))+geom_bar(stat="identity", position="dodge")+JF_theme+theme(axis.text.x = element_text(angle = 45, hjust=1))+xlab("")+geom_hline(yintercept = 150000)
total_mRNA_counts_plot

experimentTypesPlus_mRNACounts <- merge(total_mRNA_counts, experimentTypes, by="Samples")


total_mRNA_counts_facet <- ggplot(experimentTypesPlus_mRNACounts, aes(x=Samples, y=mRNA_counts, fill=Samples))+geom_bar(stat="identity", position="dodge")+JF_theme+theme(axis.text.x = element_text(angle = 45, hjust=1))+xlab("")+geom_hline(yintercept = 150000)+facet_wrap(~exp, nrow=2, scales="free_x")
total_mRNA_counts_facet
```

###Calculate the and plot reads per kilobase per million mapped (to mRNA) for the ribosome profiling experiments
```{r regionCovCalc}
#moveTo new Rmd
loadFCw_regions <- function(experimentType, CDS, five_utr, three_utr) {
sampNames=experimentType$samps
#write.csv(file="test.txt", sampNames)
CDS <- read.table(file=CDS, header= T, sep="\t", stringsAsFactors = F)
Five_prime <- read.table(file=five_utr, header= T, sep="\t", stringsAsFactors = F)
Three_prime <- read.table(file=three_utr, header= T, sep="\t", stringsAsFactors = F)
filteredCDS <- CDS %>% 
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)
filtered5p <- Five_prime %>%
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)
filtered3p <- Three_prime %>%
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)

maxSampCol <- 6+length(sampNames)
mRNA_counts <- filteredCDS[,7:maxSampCol]+filtered5p[,7:maxSampCol]+filtered3p[,7:maxSampCol]
CDS_counts <- filteredCDS[,7:maxSampCol]
FiveP_counts <- filtered5p[,7:maxSampCol]
ThreeP_counts <- filtered3p[,7:maxSampCol]
colnames(CDS_counts) <- sampNames
colnames(FiveP_counts) <- sampNames
colnames(ThreeP_counts) <- sampNames
CDS_counts <- CDS_counts %>% mutate(Geneid=filteredCDS$Geneid, Length=filteredCDS$Length, region="CDS")
FiveP_counts <-FiveP_counts %>% mutate(Geneid=filteredCDS$Geneid, Length=filtered5p$Length, region="5' UTR")
ThreeP_counts <-ThreeP_counts %>% mutate(Geneid=filteredCDS$Geneid, Length=filtered3p$Length, region="3' UTR")

All_counts <- rbind(CDS_counts, FiveP_counts, ThreeP_counts)
All_counts_tidy <- gather(All_counts, "samps","counts", -Geneid, -Length, -region)
All_counts_tidy <- merge(All_counts_tidy, experimentType, by="samps")

rpk <- All_counts_tidy %>%
  group_by(samps, region, exp) %>%
  summarize(totCounts=sum(counts),totLength=sum(Length)) %>%
  mutate(rpk=totCounts/(totLength/1000))

million_mapped <- All_counts_tidy %>%
  group_by(samps) %>%
  summarize(mRNA_mapped_mil=sum(counts)/1e6)

rpkm <- merge(rpk, million_mapped, by="samps") %>%
  mutate(rpkm=rpk/mRNA_mapped_mil)
#
}

rpkm_byRegion <- loadFCw_regions(experimentTypes, CDS, five_utr, three_utr)
rpkm_byRegion$region <- factor(rpkm_byRegion$region, levels=c("5' UTR","CDS","3' UTR"))
rpkm_byRegion <- merge(rpkm_byRegion, descriptiveNames, by="samps")
write.csv(file="test134-185.txt", head(rpkm_byRegion) , quote=F, row.names=F)
rpkm_byRegion <- rpkm_byRegion %>%
  mutate(Samples=descriptiveNames)
rpkm_byRegion_plot <- ggplot(filter(rpkm_byRegion, exp=="RP"), aes(y=rpkm,x=Samples, fill=region))+facet_wrap(~exp)+geom_bar(stat="identity", position="dodge")+JF_theme+theme(axis.text.x = element_text(angle = 45, hjust=1))+xlab("")
rpkm_byRegion_plot
#
```


###Calculate TPMs from the raw counts
```{r calculateTPM, echo=FALSE}

calcTPM <- function(RC_tidy) {
  RC <- RC_tidy %>% 
    dplyr::select(samps, Geneid, Length, counts) %>% #not robust to different types of metadata
    spread("samps","counts")
  Geneid=RC$Geneid
  lengths=RC$Length
  rawCounts = RC[,3:ncol(RC)] #rawCounts start with 7th column
  
  x <- rawCounts/lengths
  tpm.mat <- t( t(x) * 1e6 / colSums(x) )
  TPM_out <- data.frame(Geneid, tpm.mat)
  colnames(TPM_out) <- c("Geneid",colnames(rawCounts))
  TPM_out
}
TPMs <- calcTPM(featureCounts_tidy)
```

###Now plot the log10TPM+1 distributions
```{r plotTPM_distros, echo=FALSE}

tpmDensity <- function(TPMs,TPM_title="TPM_density") {
  #takes in a tidy table of tpms and returns a density plot
 JF_theme <- theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "black"), plot.title = element_text(size = 7, face = "bold"),legend.title=element_text(size=7), legend.text=element_text(size=7), axis.text = element_text(size=7), axis.title.x = element_text(size=7),axis.title.y=element_text(size=7))
  tpmPlot <- ggplot(TPMs,aes(log10TPM,fill=Samples ))+geom_histogram(alpha=0.3,position="identity",bins=100)+JF_theme+theme(legend.position = c(0.75,0.75))+ggtitle(TPM_title)
}


TPMs_tidy <- gather(TPMs, "samps","TPM",-Geneid)
TPMs_tidy <- TPMs_tidy %>%
  mutate(log10TPM=log10(TPM+1))
TPMs_tidy <- merge(TPMs_tidy, descriptiveNames, by="samps")
TPMs_tidy <- TPMs_tidy %>%
  mutate(Samples=descriptiveNames)

TPMdensity <- tpmDensity(TPMs_tidy)
TPMdensity <- TPMdensity + ylab("# of genes")
TPMdensity
```

###Plot the number of genes above TPM threshold=10 and threshold=20
```{r aboveTPMthreshold, echo=FALSE}
aboveTPM <- function(TPM_tidy, threshold=10) {
  #take in a tpm table and a tpm threshold, 
  tpm_frame <- data.frame(TPM_tidy, tpmFilter=TPM_tidy$TPM >= threshold)
  tpm_frame <- group_by(tpm_frame, Samples)
  tpm_aboveFilt <- summarize(tpm_frame, tpmFilter=sum(tpmFilter))
}

aboveTPM10 <- aboveTPM(TPMs_tidy, threshold=10)
aboveTPM20 <- aboveTPM(TPMs_tidy, threshold=20)


aboveTPM10_plot <- ggplot(aboveTPM10, aes(x=Samples, y=tpmFilter, fill=Samples))+geom_col()+ggtitle("TPM above 10")+JF_theme+xlab("")+theme(legend.position = "none")+theme(axis.text.x = element_text(angle = 45, hjust=1))+ylab("# genes above threshold")
aboveTPM20_plot <- ggplot(aboveTPM20, aes(x=Samples, y=tpmFilter, fill=Samples))+geom_col()+ggtitle("TPM above 20")+JF_theme+theme(legend.position = "none")+theme(axis.text.x = element_text(angle = 45, hjust=1))+ylab("# genes above threshold")+xlab("")

aboveTPM10_plot
aboveTPM20_plot
```

###Plot the distribtion of rawCounts in each sample:
```{r plotRawDistros, echo=FALSE}
rawDensity <- ggplot(featureCounts_tidy,aes(log10(counts+1),fill=Samples ))+geom_histogram(alpha=0.3,position="identity",bins=100)+JF_theme+theme(legend.position = c(0.75,0.75))+ggtitle("rawCounts_density")
rawDensity
```


###Plot the number of genes in each sample above rawCounts=10, 20, 40
```{r aboveRawThreshold, echo=FALSE}
aboveRaw <- function(FC_tidy, threshold=10) {
  #take in a featureCounts table and a raw threshold, 
  raw_frame <- data.frame(FC_tidy, rawFilter=FC_tidy$counts >= threshold)
  raw_frame <- group_by(raw_frame, Samples)
  raw_aboveFilt <- summarize(raw_frame, rawFilter=sum(rawFilter))
}

aboveRaw10 <- aboveRaw(featureCounts_tidy, threshold=10)
aboveRaw20 <- aboveRaw(featureCounts_tidy, threshold=20)
aboveRaw40 <- aboveRaw(featureCounts_tidy, threshold=40)

aboveRaw10_plot <- ggplot(aboveRaw10, aes(x=Samples, y=rawFilter, fill=Samples))+geom_col()+ggtitle("Raw above 10")+JF_theme+theme(legend.position = "none")+theme(axis.text.x = element_text(angle = 45, hjust=1))+ylab("# genes above threshold")+xlab("")
aboveRaw20_plot <- ggplot(aboveRaw20, aes(x=Samples, y=rawFilter, fill=Samples))+geom_col()+ggtitle("Raw above 20")+JF_theme+theme(legend.position = "none")+theme(axis.text.x = element_text(angle = 45, hjust=1))+ylab("# genes above threshold")+xlab("")
aboveRaw40_plot <- ggplot(aboveRaw40, aes(x=Samples, y=rawFilter, fill=Samples))+geom_col()+ggtitle("Raw above 40")+JF_theme+theme(legend.position = "none")+theme(axis.text.x = element_text(angle = 45, hjust=1))+ylab("# genes above threshold")+xlab("")

aboveRaw10_plot
aboveRaw20_plot
aboveRaw40_plot
```

###Code to filter a TPM-table either for genes where 1) all samples above a TPM-based filter (filterTPMs)
###Or 2) genes above a rawCount threshold (defult=10) in a given fraction of samples (default=1/4).
```{r filter_functions}
filterTPMs <- function(TPMs, threshold=10) {
  TPMs_filt <- filter_if(TPMs, is.double, all_vars(. > threshold))
}

filterRaw <- function(FC_tidy, threshold=10, frac_samples=1/4, experimentTypes, TPMs) {
  raw_frame <- data.frame(FC_tidy, rawFilter=FC_tidy$counts >= threshold)
  #First, filter for all samples where a gene passed the raw filter. Then, count the number of instances
  #Finally, filter for Genes where the number of samples exceeds the cutoff (where gene is above threshold in n/2-1 samples). Then filter the rows of the TPM matrix for gene IDs matching the genes passing the filter.
  
  filt_frame <- filter(raw_frame, rawFilter==TRUE)
  filt_frame <- filt_frame %>%
    group_by(Geneid) %>%
    add_count(Geneid)
  
  # raw_frame <- group_by(raw_frame, Geneid) %>%
  #   add_count(rawFilter)
  sampNames=experimentTypes$samps
  numSamps <- length(sampNames)
  sampCutoff <- frac_samples * numSamps
  GenesOverFilt <- filter(filt_frame, n >= sampCutoff)
  TPMs_filt <- filter(TPMs, Geneid %in% GenesOverFilt$Geneid)
}

```

###Calculate and plot Pearson's R correlation coefficient for all genes above a certain raw threshold (here, 10).
###In a certain fraction of samples (here, 1/4 of samples)
###Leave mRNA counts and date in the report, remove for the figure version for clarity.
```{r TPMCorrelationHeatmap}
TPMs_filtered <- filterRaw(FC_tidy=featureCounts_tidy,threshold=10, frac_samples=1/4, experimentTypes = experimentTypes, TPMs=TPMs)
numGenesAboveThreshold <- dim(TPMs_filtered)[1]
numGenesAboveThreshold

cormat <- cor(dplyr::select((TPMs_filtered), -Geneid))
experimentTypesPlus_mRNACounts <- merge(experimentTypes, total_mRNA_counts, by="Samples")
anno <- dplyr::select(experimentTypesPlus_mRNACounts, -samps)
rownames(anno) <- experimentTypesPlus_mRNACounts$samps
TPM_cor_pheatmap <- pheatmap(cormat, display_numbers = T, color=colorRampPalette(brewer.pal(n=9, name="YlOrRd"))(100), annotation_col=dplyr::select(anno, -Samples), treeheight_row = 10, treeheight_col = 10, show_colnames = F, labels_col = anno$Samples, show_rownames = F, fontsize=7)
TPM_cor_pheatmap

anno <- dplyr::select(experimentTypesPlus_mRNACounts, -samps, -date, -mRNA_counts)
rownames(anno) <- experimentTypesPlus_mRNACounts$samps
TPM_cor_pheatmap_simple <- pheatmap(cormat, display_numbers = T, color=colorRampPalette(brewer.pal(n=9, name="YlOrRd"))(100), annotation_col=dplyr::select(anno, -Samples), treeheight_row = 10, treeheight_col = 10, show_colnames = F, labels_col = anno$Samples, show_rownames = F, fontsize=7)
TPM_cor_pheatmap_simple


anno <- dplyr::select(experimentTypesPlus_mRNACounts, -samps, -date, -mRNA_counts, -genotype)
rownames(anno) <- experimentTypesPlus_mRNACounts$samps
TPM_cor_pheatmap_simpler <- pheatmap(cormat, display_numbers = F, color=colorRampPalette(brewer.pal(n=9, name="YlOrRd"))(100), annotation_col=dplyr::select(anno, -Samples,-cells,-rep), treeheight_row = 10, treeheight_col = 10, show_colnames = F, labels_col = anno$Samples, show_rownames = F, fontsize=7)
TPM_cor_pheatmap_simpler


```

#Filter out samples with fewer than 150000 mRNA-aligned reads (rounded to nearest 1000)
```{r TPM_heatmap_and_PCA_sample_filtering}
read_threshold=149500
featureCounts_tidy_sampFilt <- merge(featureCounts_tidy, total_mRNA_counts, by="Samples") %>%
  filter(mRNA_counts > read_threshold )

experimentTypes_sampFilt <- filter(experimentTypesPlus_mRNACounts, mRNA_counts > read_threshold)

featureCounts_tidy_sampFilt_forTPMcalc <- dplyr::select(featureCounts_tidy_sampFilt, -mRNA_counts)
TPMs_sampFilt <- calcTPM(featureCounts_tidy_sampFilt_forTPMcalc)

TPMs_sampFilt <- filterRaw(FC_tidy=featureCounts_tidy_sampFilt,threshold=20, frac_samples=1/5, experimentTypes = experimentTypes_sampFilt, TPMs=TPMs_sampFilt)
numGenesAboveThreshold <- dim(TPMs_sampFilt)[1]
numGenesAboveThreshold

cormat_sampFilt <- cor(dplyr::select((TPMs_sampFilt), -Geneid))

anno <- dplyr::select(experimentTypes_sampFilt, -samps, -date, -mRNA_counts, -genotype,-rep)
rownames(anno) <- experimentTypes_sampFilt$samps
TPM_cor_pheatmap_sampFilt <- pheatmap(cormat_sampFilt, display_numbers = T, color=colorRampPalette(brewer.pal(n=9, name="YlOrRd"))(100), annotation_col=dplyr::select(anno, -Samples), treeheight_row = 10, treeheight_col = 10, show_colnames = F, labels_row = anno$Samples, show_rownames = F, fontsize=7)
TPM_cor_pheatmap_sampFilt


TPMs_sampFilt.pca <- prcomp(t(dplyr::select((TPMs_sampFilt), -Geneid)))
TPMs_sampFilt.pca_X_tidy <- data.frame(samps=rownames(TPMs_sampFilt.pca$x), TPMs_sampFilt.pca$x)
TPMs_sampFilt.pca_X_tidy <- merge(TPMs_sampFilt.pca_X_tidy, experimentTypes_sampFilt, by="samps" )

TPMs_sampFilt.pca.var <- TPMs_sampFilt.pca$sdev^2
TPMs_sampFilt.pca.var <- round(TPMs_sampFilt.pca.var/sum(TPMs_sampFilt.pca.var)*100,1)

TPMs_sampFilt.PC1v2_plot <- ggplot(TPMs_sampFilt.pca_X_tidy, aes(x=PC1, y=PC2, color=subtype, shape=exp))+geom_point()+xlab(paste("PC1- ",TPMs_sampFilt.pca.var[1],"%", sep=""))+ylab(paste("PC2- ",TPMs_sampFilt.pca.var[2],"%", sep=""))
TPMs_sampFilt.PC1v2_plot
TPMs_sampFilt.PC2v3_plot <- ggplot(TPMs_sampFilt.pca_X_tidy, aes(x=PC2, y=PC3, color=subtype, shape=exp))+geom_point()+xlab(paste("PC2- ",TPMs_sampFilt.pca.var[2],"%", sep=""))+ylab(paste("PC3- ",TPMs_sampFilt.pca.var[3],"%", sep=""))
TPMs_sampFilt.PC2v3_plot

TPMs_sampFilt_tidy <- gather(TPMs_sampFilt, "samps","TPM",-Geneid)
TPMs_sampFilt_tidy <- merge(TPMs_sampFilt_tidy, experimentTypes_sampFilt, by="samps")

TPMs_sampFilt_vioPlot <- ggplot(TPMs_sampFilt_tidy, aes(x=Samples, y=log10(TPM+1), fill=subtype))+geom_violin()+facet_wrap(~exp, scales="free_x") 
TPMs_sampFilt_vioPlot
```

###Make the counts matrix for DESeq2.
```{r makeCountsMatrixForDESeq2}

DESeq2_counts_matrix <- featureCounts_tidy_sampFilt %>%
    dplyr::select(samps, Geneid, counts) %>%
    spread("samps","counts") %>%
    filter(Geneid %in% TPMs_sampFilt$Geneid)

rownames(DESeq2_counts_matrix) <- DESeq2_counts_matrix$Geneid
DESeq2_counts_matrix <- dplyr::select(DESeq2_counts_matrix, -Geneid)
colData <- dplyr::select(experimentTypes_sampFilt, samps, exp,subtype)
colData <- dplyr::arrange(colData, samps)

rownames(colData) <- colData$samps
colData <- dplyr::select(colData, -samps)

runDESeq_bySubtype <- function(FC) {
  DESeq2_counts_matrix <- FC %>%
    dplyr::select(samps, Geneid, counts) %>%
    spread("samps","counts")
  rownames(DESeq2_counts_matrix) <- DESeq2_counts_matrix$Geneid
  DESeq2_counts_matrix <- dplyr::select(DESeq2_counts_matrix, -Geneid)
  
  # colData <- filter(metaData, samps %in% FC$samps & region %in% regions & stage %in% stages & strain %in% strains) %>% dplyr::select(samps, stage, strain, region)
  colData <- FC %>%
    group_by(samps, exp, subtype) %>%
    summarize(n=n()) %>%
    dplyr::select(-n) %>%
    dplyr::arrange(samps)
  colData <- data.frame(colData)
rownames(colData) <- colData$samps
colData <- dplyr::select(colData, -samps)
colData <- droplevels(colData)

dds <- DESeqDataSetFromMatrix(DESeq2_counts_matrix, colData = colData, design=~subtype)
dds <- DESeq(dds)
}

test <- runDESeq_bySubtype(featureCounts_tidy_sampFilt)

```

###Run DESeq2, including an exp:subtype interaction term. The genes with this 
###interaction term significant are the ones differentially translated. 
###However, this includes exclusively translated genes, and buffered genes.
###Need to run DESeq2 separately on the AF (RNA) and RP (Ribo-Seq) data sets
###To be able to parse out translational regulation from buffering.
###Update 4/21/21: running one full model: ~exp+subtype+exp:subtype
###The main effect of exp is translation efficiency (TE) in CPN
###The main effect of exp+exp:subtype interaction is SCPN TE
###The main effect of subtype is the RP/AF ratio in CPN
###The main effect of subtype+exp:subtype interaction is RP/AF in SCPN
###Awkwardly renaming the results objects to match previous names
###Using genes that are significant in the subtype main effect as "RNA"
###and genes significant in the exp comparison as "Ribo". Significant in both
###is "RNA+Ribo"
###Write out the Geneids and log2FC for pre-ranked GSEA analysis
```{r DESeq2}
sig=0.05
TE_sig=0.1
dds <- DESeqDataSetFromMatrix(countData = DESeq2_counts_matrix, colData = colData, design=~exp+subtype+subtype:exp)
dds$exp <- factor(dds$exp, levels=c("AF","RP"))
dds$subtype <- factor(dds$subtype, levels=c("CPN","SCPN"))
dds <- DESeq(dds)
res_diffTE <- results(dds, alpha=TE_sig, name="expRP.subtypeSCPN", independentFiltering =T)


#main effect of exp, RP vs AF in CPN
res_TE_CPN <- results(dds, contrast=c("exp", "RP","AF" ), alpha=sig)
#exp effect for SCPN
res_TE_SCPN <- results(dds, contrast=list(c("exp_RP_vs_AF", "expRP.subtypeSCPN" )), alpha=sig)
#effect of subtype in AF, RP
res_subtype_AF <- results(dds, contrast=c("subtype","SCPN","CPN"), alpha=sig)
res_subtype_RP <- results(dds, contrast=list(c("subtype_SCPN_vs_CPN","expRP.subtypeSCPN")), alpha=sig)

res_AF_df <- data.frame(res_subtype_AF, contrast="mRNA", Geneid=rownames(res_subtype_AF)) %>% mutate(mRNA_sig=case_when(padj < sig & log2FoldChange < 0 ~ "lower", padj < sig & log2FoldChange > 0 ~ "higher", padj > sig ~ "not significant"))
res_RP_df <- data.frame(res_subtype_RP, contrast="RPF", Geneid=rownames(res_subtype_RP)) %>% mutate(RP_sig=case_when(padj < sig & log2FoldChange < 0 ~ "lower", padj < sig & log2FoldChange > 0 ~ "higher", padj > sig ~ "not significant"))
res_diffTE_df <- data.frame(res_diffTE, contrast="subtypeTE", Geneid=rownames(res_diffTE)) %>% mutate(diffTE_sig=case_when(padj < TE_sig & log2FoldChange < 0 ~ "lower", padj < TE_sig & log2FoldChange > 0 ~ "higher", padj > TE_sig ~ "not significant"))
res_TE_CPN_df <- data.frame(res_TE_CPN, contrast="TE_CPN", Geneid=rownames(res_TE_CPN)) %>% mutate(TE_CPN_sig=case_when(padj < sig & log2FoldChange < 0 ~ "lower", padj < sig & log2FoldChange > 0 ~ "higher", padj > sig ~ "not significant"))
res_TE_SCPN_df <- data.frame(res_TE_SCPN, contrast="TE_SCPN", Geneid=rownames(res_TE_SCPN)) %>% mutate(TE_SCPN_sig=case_when(padj < sig & log2FoldChange < 0 ~ "lower", padj < sig & log2FoldChange > 0 ~ "higher", padj > sig ~ "not significant"))

mRNA_MAplot <- ggplot(filter(res_AF_df, is.na(mRNA_sig)==F), aes(y=log2FoldChange,x=log10(baseMean), color=mRNA_sig))+JF_theme+geom_point(aes=0.5, size=0.2)+scale_color_manual(values=c("red","blue","grey"))+ylim(-6,6)+theme(legend.position = c(1 , 0.8), legend.key.size = unit(0.5,"line"))
RPF_MAplot <- ggplot(filter(res_RP_df, is.na(RP_sig)==F) , aes(y=log2FoldChange,x=log10(baseMean), color=RP_sig))+JF_theme+geom_point(aes=0.5, size=0.2)+scale_color_manual(values=c("red","blue","grey"))+ylim(-6,6)+theme(legend.position = c(1, 0.8), legend.key.size = unit(0.5,"line"))
TE_MAplot <- ggplot(filter(res_diffTE_df, is.na(diffTE_sig)==F) , aes(y=log2FoldChange,x=log10(baseMean), color=diffTE_sig))+JF_theme+geom_point(aes=0.5, size=0.2)+scale_color_manual(values=c("red","blue","grey"))+ylim(-6,6)+theme(legend.position = c(1, 0.8), legend.key.size = unit(0.5,"line"))

subtypeComparison_res_df <- bind_rows(res_AF_df, res_RP_df, res_diffTE_df, res_TE_CPN_df, res_TE_SCPN_df) 
subtypeComparison_res_df  <- subtypeComparison_res_df %>% left_join(mouse_gene_symbols_wGeneid, by="Geneid")
RPvsAF_df <- subtypeComparison_res_df %>% filter(contrast %in% c("mRNA","RPF") & is.na(padj)==F) %>% dplyr::select(Geneid, log2FoldChange, contrast) %>% spread(key="contrast",value=log2FoldChange) %>% left_join(res_diffTE_df, by="Geneid") %>% dplyr::select(Geneid, mRNA, RPF, diffTE_sig)
RPvsAF_scatterplot <- ggplot(RPvsAF_df, aes(x=mRNA,y=RPF))+JF_theme+geom_point(alpha=0.5, size=0.2)+geom_hline(yintercept=0)+geom_vline(xintercept=0)+xlab("mRNA log2FoldChange (SCPN/CPN)")+ylab("RPF log2FoldChange (SCPN/CPN)")+ggtitle(paste("Pearson's R=",round(cor(RPvsAF_df$mRNA,RPvsAF_df$RPF,method="pearson"),3)))+theme(legend.position = c(0.75, 0.75))
RPvsAF_scatterplot 

TE_CPNvsSCPN_df <- subtypeComparison_res_df %>% filter(contrast %in% c("TE_CPN","TE_SCPN") & is.na(padj)==F) %>% dplyr::select(Geneid, log2FoldChange, contrast) %>% spread(key="contrast",value=log2FoldChange) %>% left_join(res_diffTE_df, by="Geneid") %>% dplyr::select(Geneid, TE_CPN, TE_SCPN, diffTE_sig) %>% left_join(dplyr::select(res_TE_SCPN_df, Geneid, TE_SCPN_sig), by="Geneid") %>% left_join(dplyr::select(res_TE_CPN_df, Geneid, TE_CPN_sig), by="Geneid")

TE_CPNvsSCPN_scatterplot <- ggplot(TE_CPNvsSCPN_df, aes(x=TE_CPN,y=TE_SCPN, color=diffTE_sig))+JF_theme+geom_point(alpha=0.5, size=0.2)+geom_hline(yintercept=0)+geom_vline(xintercept=0)+scale_color_manual(values=c("red","blue","black","grey"))+xlab("TE in CPN")+ylab("TE in SCPN")+ggtitle(paste("Pearson's R=",round(cor(TE_CPNvsSCPN_df$TE_CPN,TE_CPNvsSCPN_df$TE_SCPN,method="pearson"),3)))+theme(legend.position = c(0.85, 0.3), legend.key.size = unit(0.5,"line"))
TE_CPNvsSCPN_scatterplot

TE_CPNvsSCPN_scatterplot_TE_CPN_sig_colored <- ggplot(TE_CPNvsSCPN_df, aes(x=TE_CPN,y=TE_SCPN, color=TE_CPN_sig))+JF_theme+geom_point(alpha=0.5, size=0.2)+geom_hline(yintercept=0)+geom_vline(xintercept=0)+scale_color_manual(values=c("red","blue","black","grey"))+xlab("TE in CPN")+ylab("TE in SCPN")+ggtitle("T.E. in SCPN vs. T.E. in CPN, skewed TE in CPN highlighted ")+theme(legend.position = c(0.85, 0.3), legend.key.size = unit(0.5,"line"))
TE_CPNvsSCPN_scatterplot_TE_CPN_sig_colored 

TE_CPNvsSCPN_scatterplot_TE_SCPN_sig_colored <- ggplot(TE_CPNvsSCPN_df, aes(x=TE_CPN,y=TE_SCPN, color=TE_SCPN_sig))+JF_theme+geom_point(alpha=0.5, size=0.2)+geom_hline(yintercept=0)+geom_vline(xintercept=0)+scale_color_manual(values=c("red","blue","black","grey"))+xlab("TE in CPN")+ylab("TE in SCPN")+ggtitle("T.E. in SCPN vs. T.E. in CPN, skewed TE in CPN highlighted ")+theme(legend.position = c(0.85, 0.3), legend.key.size = unit(0.5,"line"))
TE_CPNvsSCPN_scatterplot_TE_SCPN_sig_colored 

log2FC_histo_plot <- filter(subtypeComparison_res_df, (contrast %in% c("TE_SCPN","TE_CPN")==F)) %>%
  ggplot(aes(x=log2FoldChange, fill=contrast))+JF_theme+geom_histogram(alpha=0.5, bins=400, position="identity")+ylab("gene count")
log2FC_histo_plot

log2FC_ecdf_plot <- filter(subtypeComparison_res_df, (contrast %in% c("TE_SCPN","TE_CPN")==F)) %>%
  ggplot(aes(x=log2FoldChange, color=contrast))+JF_theme+stat_ecdf()+ylab("cumlative_fraction")
log2FC_ecdf_plot

```

###Are AF and RP dispersions different?
###Answer: yes. dispersion correlation is 0.2857. Have to draw them from different distributions. 

```{r AFvsRP_DESeq2}

AF_dds <- runDESeq_bySubtype(filter(featureCounts_tidy_sampFilt, exp=="AF"))
RP_dds <- runDESeq_bySubtype(filter(featureCounts_tidy_sampFilt, exp=="RP"))

AF_fullDEseq <- data.frame(mcols(AF_dds), exp="AF", Geneid=rownames(mcols(AF_dds)))
RP_fullDEseq <- data.frame(mcols(RP_dds),exp="RP", Geneid=rownames(mcols(RP_dds))) 
AF_RP_full_DEseq <- bind_rows(RP_fullDEseq, AF_fullDEseq)  %>% filter(Geneid %in% subtypeComparison_res_df$Geneid)

AF_RP_distGeneEst <- AF_RP_full_DEseq %>% dplyr::select(Geneid, dispGeneEst, exp) %>% spread(key=exp, value=dispGeneEst)
AFvsRP_dispGeneEstPlot <- ggplot(AF_RP_distGeneEst, aes(x=AF,y=RP))+JF_theme+geom_point()+xlim(c(0,2.5))+ylim(c(0,2.5))

AF_RP_dispersions <- AF_RP_full_DEseq %>% dplyr::select(Geneid, dispersion, exp) %>% spread(key=exp, value=dispersion)
AFvsRP_dispersionPlot <- ggplot(AF_RP_dispersions, aes(x=AF,y=RP))+JF_theme+geom_point()

```

###Make AF and RP sim data sets using the estimated dispersion parameters.
###Deep knowledge here: baseMean represents the mean counts per sample after normalizing for libSize!
###Also size=1/dispersion, where dispersion is the DESeq2 dispersion. 
###Got 3/9967 diff exp genes at padj < 0.1 for simulated AF data.

```{r simulate_using_DESeq2}
AF_fullDEseq_filt <- AF_fullDEseq %>% filter(Geneid %in% subtypeComparison_res_df$Geneid)

drawFromNbinom <- function(baseMean, dispersion,Geneid, n=8) {
  #returnList=list("bM"=baseMean,"disp"=dispersion)
  draws <- rnbinom(n=n, size=1/dispersion, mu=baseMean)
  return(c(Geneid, draws))
}

simFrom_fullDESeq <- function(fullDESeq, seed=1182, cond_n=4) {
  set.seed(seed)
  sim <- mapply(drawFromNbinom, fullDESeq$baseMean, fullDESeq$dispersion, fullDESeq$Geneid, cond_n*2)
  sim <- t(sim)
  simSampNames <- c(paste(paste("S",seq(1:cond_n),sep=""),"n1",sep="_")
    ,paste(paste("S",seq(1:cond_n),sep=""),"n2",sep="_")) 
  colnames(sim) <- c("Geneid",simSampNames)
  rownames(sim) <- sim[,1]
  sim <- matrix(as.numeric(sim[,-1]), ncol=ncol(sim[,-1]), dimnames=list(rownames(sim),colnames(sim[,-1])))
  sim <- data.frame(sim)
}

AF_sim <- simFrom_fullDESeq(fullDESeq=AF_fullDEseq_filt, cond_n=4)
AF_sim_DEseq_res  <- DESeq2_RNAonly_runme(test)

set.seed(62189)

AF_sim <- mapply(drawFromNbinom, AF_fullDEseq_filt$baseMean, AF_fullDEseq_filt$dispersion, AF_fullDEseq_filt$Geneid)
AF_sim <- t(AF_sim)
sim_sampNames <- c("S1_n1","S2_n1","S3_n1","S4_n1","S5_n2","S6_n2","S7_n2","S8_n2")
colnames(AF_sim) <- c("Geneid", sim_sampNames)
rownames(AF_sim) <- AF_sim[,1]
AF_sim <- matrix(as.numeric(AF_sim[,-1]), ncol=ncol(AF_sim[,-1]), dimnames=list(rownames(AF_sim),colnames(AF_sim[,-1])))
AF_sim <- data.frame(AF_sim)

#AF_sim <- data.frame(as.numeric(AF_sim[,-1]))

DESeq2_RNAonly_runme <- function(countsData) {
  colData <- data.frame(condition=str_extract(colnames(countsData), regex("n.")))
  #return(colData)
  rownames(colData) <- colnames(countsData)
  dds <- DESeqDataSetFromMatrix(countData = countsData, colData = colData, design=~condition)
  dds <- DESeq(dds)
  res <- data.frame(results(dds))
}
sig_threshold <- 0.1

AF_sim_DEseq_res <- DESeq2_RNAonly_runme(AF_sim)


```



```{r enrichGO}
rna_universe <- subtypeComparison_res_df %>%
 dplyr::select(Geneid) %>% distinct()
rna_universe <- rna_universe$Geneid


run_enrichGO <- function(genesOfInterest, gene_uni=rna_universe, min_count=15) {
  GO_res <- enrichGO(gene=genesOfInterest, universe=gene_uni,  OrgDb = "org.Mm.eg.db", keyType = "ENSEMBL", ont = "ALL", pAdjustMethod = "BH", readable = T, pool = F,) %>% data.frame() %>% filter(Count > min_count) %>% separate(GeneRatio, into = c("Significant", "denom")) %>%  mutate(GeneRatio = as.numeric(Significant) / as.numeric(denom)) %>% separate(BgRatio, into=c("notSignificant","denom")) %>% mutate(BgRatio=as.numeric(notSignificant)/as.numeric(denom))
}

run_enrichGO_raw_output <- function(genesOfInterest, gene_uni=rna_universe, min_count=15) {
  #same as run_enrichGO, but does not convert the output to a dataframe. 
  GO_res <- enrichGO(gene=genesOfInterest, universe=gene_uni,  OrgDb = "org.Mm.eg.db", keyType = "ENSEMBL", ont = "ALL", pAdjustMethod = "BH", readable = T, pool = F,)
}

SCPNvsCPN_TE_higher <- (subtypeComparison_res_df %>% filter(diffTE_sig == "higher"))$Geneid
SCPNvsCPN_TE_lower <- (subtypeComparison_res_df %>% filter(diffTE_sig == "lower"))$Geneid
diffTE_genes <- union(SCPNvsCPN_TE_higher, SCPNvsCPN_TE_lower)
CPN_TE_high <- (subtypeComparison_res_df %>% filter(TE_CPN_sig == "higher"))$Geneid
SCPN_TE_high <- (subtypeComparison_res_df %>% filter(TE_SCPN_sig == "higher"))$Geneid
CPN_TE_low <- (subtypeComparison_res_df %>% filter(TE_CPN_sig == "lower"))$Geneid
SCPN_TE_low <- (subtypeComparison_res_df %>% filter(TE_SCPN_sig == "lower"))$Geneid
CPN_TE_notSig <- (subtypeComparison_res_df %>% filter(contrast=="TE_CPN" & (TE_CPN_sig %in% c("higher","lower"))==F))$Geneid
SCPN_TE_notSig <- (subtypeComparison_res_df %>% filter(contrast=="TE_SCPN" & (TE_SCPN_sig %in% c("higher","lower"))==F))$Geneid


SCPN_mRNA <- (subtypeComparison_res_df %>% filter(mRNA_sig == "higher"))$Geneid
CPN_mRNA <- (subtypeComparison_res_df %>% filter(mRNA_sig == "lower"))$Geneid
SCPN_RPF <- (subtypeComparison_res_df %>% filter(RP_sig == "higher"))$Geneid
CPN_RPF <- (subtypeComparison_res_df %>% filter(RP_sig == "lower"))$Geneid

CPN_SCPN_TE_upset_list <- list("CPN_high"=CPN_TE_high, "CPN_low"=CPN_TE_low, "SCPN_high"=SCPN_TE_high, "SCPN_low"=SCPN_TE_low)
CPN_SCPN_TE_upset <- upset(fromList(CPN_SCPN_TE_upset_list), order.by="freq")
CPN_SCPN_TE_upset 

SCPN_CPN_AF_RP_upset_list <- list("SCPN_mRNA"=SCPN_mRNA, "CPN_mRNA"=CPN_mRNA, "SCPN_RPF"=SCPN_RPF, "CPN_RPF"=CPN_RPF)
SCPN_CPN_AF_RP_upset <- upset(fromList(SCPN_CPN_AF_RP_upset_list), order.by="freq")
SCPN_CPN_AF_RP_upset 

diffTE_upset_list <- list("SCPN_TE"=SCPNvsCPN_TE_higher, "CPN_TE"=SCPNvsCPN_TE_lower, "SCPN_mRNA"=SCPN_mRNA, "CPN_mRNA"=CPN_mRNA, "SCPN_RPF"=SCPN_RPF, "CPN_RPF"=CPN_RPF )
diffTE_upset <- upset(fromList(diffTE_upset_list), keep.order = T, nsets=length(diffTE_upset_list))
diffTE_upset 
diffTE_focus_list <- list(list("SCPN_TE","SCPN_mRNA"),list("SCPN_TE","SCPN_RPF"),
                          list("SCPN_TE","CPN_mRNA"),list("SCPN_TE","CPN_RPF"),
                          list("CPN_TE","CPN_mRNA"),list("CPN_TE","CPN_RPF"),
                          list("CPN_TE","SCPN_mRNA"),list("CPN_TE","SCPN_RPF"),
                          list("CPN_TE","SCPN_mRNA","CPN_RPF"), list("SCPN_TE","SCPN_RPF","SCPN_mRNA"),list("CPN_TE","CPN_RPF","CPN_mRNA"), list("SCPN_TE","CPN_mRNA","SCPN_RPF"))
diffTE_upset_TE_focus <- upset(fromList(diffTE_upset_list), order.by = "freq",intersections=diffTE_focus_list)


subtypeComp_GO_list <- list("SCPN_vsCPN_TEhigher"=SCPNvsCPN_TE_higher,"SCPN_vsCPN_TElower"=SCPNvsCPN_TE_lower,
                            "CPN_TE_high"=CPN_TE_high, "SCPN_TE_high"=SCPN_TE_high,
                            "CPN_TE_low"=CPN_TE_low, "SCPN_TE_low"=SCPN_TE_low,
                            "diffTE"=diffTE_genes,
                            "SCPN_mRNA"=SCPN_mRNA, "CPN_mRNA"=CPN_mRNA, 
                            "SCPN_RPF"=SCPN_RPF, "CPN_RPF"=CPN_RPF)

ego_subtypes <- lapply(subtypeComp_GO_list, run_enrichGO)
ego_subtypes_raw <- lapply(subtypeComp_GO_list, run_enrichGO_raw_output)

names(ego_subtypes ) <- names(ego_subtypes)

plot_ego <- function(ego, numToPlot=5, onto=c("BP","CC","MF"), title="") {
  ego_filt <- ego %>% filter(ONTOLOGY %in% onto) %>% group_by(ONTOLOGY) %>% slice_min(n=numToPlot, order_by=qvalue)
  ego_filt$Description <- factor(ego_filt$Description, levels=ego_filt$Description)
  ego_plot <- ggplot(ego_filt, aes(x=GeneRatio,y=Description, color=-log10(qvalue),size=Count))+JF_theme+geom_point()+facet_wrap(~ONTOLOGY, nrow=3, scales="free_y")+ggtitle(title)
}

ego_subtypes_plots <- lapply(ego_subtypes, plot_ego)
ego_subtypes_plots_BP_CC <- lapply(ego_subtypes, plot_ego, onto=c("BP","CC"))

ego_subtypes_barplots <- lapply(ego_subtypes_raw, barplot, showCategory=5, font.size=7, label_format=16)

TE_CPN_higher_barplot <- ego_subtypes_barplots$CPN_TE_high + ggtitle("TE CPN, skew higher")+JF_theme+theme(legend.position = "top")
TE_CPN_higher_barplot
TE_SCPN_higher_barplot <- ego_subtypes_barplots$SCPN_TE_high + ggtitle("TE SCPN, skew higher")+JF_theme+theme(legend.position = "top")
TE_SCPN_higher_barplot
TE_CPN_lower_barplot <- ego_subtypes_barplots$CPN_TE_low + ggtitle("TE CPN, skew lower")+JF_theme+theme(legend.position = "top")
TE_CPN_lower_barplot
TE_SCPN_lower_barplot <- ego_subtypes_barplots$SCPN_TE_low + ggtitle("TE SCPN, skew lower")+JF_theme+theme(legend.position = "top")
TE_SCPN_lower_barplot
###Tried treeplot, ran into dependency issues.


```

```{r subtypesFigures_layout}

###Need to manually put in the upset plots as these are not ggplots. 


subtypesComp_Figure_upsetSpacers <- as.ggplot(TPM_cor_pheatmap_simpler) / (RPvsAF_scatterplot | mRNA_MAplot | RPF_MAplot ) / (TE_MAplot + TE_CPNvsSCPN_scatterplot) /(plot_spacer()+plot_spacer())+theme(text=element_text("ArialMT")) + plot_annotation(tag_levels = list(c("A","B","C","D","E","F","G","H"))) & theme(plot.tag=element_text(size=12, face="bold"))+theme(legend.key.size=unit(0.5,"line"))

pdfName=paste(outDir, "SubtypeComparison_fig.pdf", sep="/")
pdf(file=pdfName, width=6.5, height=9)
plot(subtypesComp_Figure_upsetSpacers)
dev.off()



withinSubtypeTE <- (plot_spacer()| TE_CPNvsSCPN_scatterplot_TE_CPN_sig_colored | TE_CPNvsSCPN_scatterplot_TE_SCPN_sig_colored) /(TE_CPN_higher_barplot + TE_CPN_lower_barplot) / (TE_SCPN_higher_barplot + TE_SCPN_lower_barplot) / (plot_spacer() | plot_spacer()) /(plot_spacer() | plot_spacer())

pdfName=paste(outDir, "withinSubtypeTE_fig.pdf", sep="/")
pdf(file=pdfName, width=6.5, height=9)
plot(withinSubtypeTE)
dev.off()

```
```{r uORFs}
CPN_ORFs_file="CPN_RiboCode/RiboCode_ORFs_collapsed.txt"
SCPN_ORFs_file="SCPN_RiboCode/RiboCode_ORFs_collapsed.txt"

CPN_ORFs_res_collapsed <- read.table(file=CPN_ORFs_file, sep="\t", header=T) %>% mutate(subtype="CPN")
SCPN_ORFs_res_collapsed <- read.table(file=SCPN_ORFs_file, sep="\t", header=T) %>% mutate(subtype="SCPN")
ORF_res <- bind_rows(CPN_ORFs_res_collapsed, SCPN_ORFs_res_collapsed)

#label novel uORFs mapping to pseudogenes
novelORF_in_pseudogene <- ORF_res$ORF_type=="novel" & (ORF_res$transcript_type %in% c("polymorphic_pseudogene","processed_pseudogene","pseudogene","transcribed_processed_pseudogene","transcribed_unitary_pseudogene","transcribed_unprocessed_pseudogene","translated_processed_pseudogene","unprocessed_pseudogene"))==T
ORF_res[novelORF_in_pseudogene==T,]$ORF_type <- "pseudogene" 

#plot the counts of ORF types by subtype
ORF_type_counts <- ORF_res %>% group_by(subtype, ORF_type) %>% summarize(ORF_count=n())

uORF_res <- filter(ORF_res, ORF_type=="uORF" | ORF_type=="Overlap_uORF")

ORF_types_plot <- ggplot(ORF_type_counts, aes(x=ORF_type, y=ORF_count, fill=ORF_type, label=ORF_count))+JF_theme+geom_col()+geom_text()+facet_wrap(~subtype)+theme(axis.text.x = element_text(angle = 45, hjust=1))+theme(legend.position = "none")


###Overlap between CPN and SCPN is a bit lower than I expected. Tried re-runing RiboCode with a more stringent pval (0.01 vs 0.05)
###Still low. I wonder a bit if this might be because a decent % of uORFs are in genes that are differentially expressed between subtypes.

uORF_ID_list <- list("CPN_uORF"=filter(ORF_res, ORF_type=="uORF" & subtype=="CPN")$ORF_ID,
                     "CPN_Overlap_uORF"=filter(ORF_res, ORF_type=="Overlap_uORF" & subtype=="CPN")$ORF_ID,
                     "SCPN_uORF"=filter(ORF_res, ORF_type=="uORF" & subtype=="SCPN")$ORF_ID,
                     "SCPN_Overlap_uORF"=filter(ORF_res, ORF_type=="Overlap_uORF" & subtype=="SCPN")$ORF_ID)
uORF_ID_upset <- upset(fromList(uORF_ID_list),order.by = "freq")
uORF_ID_upset
###Hypothesis wrong. Even among genes that are not diff exp, only about 33% of uORFs are shared. 
uORF_ID_notdiffExp_list <- list("CPN_uORF"=filter(ORF_res, ORF_type=="uORF" & subtype=="CPN" & (gene_id %in% c(subtypeComp_GO_list$CPN_mRNA, subtypeComp_GO_list$SCPN_mRNA))==F)$ORF_ID,
                     "CPN_Overlap_uORF"=filter(ORF_res, ORF_type=="Overlap_uORF" & subtype=="CPN" & (gene_id %in% c(subtypeComp_GO_list$CPN_mRNA, subtypeComp_GO_list$SCPN_mRNA))==F)$ORF_ID,
                     "SCPN_uORF"=filter(ORF_res, ORF_type=="uORF" & subtype=="SCPN" & (gene_id %in% c(subtypeComp_GO_list$CPN_mRNA, subtypeComp_GO_list$SCPN_mRNA))==F)$ORF_ID,
                     "SCPN_Overlap_uORF"=filter(ORF_res, ORF_type=="Overlap_uORF" & subtype=="SCPN" & (gene_id %in% c(subtypeComp_GO_list$CPN_mRNA, subtypeComp_GO_list$SCPN_mRNA))==F)$ORF_ID)
uORF_ID_notdiffExp_upset <- upset(fromList(uORF_ID_notdiffExp_list),order.by = "freq")
uORF_ID_notdiffExp_upset



uORF_transcriptID_list <- list("CPN_uORF"=filter(ORF_res, ORF_type=="uORF" & subtype=="CPN")$transcript_id,
                     "CPN_Overlap_uORF"=filter(ORF_res, ORF_type=="Overlap_uORF" & subtype=="CPN")$transcript_id,
                     "SCPN_uORF"=filter(ORF_res, ORF_type=="uORF" & subtype=="SCPN")$transcript_id,
                     "SCPN_Overlap_uORF"=filter(ORF_res, ORF_type=="Overlap_uORF" & subtype=="SCPN")$transcript_id)
uORF_transcriptID_upset <- upset(fromList(uORF_transcriptID_list))
uORF_transcriptID_upset

uORF_geneID_list <- list("CPN_uORF"=filter(ORF_res, ORF_type=="uORF" & subtype=="CPN")$gene_id,
                     "CPN_Overlap_uORF"=filter(ORF_res, ORF_type=="Overlap_uORF" & subtype=="CPN")$gene_id,
                     "SCPN_uORF"=filter(ORF_res, ORF_type=="uORF" & subtype=="SCPN")$gene_id,
                     "SCPN_Overlap_uORF"=filter(ORF_res, ORF_type=="Overlap_uORF" & subtype=="SCPN")$gene_id)
uORF_geneID_upset <- upset(fromList(uORF_geneID_list))
uORF_geneID_upset

uORF_GO_list <- list("CPN_uORFs" = filter(ORF_res, (ORF_type=="uORF"| ORF_type=="Overlap_uORF" ) & subtype=="CPN")$gene_id,
                     "SCPN_uORFs"= filter(ORF_res, (ORF_type=="uORF"| ORF_type=="Overlap_uORF" ) & subtype=="SCPN")$gene_id)
uORF_ego_raw <- lapply(uORF_GO_list, run_enrichGO_raw_output)

###See the strong enrichment for "synapse and structure" genes seen previously.
uORF_ego_barplot <- lapply(uORF_ego_raw, barplot, showCategory=5,font.size=7, label_format=24)
uORF_ego_barplot$CPN_uORFs <- uORF_ego_barplot$CPN_uORFs+theme(legend.position = "top")+ggtitle("top GO terms, CPN")+JF_theme
uORF_ego_barplot$SCPN_uORFs <- uORF_ego_barplot$SCPN_uORFs+theme(legend.position = "top")+ggtitle("top GO terms uORFs, SCPN")+JF_theme



subtypeComparison_res_df_plus_uORFs <- subtypeComparison_res_df %>%
  mutate(uORF_status=case_when(
    Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="uORF")$gene_id ~ "CPN_uORF",
    Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="Overlap_uORF")$gene_id ~ "CPN_Overlap_uORF",
    Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="uORF")$gene_id ~ "SCPN_uORF",
    Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="Overlap_uORF")$gene_id ~ "SCPN_Overlap_uORF"
  )) %>%
  mutate(uORF_status_intersect=case_when(
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="uORF")$gene_id==T) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="uORF")$gene_id==F) ~ "CPN_uORF",
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="uORF")$gene_id==F) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="uORF")$gene_id==T) ~ "SCPN_uORF",
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="uORF")$gene_id==T) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="uORF")$gene_id==T) ~ "SCPN&CPN_uORF",
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="Overlap_uORF")$gene_id==T) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="Overlap_uORF")$gene_id==F) ~ "CPN_Overlap_uORF",
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="Overlap_uORF")$gene_id==F) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="Overlap_uORF")$gene_id==T) ~ "SCPN_Overlap_uORF",
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="Overlap_uORF")$gene_id==T) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="Overlap_uORF")$gene_id==T) ~ "SCPN&CPN_Overlap_uORF",
))

res_TE_CPN_df <- res_TE_CPN_df %>% mutate(uORF_status=case_when(
  Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="uORF")$gene_id ~ "CPN_uORF",
  Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="Overlap_uORF")$gene_id ~ "CPN_Overlap_uORF",
  T ~ "no_uORF"
)) %>% mutate(uORF_simple=case_when(
  uORF_status != "no_uORF" ~ "uORF",
  uORF_status == "no_uORF" ~ "no_uORF",
))
res_TE_SCPN_df <- res_TE_SCPN_df %>% mutate(uORF_status=case_when(
  Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="uORF")$gene_id ~ "SCPN_uORF",
    Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="Overlap_uORF")$gene_id ~ "SCPN_Overlap_uORF",
  T ~ "no_uORF"
)) %>% mutate(uORF_simple=case_when(
  uORF_status != "no_uORF" ~ "uORF",
  uORF_status == "no_uORF" ~ "no_uORF",
))
res_diffTE_df <- res_diffTE_df %>% mutate(uORF_status=case_when(
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="uORF")$gene_id==T) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="uORF")$gene_id==F) ~ "CPN_uORF",
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="uORF")$gene_id==F) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="uORF")$gene_id==T) ~ "SCPN_uORF",
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="uORF")$gene_id==T) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="uORF")$gene_id==T) ~ "SCPN&CPN_uORF",
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="Overlap_uORF")$gene_id==T) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="Overlap_uORF")$gene_id==F) ~ "CPN_Overlap_uORF",
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="Overlap_uORF")$gene_id==F) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="Overlap_uORF")$gene_id==T) ~ "SCPN_Overlap_uORF",
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="Overlap_uORF")$gene_id==T) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="Overlap_uORF")$gene_id==T) ~ "SCPN&CPN_Overlap_uORF",
))


TE_CPN_uORF_plot <- ggplot(res_TE_CPN_df, aes(x=log2FoldChange, color=uORF_status))+JF_theme+stat_ecdf(size=0.1)+facet_wrap(~contrast)+xlim(c(-2,2))+ylab("cumulative_fraction")+ggtitle("T.E. by uORF status, CPN")+xlab("T.E.")+theme(legend.position="top")+geom_hline(yintercept=0.5,size=0.2)+geom_vline(xintercept=0, size=0.2)
TE_CPN_uORF_plot
TE_SCPN_uORF_plot <- ggplot(res_TE_SCPN_df, aes(x=log2FoldChange, color=uORF_status))+JF_theme+stat_ecdf(size=0.1)+facet_wrap(~contrast)+xlim(c(-2,2))+ylab("cumulative_fraction")+ggtitle("T.E. by uORF status, SCPN")+xlab("T.E.")+theme(legend.position="top")+geom_hline(yintercept=0.5,size=0.2)+geom_vline(xintercept=0, size=0.2)
TE_SCPN_uORF_plot


TE_distros_for_GO_ID <- function(CPN_df, SCPN_df, GO_IDs, allGO=ensembl_to_allGO, GO_name) {
  GO_ID_genes <- filter(allGO, go_id %in% GO_IDs)
  CPN_df <- CPN_df %>% mutate(gene_set=case_when(
    (Geneid %in% GO_ID_genes$ensembl_id)==T ~ GO_name,
    (Geneid %in% GO_ID_genes$ensembl_id)==F ~ paste("not",GO_name, sep="_")
  ))
  SCPN_df <- SCPN_df %>% mutate(gene_set=case_when(
    (Geneid %in% GO_ID_genes$ensembl_id)==T ~ GO_name,
    (Geneid %in% GO_ID_genes$ensembl_id)==F ~ paste("not",GO_name, sep="_")
  ))
  CPN_GO_plot <- ggplot(CPN_df, aes(x=log2FoldChange, color=interaction( uORF_simple,gene_set)))+JF_theme+stat_ecdf(size=0.1)+xlim(c(-2,2))+ylab("cumulative_fraction")+ggtitle("T.E. by uORF status, CPN")+xlab("T.E.")+geom_hline(yintercept=0.5, size=0.2)+geom_vline(xintercept=0, size=0.2)+theme(legend.position = "top")
  SCPN_GO_plot <- ggplot(SCPN_df, aes(x=log2FoldChange, color=interaction( uORF_simple,gene_set)))+JF_theme+stat_ecdf(size=0.1)+xlim(c(-2,2))+ylab("cumulative_fraction")+ggtitle("T.E. by uORF status, SCPN")+xlab("T.E.")+geom_hline(yintercept=0.5, size=0.2)+geom_vline(xintercept=0, size=0.2)+theme(legend.position = "top")
  return_list <- list("CPN_plot"=CPN_GO_plot, "SCPN_plot"=SCPN_GO_plot)
  return(return_list)
}

synapse_organization_TE_plots <- TE_distros_for_GO_ID(CPN_df = res_TE_CPN_df, SCPN_df = res_TE_SCPN_df, GO_IDs="GO:0050808", GO_name="Synapse_Org")
axonogenesis_TE_plots <- TE_distros_for_GO_ID(CPN_df = res_TE_CPN_df, SCPN_df = res_TE_SCPN_df, GO_IDs="GO:0007409", GO_name="Axonogenesis")
axon_development_TE_plots <-  TE_distros_for_GO_ID(CPN_df = res_TE_CPN_df, SCPN_df = res_TE_SCPN_df, GO_IDs="GO:0061564", GO_name="Axon dev.")
mRNA_processing_TE_plots <- TE_distros_for_GO_ID(CPN_df = res_TE_CPN_df, SCPN_df = res_TE_SCPN_df, GO_IDs="GO:0006397", GO_name="mRNA processing")
RNP_biogenesis_TE_plots <-  TE_distros_for_GO_ID(CPN_df = res_TE_CPN_df, SCPN_df = res_TE_SCPN_df, GO_IDs="GO:0022613", GO_name="ribonucleoprotein complex biogenesis")
synapse_assembly_TE_plots <- TE_distros_for_GO_ID(CPN_df = res_TE_CPN_df, SCPN_df = res_TE_SCPN_df, GO_IDs="GO:0007416", GO_name="synapse assembly")

res_TE_CPN_df <- res_TE_CPN_df %>% mutate(TE_quintile=ntile(log2FoldChange,n=5))
res_TE_SCPN_df <- res_TE_SCPN_df %>% mutate(TE_quintile=ntile(log2FoldChange,n=5))

TE_CPN_uORF_byQuntile_plot <- ggplot(res_TE_CPN_df, aes(x=log2FoldChange, color=uORF_status))+JF_theme+stat_ecdf(size=0.1)+facet_wrap(~TE_quintile,scales="free_x")+ylab("cumulative_fraction")+ggtitle("T.E. by uORF status, CPN")+xlab("T.E.")+theme(legend.position="top")+geom_hline(yintercept=0.5)+geom_vline(xintercept=0)+geom_hline(yintercept=0.5,size=0.2)+geom_vline(xintercept=0, size=0.2)


```

```{r ribotricer_uORFs}
RT_CPN_file <- "CPN_RiboCode/ribotracer_detected_ORFs_translating_ORFs.tsv"
RT_SCPN_file <- "SCPN_RiboCode/ribotracer_detected_ORFs_translating_ORFs.tsv"

RT_CPN <- read.table(RT_CPN_file,header=T, sep="\t")

collapse_ribotricer <- function(RT) {
  ###Take in a ribotricer file, split ORF_ID into its fields, append gene_name 
  ##and collapse ORFs with the same stop codon or stop condon, take the ORF with the longest length
  RT_collapsed <- RT %>% 
    separate(col=ORF_ID, sep="_", into=c("tx_id","start","stop","obs_length"), remove=F) %>%
    mutate(ORF_by_geneid=case_when(
      strand=="+" ~ paste(gene_id,start,stop,obs_length, sep="_"),
      strand=="-" ~ paste(gene_id,stop,start,obs_length, sep="_"))) %>%
    group_by(gene_id, ORF_type, start) %>% top_n(1,wt=length) %>%
    group_by(gene_id, ORF_type, stop) %>% top_n(1,wt=length)
}

test <- collapse_ribotricer(RT_CPN)

```



```{r uORFs_figure}

uORFs_figure <- ORF_types_plot / (TE_CPN_uORF_plot + TE_SCPN_uORF_plot) / (uORF_ego_barplot$CPN_uORFs + uORF_ego_barplot$SCPN_uORFs) / (synapse_assembly_TE_plots$CPN_plot + synapse_assembly_TE_plots$SCPN_plot) / plot_spacer()

pdfName=paste(outDir, "uORF_fig.pdf", sep="/")
pdf(file=pdfName, width=6.5, height=9)
plot(uORFs_figure)
dev.off()

```

###Run RiboWaltz to evaluate 3-nt periodicity in the data sets. 
###Requires a directory of transcriptome-aligned BamFiles and a gtf
```{r bamToList, echo=FALSE}
#Try loading in all the files and see how it goes. After a more careful
#reading of the RiboWaltz vignette, it looks like it will process samples separately.
#make a named vector mapping bamFile names (no path or extension needed) to descriptiveNames
RP_samps <- filter(experimentTypes, exp=="RP")
RP_samps <- data.frame(RP_samps)
RP_descriptiveNames <- as.character(RP_samps$Samples)
names(RP_descriptiveNames) <- as.character(RP_samps$samps)
write.csv(file="test337.txt", RP_descriptiveNames)
annotation_dt <- create_annotation(gtfpath = gtf, dataSource = "GRCm38.95",organism="Mus musculus")
reads_list <- bamtolist(bamfolder = bamFiles, annotation = annotation_dt, transcript_align = FALSE, name_samples = RP_descriptiveNames) 
```

```{r lengthFilter, echo=FALSE}
#From last run, need to set length_filter_mode="custom" for "bad" or
#RNA-seq libraries to get scored. 
filtered_list <- riboWaltz::length_filter(data = reads_list, length_filter_mode = "custom", length_range = c(27:33))
```

```{r psite, echo=FALSE}
#perform p-site analysis for all samples, save data to a sub-folder of the transcriptome alignment.
dir.create(outDir)
psite_dir <- paste(outDir,'/psite', sep="")
dir.create(psite_dir)
psite <- psite(data=filtered_list, plot=T, plot_dir = psite_dir, plot_format = "pdf")
```

###Make plots that stratify by frame:
####At first, I plotted using reads_list. This makes plots across all reads lengths
####Plotting using the length-filtered reads (filtered_list) makes much nicer graphs.
```{r stratify, echo=FALSE}
reads_psite_list <- psite_info(filtered_list,psite)
frames_stratified <- frame_psite_length(reads_psite_list, region="all",cl=90)
strat_dir <- paste(outDir,"/frame_counts",sep="")
dir.create(strat_dir)
#setwd(strat_dir)
pdfName <- paste(strat_dir, "Frame_counts_all_samples.pdf", sep="/" )
frames_stratified$plot
pdf(pdfName)
frames_stratified$plot
dev.off()
#The plot looks squished if I try plotting all of them on the same plot. Try looping over the sampleNames and making individual plots for each one. Update: If I only include filtered reads, the graphs aren't too squished and are readable
sampleNames=unique(psite$sample)
for(sampleName in sampleNames) {
  frame_stratified=frame_psite_length(reads_psite_list,  sample=sampleName, region="all",cl=90)
  pdfName=paste(sampleName,".pdf",sep="")
  pdf(file=pdfName)
  plot(frame_stratified$plot)
  dev.off()
}
```

###P-site heatmaps (both directly from RiboWaltz, and by re-plotting the RiboWaltz data for a nicer-looking plot) for all samples in current set. 
###Currently using scale factors as 1/CDS_reads:
```{r psite_heatmap, echo=FALSE}
heat_dir <- paste(outDir,"/heatmaps",sep="")
dir.create(heat_dir)
#setwd(heat_dir)
#calculate scale factors as the inverse of CDS_counts
scale_factors <- filter(featureCounts_tidy, exp=="RP") %>%
  group_by(samps) %>%
  summarize(CDS_reads=sum(counts)) %>%
  mutate(scale_factors=1/CDS_reads) %>%
  dplyr::select(scale_factors)
scale_factors <- c(t(scale_factors))
names(scale_factors) <- sampleNames
#Build up the list of sample names
sampNamesList <- list()
for (sampName in sampleNames) {
  sampNamesList[[sampName]] <- c(sampName)
}
psite_heatmap <- metaheatmap_psite(reads_psite_list, annotation=annotation_dt,log=F,utr5l=15, cdsl=30, utr3l=15, sample=sampNamesList, scale_factors=scale_factors)
psite_heatmap$plot
pdfName=paste(heat_dir, "psite_heatmap.pdf",sep="/")
pdf(file=pdfName, width=11, height=8.5)
plot(psite_heatmap$plot)
dev.off()
#Example plotting function that looks nicer than a heatmap:
psite_signal_plot <- ggplot(psite_heatmap$dt, aes(x=distance,y=reads,color=sample))+geom_line()+JF_theme+facet_wrap(~reg)
psite_signal_plot
psite_dt <- psite_heatmap$dt
psite_dt$reg <- plyr::revalue(psite_dt$reg, c("Distance from start (nt)"="from start (nt):","Distance from stop (nt)"="from stop (nt):"))
psite_signal_heat <- ggplot(psite_dt, aes(x=distance,fill=reads,y=sample))+geom_tile()+facet_wrap(~reg)+scale_fill_gradient(low="white", high="black")+JF_theme+xlab("distance from start or stop (nt)")+ylab("")+xlab("")
psite_signal_heat
#look at how scaling is done. Try row normalization of raw read counts to compute frequencies within the regions of interest
psite_heatmap_raw <- metaheatmap_psite(reads_psite_list, annotation=annotation_dt,log=F,utr5l=15, cdsl=30, utr3l=15, sample=sampNamesList)
psite_dt_raw <- psite_heatmap_raw$dt
psite_dt_raw$reg <- plyr::revalue(psite_dt_raw$reg, c("Distance from start (nt)"="from start (nt):","Distance from stop (nt)"="from stop (nt):"))
psite_region_sums <- psite_dt_raw %>%
  group_by(sample, reg) %>%
  summarize(region_sum=sum(reads))
psite_dt_raw <- merge(psite_dt_raw, psite_region_sums, by=c("sample","reg"))
psite_dt_raw <- psite_dt_raw %>%
  mutate(freq=reads/region_sum) %>% 
  mutate(Samples=sample)
titration_metadata <- dplyr::select(experimentTypes, Samples, exp, date, cells) %>% filter(exp=="RP")
psite_dt_raw <- merge(psite_dt_raw, titration_metadata, by="Samples")
psite_dt_raw <- psite_dt_raw %>%
  mutate(Samples= fct_reorder(Samples,as.numeric(cells), .desc=F))
#psite_dt_raw <- merge(psite_dt_raw, dplyr::filter(experimentTypes, Samples, exp_date), by="Samples")
psite_freqHeat <- ggplot(psite_dt_raw, aes(x=distance,fill=freq,y=sample))+geom_tile()+facet_wrap(~reg)+scale_fill_gradient(low="white", high="black")+JF_theme+xlab("distance from start or stop (nt)")+ylab("")+xlab("")+theme(legend.position="bottom")
psite_freqHeat
pdfName=paste(heat_dir, "psite_freq_heatmap.pdf",sep="/")
pdf(file=pdfName, width=11, height=8.5)
plot(psite_freqHeat)
dev.off()
```

###Next, calculate the percent of P-sites within each reading frame for each sample.
```{r framePsite, echo=FALSE}
frame_dir <- paste(outDir,"/frame_psite_percentage",sep="")
dir.create(frame_dir)
#setwd(frame_dir)
frame_psites <- frame_psite(reads_psite_list, region="all")
frame_psites$plot
pdfName <- paste(frame_dir, "Frame_percentages_all_samples.pdf", sep="/")
pdf(pdfName)
plot(frame_psites$plot)
dev.off()
sampleNames=unique(psite$sample)
for(sampleName in sampleNames) {
  aFramePsite=frame_psite(reads_psite_list, sample=sampleName, region="all")
  pdfName=paste(sampleName,".pdf",sep="")
  pdf(file=pdfName)
  plot(aFramePsite$plot)
  dev.off()
}
```

###And make P-site frequency plots for comparison across samples:
```{r psitePlotting, echo=FALSE}
plotFramePsites <- function(samples, frame_psites, sampNames=samples) {
  #function to take in a list of samples, and frame_psites Ribowaltz object    #Plots the percentage of reads over CDSs in each frame, grouped by samples.
  FPs <- frame_psites$dt
  FPsToPlot <- filter(FPs, sample %in% samples & region == "CDS" )
  FPplot <- ggplot(data=FPsToPlot, aes(x=sample,y=percentage,fill=as.factor(frame)))
  FPplot <- FPplot+geom_bar(stat="identity", position=position_dodge())+theme_minimal(base_size=18)+scale_fill_discrete(name="frame")+ylim(0,100)+scale_x_discrete(labels=sampNames)+JF_theme
  FPplot <- FPplot+theme(axis.text.x = element_text(angle = 45, hjust=1))+xlab("")+ylab("%P-sites in each frame")
                           
}
#setwd(frame_dir)
psiteFreq<- plotFramePsites(sampleNames,frame_psites)
pdfName=paste(frame_dir, "psite_freq.pdf",sep="/")
pdf(file=pdfName, width=11, height=8.5)
plot(psiteFreq)
dev.off()
psiteFreq
```

###Plots for reviewer #3's questions about P-site frequency broken down by TE status.
```{r Reviewer3_chunk}

CPN_TE_high_tx <- ensmusg2ensmust %>% filter(Geneid %in% CPN_TE_high)
SCPN_TE_high_tx <- ensmusg2ensmust %>% filter(Geneid %in% SCPN_TE_high)
CPN_TE_low_tx <- ensmusg2ensmust %>% filter(Geneid %in% CPN_TE_low)
SCPN_TE_low_tx <- ensmusg2ensmust %>% filter(Geneid %in% SCPN_TE_low)
CPN_TE_notSig_tx <- ensmusg2ensmust %>% filter( (Geneid %in% CPN_TE_high)==F & (Geneid %in% CPN_TE_low)==F)
SCPN_TE_notSig_tx <- ensmusg2ensmust %>% filter( (Geneid %in% SCPN_TE_high)==F & (Geneid %in% SCPN_TE_low)==F)

transcripts_by_TE_list <- list("CPN_TE_high"=CPN_TE_high_tx$ensembl_transcript_id, "CPN_TE_low"=CPN_TE_low_tx$ensembl_transcript_id, "SCPN_TE_high"=SCPN_TE_high_tx$ensembl_transcript_id, "SCPN_TE_low"=SCPN_TE_low_tx$ensembl_transcript_id, "CPN_TE_noSig"=CPN_TE_notSig_tx$ensembl_transcript_id, "SCPN_TE_noSig"=SCPN_TE_notSig_tx$ensembl_transcript_id )

split_psites_by_TE_list <- function(transcript_list, RPL=reads_psite_list) {
  split_RPL <- lapply(RPL, function(x) {
    filter(x, transcript %in% transcript_list)
  })
}

splitPs <- lapply(transcripts_by_TE_list, split_psites_by_TE_list)

splitP_frame_psites <- lapply(splitPs, frame_psite, region="all")
splitP_frame_psite_bySample_plots <- lapply(splitP_frame_psites, plotFramePsites, samples=sampleNames)
splitP_frame_psite_bySample_plots$CPN_TE_high <- splitP_frame_psite_bySample_plots$CPN_TE_high+ggtitle("CPN_TE_high")+theme(legend.position="top")
splitP_frame_psite_bySample_plots$SCPN_TE_high <- splitP_frame_psite_bySample_plots$SCPN_TE_high+ggtitle("SCPN_TE_high")+theme(legend.position="top")
splitP_frame_psite_bySample_plots$CPN_TE_low <- splitP_frame_psite_bySample_plots$CPN_TE_low+ggtitle("CPN_TE_low")+theme(legend.position="top")
splitP_frame_psite_bySample_plots$SCPN_TE_low <- splitP_frame_psite_bySample_plots$SCPN_TE_low+ggtitle("SCPN_TE_low")+theme(legend.position="top")
splitP_frame_psite_bySample_plots$CPN_TE_noSig <- splitP_frame_psite_bySample_plots$CPN_TE_noSig+ggtitle("SCPN_TE_noSig")+theme(legend.position="top")
splitP_frame_psite_bySample_plots$SCPN_TE_noSig <- splitP_frame_psite_bySample_plots$SCPN_TE_noSig+ggtitle("SCPN_TE_noSig")+theme(legend.position="top")

split_Reads <- lapply(transcripts_by_TE_list, split_psites_by_TE_list, RPL=reads_list)

cds_distro_from_reads_list <- function(RL=reads_list) {
  RL_length_distros <- lapply(RL, function(x) {
  RL_lengths <- data.frame(x) %>% group_by(length) %>% summarize(length_counts_bySample=n())
  })
  RL_lengths <- bind_rows(RL_length_distros, .id="sample") %>% mutate(subtype=case_when(regexpr('^CPN', sample)==T~"CPN",regexpr('^SCPN', sample)==T~"SCPN"))
  lengths_bySubtype <- RL_lengths %>% group_by(subtype, length) %>% summarize(length_counts_bySubtype=sum(length_counts_bySample))
  readSums_bySubtype <- lengths_bySubtype %>% group_by(subtype) %>% summarize(readSums=sum(length_counts_bySubtype))
  lengths_bySubtype <- lengths_bySubtype %>% left_join(readSums_bySubtype, by="subtype") %>% mutate(freq_bySubtype=length_counts_bySubtype/readSums)
}

split_Read_distros <- lapply(split_Reads, cds_distro_from_reads_list)  

split_Read_distros <- bind_rows(split_Read_distros, .id="TE_status")

CPN_length_distro_by_TE_plot <- ggplot(filter(split_Read_distros, subtype=="CPN" & regexpr("^CPN",TE_status)==T), aes(x=length,y=freq_bySubtype, color=TE_status))+JF_theme+geom_line()+theme(legend.position="top")+ggtitle("read length distro over CDS by CPN T.E. status")+xlim(c(20,40))
CPN_length_distro_by_TE_plot
SCPN_length_distro_by_TE_plot <- ggplot(filter(split_Read_distros, subtype=="SCPN" & regexpr("^SCPN",TE_status)==T), aes(x=length,y=freq_bySubtype, color=TE_status))+JF_theme+geom_line()+theme(legend.position="top")+ggtitle("read length distro over CDS by SCPN T.E. status")+xlim(c(20,40))
SCPN_length_distro_by_TE_plot 

#moveTo new Rmd
loadFC2tidyCounts <- function(experimentType, CDS, five_utr, three_utr) {
sampNames=experimentType$samps
#write.csv(file="test.txt", sampNames)
CDS <- read.table(file=CDS, header= T, sep="\t", stringsAsFactors = F)
Five_prime <- read.table(file=five_utr, header= T, sep="\t", stringsAsFactors = F)
Three_prime <- read.table(file=three_utr, header= T, sep="\t", stringsAsFactors = F)
filteredCDS <- CDS %>% 
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)
filtered5p <- Five_prime %>%
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)
filtered3p <- Three_prime %>%
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)

maxSampCol <- 6+length(sampNames)
mRNA_counts <- filteredCDS[,7:maxSampCol]+filtered5p[,7:maxSampCol]+filtered3p[,7:maxSampCol]
CDS_counts <- filteredCDS[,7:maxSampCol]
FiveP_counts <- filtered5p[,7:maxSampCol]
ThreeP_counts <- filtered3p[,7:maxSampCol]
colnames(CDS_counts) <- sampNames
colnames(FiveP_counts) <- sampNames
colnames(ThreeP_counts) <- sampNames
CDS_counts <- CDS_counts %>% mutate(Geneid=filteredCDS$Geneid, Length=filteredCDS$Length, region="CDS")
FiveP_counts <-FiveP_counts %>% mutate(Geneid=filteredCDS$Geneid, Length=filtered5p$Length, region="5' UTR")
ThreeP_counts <-ThreeP_counts %>% mutate(Geneid=filteredCDS$Geneid, Length=filtered3p$Length, region="3' UTR")

All_counts <- rbind(CDS_counts, FiveP_counts, ThreeP_counts)
All_counts_tidy <- gather(All_counts, "samps","counts", -Geneid, -Length, -region)
All_counts_tidy <- merge(All_counts_tidy, experimentType, by="samps")
}

All_counts_tidy <- loadFC2tidyCounts(experimentTypes, CDS, five_utr, three_utr)

makeRegionCov <- function(ACT, sampNames=descriptiveNames) {
  rpk <- ACT %>%
  group_by(samps, region, exp) %>%
  summarize(totCounts=sum(counts),totLength=sum(Length)) %>%
  mutate(rpk=totCounts/(totLength/1000))

million_mapped <- ACT %>%
  group_by(samps) %>%
  summarize(mRNA_mapped_mil=sum(counts)/1e6)

rpkm <- merge(rpk, million_mapped, by="samps") %>%
  mutate(rpkm=rpk/mRNA_mapped_mil)

rpkm$region <- factor(rpkm$region, levels=c("5' UTR","CDS","3' UTR"))
rpkm <- merge(rpkm, sampNames, by="samps")
# rpkm_byRegion <- rpkm_byRegion %>%
#   mutate(Samples=sampNames)
}

test <- makeRegionCov(ACT=All_counts_tidy)

geneids_by_TE_list <- list("CPN_TE_high"=CPN_TE_high, "CPN_TE_low"=CPN_TE_low, "CPN_TE_noSig"=CPN_TE_notSig,
                           "SCPN_TE_high"=SCPN_TE_high, "SCPN_TE_low"=SCPN_TE_low, "SCPN_TE_noSig"=SCPN_TE_notSig)
split_ACT <- lapply(geneids_by_TE_list, function(x, ACT=All_counts_tidy) {
  filter(ACT, Geneid %in% x)
})
split_ACT_regionCovs <- lapply(split_ACT, makeRegionCov)
split_ACT_regionCovs <- split_ACT_regionCovs %>% bind_rows(.id="TE_status") %>% mutate(subtype=case_when(regexpr('^CPN', descriptiveNames)==T~"CPN",regexpr('^SCPN', descriptiveNames)==T~"SCPN"))

CPN_regionCov_by_TE_status <- ggplot(filter(split_ACT_regionCovs, subtype=="CPN" & exp=="RP" & regexpr("^CPN",TE_status)==T), aes(x=TE_status, y=rpkm, fill=region))+JF_theme+geom_col(position="dodge")+theme(legend.position = "top")+facet_wrap(~descriptiveNames, nrow=1)+ggtitle("Coverage by region by T.E. status, CPN")+theme(axis.text.x = element_text(angle = 45, hjust=1))
CPN_regionCov_by_TE_status 
SCPN_regionCov_by_TE_status <- ggplot(filter(split_ACT_regionCovs, subtype=="SCPN" & exp=="RP" & regexpr("^SCPN",TE_status)==T), aes(x=TE_status, y=rpkm, fill=region))+JF_theme+geom_col(position="dodge")+theme(legend.position = "top")+facet_wrap(~descriptiveNames, nrow=1)+ggtitle("Coverage by region by T.E. status, SCPN")+theme(axis.text.x = element_text(angle = 45, hjust=1))
SCPN_regionCov_by_TE_status

TE_riboWaltz_QC_fig <- (CPN_length_distro_by_TE_plot | SCPN_length_distro_by_TE_plot) / (splitP_frame_psite_bySample_plots$CPN_TE_high + splitP_frame_psite_bySample_plots$CPN_TE_noSig + splitP_frame_psite_bySample_plots$CPN_TE_low) / (splitP_frame_psite_bySample_plots$SCPN_TE_high + splitP_frame_psite_bySample_plots$SCPN_TE_noSig + splitP_frame_psite_bySample_plots$SCPN_TE_low) / (CPN_regionCov_by_TE_status | SCPN_regionCov_by_TE_status ) + plot_annotation(tag_levels = list(c("A","B","C","D","E","F","G","H","I","J"))) & theme(plot.tag=element_text(size=12, face="bold"))+theme(legend.key.size=unit(0.5,"line"))

pdfName=paste(outDir, "TE_riboWaltz_QC_fig.pdf", sep="/")
pdf(file=pdfName, width=6.5, height=9)
plot(TE_riboWaltz_QC_fig)
dev.off()

```
```{r sesh}
sessionInfo()

```

