---
title: "CountsPlotting"
author: "John E Froberg"
date: "Sys.Date()"
output:
  pdf_document: default
  html_document: default
---

```{r setup}
knitr::opts_chunk$set(echo=FALSE, message=FALSE)

library('devtools')
library('tidyverse')
library('riboWaltz')
library("rmarkdown")
library("patchwork")
library("pheatmap")
library("RColorBrewer")
library("ggplotify")
library("topGO")
library("ggstance")
library("DESeq2")
library("biomaRt")
library("UpSetR")
library("GGally")
library("clusterProfiler")
library("org.Mm.eg.db")
library("enrichplot")
library("pROC")
#install_github("xryanglab/xtail")
library("xtail")
#BiocManager::install("anota2seq")
library("anota2seq")

# #variables that are normally passed in from the Snakemake run:
experimentTypeFile="experimentTypesFull_SubFall2021.csv"
lengthCountsFile="dedup_lengthDistroTidy.txt"
CDS="dedupBams/featureCounts_CDS_summary.txt"
three_utr="dedupBams/featureCounts_three_prime_utr_summary.txt"
five_utr="dedupBams/featureCounts_five_prime_utr_summary.txt"
gtf="/n/macklis_lab/users/jfroberg/gtfs/Mus_musculus.GRCm38.95_chrNamed_headFix.gtf"
bamFiles="dedup_RPbams"
outDir="interactivePlots"
dir.create(outDir)
RiboCodeFile="RiboCode_ORFs_out.txt"


JF_theme <- theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "black"), plot.title = element_text(size = 7, face = "bold"),legend.title=element_text(size=7), legend.text=element_text(size=7), axis.text = element_text(size=7), axis.title.x = element_text(size=7),axis.title.y=element_text(size=7))
```


```{r getGeneNames_from_biomaRt}
mouse_ensembl <- useMart("ensembl", dataset = "mmusculus_gene_ensembl")
mouse_gene_symbols <- getBM(attributes=c("ensembl_gene_id","external_gene_name"), mart=mouse_ensembl)
mouse_gene_symbols_wGeneid <- mouse_gene_symbols %>% mutate(Geneid=ensembl_gene_id)
ensmusg2ensmust <- getBM(attributes=c("ensembl_gene_id","external_gene_name", "ensembl_transcript_id"), mart=mouse_ensembl) %>% mutate(Geneid=ensembl_gene_id)

ensembl_to_entrez <- org.Mm.egENSEMBL2EG %>%
  as.data.frame()
ensembl_to_allGO <- as.data.frame(org.Mm.egGO2ALLEGS) %>%
  left_join(ensembl_to_entrez, by = "gene_id")


# mouse_gene_symbols_resExp <- getBM(attributes=c("ensembl_gene_id","external_gene_name"), filters="ensembl_gene_id", values=resExp_df$Geneid, mart=mouse_ensembl)
# mouse_gene_symbols_resExp$Geneid <- mouse_gene_symbols_resExp$ensembl_gene_id
```


###Load in the metadata from the experiment types file. 
###Compute "descriptive names" by dropping the "samps" and "date" fields and pasting together.
###Save descriptiveNames as Upper Case "Samples" in experimentTypes.
```{r loadMetadata, echo=FALSE}
experimentTypes=read.csv(experimentTypeFile)
 descriptiveNames <- experimentTypes %>%
  unite(col="descriptiveNames", -samps, -date,-exp, sep=".")
 descriptiveNames <- dplyr::select(descriptiveNames, samps, descriptiveNames)
 experimentTypes$Samples <- descriptiveNames$descriptiveNames
 experimentTypes
```

###Convert length counts to relative frequencies, add in the metadata
```{r countsToFreqwMeta}
lengthCounts <- read.csv(lengthCountsFile)
totalCounts <- lengthCounts %>%
  group_by(samps,anno) %>%
  summarize(sum=sum(count))

lengthDistro <- merge(lengthCounts, totalCounts, by=c("samps", "anno"))
lengthDistro <- lengthDistro %>%
  mutate(freq=count/sum)

#metadata <- read.csv("metadata.csv")
lengthDistro <- merge(lengthDistro, experimentTypes, by="samps")

```
###Find the length with max freq for all samples
```{r maxFreq}
maxFreq <- lengthDistro %>%
  group_by(samps, anno) %>%
  summarize(max=max(freq))

max <- merge(maxFreq, lengthDistro, by=c("samps","anno"))
maxLengthDistro <- filter(max, freq==max)
maxLengthCDS_RPs <- filter(maxLengthDistro, exp=="RP" & anno=="CDS")
```


###Plot Freq vs length for the CDS with various groupings of the samples
```{r freqPlotting}
CDS_riboProf <- ggplot(filter(lengthDistro, exp=="RP",anno=="CDS"), aes(x=length, y=freq, color=Samples))+geom_line(size=1)+JF_theme+xlim(20,40)+xlab("length (nt)")+theme(axis.title.x = element_text(vjust=0))

CDS_riboProf_byDate <- ggplot(filter(lengthDistro, exp=="RP",anno=="CDS"), aes(x=length, y=freq, color=Samples))+geom_line(size=1)+facet_wrap(~date)+xlim(20,40)+JF_theme

CDS_riboProf_byDate_noxlim <- ggplot(filter(lengthDistro, exp=="RP",anno=="CDS"), aes(x=length, y=freq, color=descriptiveNames))+geom_line(size=1)+facet_wrap(~date)+ylim(0,0.5)+JF_theme

if ( "AF" %in% levels(lengthDistro$exp) ) {
CDS_AF_byDate <-ggplot(filter(lengthDistro, exp=="AF",anno=="CDS"), aes(x=length, y=freq, color=Samples))+geom_line(size=1)+facet_wrap(~date)+ylim(0,0.5)+JF_theme
CDS_AF_byDate
}


allAnno_RiboProf_bySample <- ggplot(filter(lengthDistro, exp=="RP", anno %in% c("CDS","three_prime_utr","snoRNA")), aes(x=length, y=freq, color=anno))+geom_line(size=1)+facet_wrap(~Samples, nrow=ceiling(nlevels(factor(lengthDistro$Samples))/6))+xlim(20,40)+JF_theme+xlab("length (nt)")

CDS_riboProf
CDS_riboProf_byDate
allAnno_RiboProf_bySample
```



###Load in the featureCounts. Use only CDS reads for RiboProf, 3'+5'UTR+CDS reads for AlkFrag libraries
```{r loadFeatureCounts, echo=FALSE}
loadFC <- function(experimentType, CDS, five_utr, three_utr) {
sampNames=experimentType$samps
CDS <- read.table(file=CDS, header= T, sep="\t", stringsAsFactors = F)
Five_prime <- read.table(file=five_utr, header= T, sep="\t", stringsAsFactors = F)
Three_prime <- read.table(file=three_utr, header= T, sep="\t", stringsAsFactors = F)
filteredCDS <- CDS %>% 
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)
filtered5p <- Five_prime %>%
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)
filtered3p <- Three_prime %>%
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)

#move to new Rmd
maxSampCol <- 6+length(sampNames)
mRNA_counts <- filteredCDS[,7:maxSampCol]+filtered5p[,7:maxSampCol]+filtered3p[,7:maxSampCol]
CDS_counts <- filteredCDS[,7:maxSampCol]
#

colnames(mRNA_counts) <- sampNames
colnames(CDS_counts) <- sampNames
mRNA_counts <- mRNA_counts %>%
  mutate(Geneid=filteredCDS$Geneid, Length=filteredCDS$Length)
CDS_counts <- CDS_counts %>%
  mutate(Geneid=filteredCDS$Geneid, Length=filteredCDS$Length)
mRNA_counts_tidy <- gather(mRNA_counts,"samps","counts", -Geneid, -Length)
CDS_counts_tidy <- gather(CDS_counts,"samps","counts", -Geneid, -Length)
mRNA_counts_tidy <- merge(mRNA_counts_tidy, experimentType, by="samps")
CDS_counts_tidy <- merge(CDS_counts_tidy, experimentType, by="samps")

raw_counts_tidy <- rbind(filter(mRNA_counts_tidy, exp=="AF"),filter(CDS_counts_tidy, exp=="RP"))
}


featureCounts_tidy <- loadFC(experimentTypes, CDS, five_utr, three_utr)
total_mRNA_counts <- featureCounts_tidy %>%
  group_by(Samples) %>%
  summarize(mRNA_counts=sum(counts))
total_mRNA_counts
total_mRNA_counts_plot <- ggplot(total_mRNA_counts, aes(x=Samples, y=mRNA_counts, fill=Samples))+geom_bar(stat="identity", position="dodge")+JF_theme+theme(axis.text.x = element_text(angle = 45, hjust=1))+xlab("")+geom_hline(yintercept = 150000)
total_mRNA_counts_plot

experimentTypesPlus_mRNACounts <- merge(total_mRNA_counts, experimentTypes, by="Samples")


total_mRNA_counts_facet <- ggplot(experimentTypesPlus_mRNACounts, aes(x=Samples, y=mRNA_counts, fill=Samples))+geom_bar(stat="identity", position="dodge")+JF_theme+theme(axis.text.x = element_text(angle = 45, hjust=1))+xlab("")+geom_hline(yintercept = 150000)+facet_wrap(~exp, nrow=2, scales="free_x")
total_mRNA_counts_facet
```

###Calculate the and plot reads per kilobase per million mapped (to mRNA) for the ribosome profiling experiments
```{r regionCovCalc}
#moveTo new Rmd
loadFCw_regions <- function(experimentType, CDS, five_utr, three_utr) {
sampNames=experimentType$samps
#write.csv(file="test.txt", sampNames)
CDS <- read.table(file=CDS, header= T, sep="\t", stringsAsFactors = F)
Five_prime <- read.table(file=five_utr, header= T, sep="\t", stringsAsFactors = F)
Three_prime <- read.table(file=three_utr, header= T, sep="\t", stringsAsFactors = F)
filteredCDS <- CDS %>% 
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)
filtered5p <- Five_prime %>%
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)
filtered3p <- Three_prime %>%
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)

maxSampCol <- 6+length(sampNames)
mRNA_counts <- filteredCDS[,7:maxSampCol]+filtered5p[,7:maxSampCol]+filtered3p[,7:maxSampCol]
CDS_counts <- filteredCDS[,7:maxSampCol]
FiveP_counts <- filtered5p[,7:maxSampCol]
ThreeP_counts <- filtered3p[,7:maxSampCol]
colnames(CDS_counts) <- sampNames
colnames(FiveP_counts) <- sampNames
colnames(ThreeP_counts) <- sampNames
CDS_counts <- CDS_counts %>% mutate(Geneid=filteredCDS$Geneid, Length=filteredCDS$Length, region="CDS")
FiveP_counts <-FiveP_counts %>% mutate(Geneid=filteredCDS$Geneid, Length=filtered5p$Length, region="5' UTR")
ThreeP_counts <-ThreeP_counts %>% mutate(Geneid=filteredCDS$Geneid, Length=filtered3p$Length, region="3' UTR")

All_counts <- rbind(CDS_counts, FiveP_counts, ThreeP_counts)
All_counts_tidy <- gather(All_counts, "samps","counts", -Geneid, -Length, -region)
All_counts_tidy <- merge(All_counts_tidy, experimentType, by="samps")

rpk <- All_counts_tidy %>%
  group_by(samps, region, exp) %>%
  summarize(totCounts=sum(counts),totLength=sum(Length)) %>%
  mutate(rpk=totCounts/(totLength/1000))

million_mapped <- All_counts_tidy %>%
  group_by(samps) %>%
  summarize(mRNA_mapped_mil=sum(counts)/1e6)

rpkm <- merge(rpk, million_mapped, by="samps") %>%
  mutate(rpkm=rpk/mRNA_mapped_mil)
#
}

rpkm_byRegion <- loadFCw_regions(experimentTypes, CDS, five_utr, three_utr)
rpkm_byRegion$region <- factor(rpkm_byRegion$region, levels=c("5' UTR","CDS","3' UTR"))
rpkm_byRegion <- merge(rpkm_byRegion, descriptiveNames, by="samps")
write.csv(file="test134-185.txt", head(rpkm_byRegion) , quote=F, row.names=F)
rpkm_byRegion <- rpkm_byRegion %>%
  mutate(Samples=descriptiveNames)
rpkm_byRegion_plot <- ggplot(filter(rpkm_byRegion, exp=="RP"), aes(y=rpkm,x=Samples, fill=region))+facet_wrap(~exp)+geom_bar(stat="identity", position="dodge")+JF_theme+theme(axis.text.x = element_text(angle = 45, hjust=1))+xlab("")
rpkm_byRegion_plot
#
```


###Calculate TPMs from the raw counts
```{r calculateTPM, echo=FALSE}

calcTPM <- function(RC_tidy) {
  RC <- RC_tidy %>% 
    dplyr::select(samps, Geneid, Length, counts) %>% #not robust to different types of metadata
    spread("samps","counts")
  Geneid=RC$Geneid
  lengths=RC$Length
  rawCounts = RC[,3:ncol(RC)] #rawCounts start with 7th column
  
  x <- rawCounts/lengths
  tpm.mat <- t( t(x) * 1e6 / colSums(x) )
  TPM_out <- data.frame(Geneid, tpm.mat)
  colnames(TPM_out) <- c("Geneid",colnames(rawCounts))
  TPM_out
}
TPMs <- calcTPM(featureCounts_tidy)
```

###Now plot the log10TPM+1 distributions
```{r plotTPM_distros, echo=FALSE}

tpmDensity <- function(TPMs,TPM_title="TPM_density") {
  #takes in a tidy table of tpms and returns a density plot
 JF_theme <- theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "black"), plot.title = element_text(size = 7, face = "bold"),legend.title=element_text(size=7), legend.text=element_text(size=7), axis.text = element_text(size=7), axis.title.x = element_text(size=7),axis.title.y=element_text(size=7))
  tpmPlot <- ggplot(TPMs,aes(log10TPM,fill=Samples ))+geom_histogram(alpha=0.3,position="identity",bins=100)+JF_theme+theme(legend.position = c(0.75,0.75))+ggtitle(TPM_title)
}


TPMs_tidy <- gather(TPMs, "samps","TPM",-Geneid)
TPMs_tidy <- TPMs_tidy %>%
  mutate(log10TPM=log10(TPM+1))
TPMs_tidy <- merge(TPMs_tidy, descriptiveNames, by="samps")
TPMs_tidy <- TPMs_tidy %>%
  mutate(Samples=descriptiveNames)

TPMdensity <- tpmDensity(TPMs_tidy)
TPMdensity <- TPMdensity + ylab("# of genes")
TPMdensity
```

###Plot the number of genes above TPM threshold=10 and threshold=20
```{r aboveTPMthreshold, echo=FALSE}
aboveTPM <- function(TPM_tidy, threshold=10) {
  #take in a tpm table and a tpm threshold, 
  tpm_frame <- data.frame(TPM_tidy, tpmFilter=TPM_tidy$TPM >= threshold)
  tpm_frame <- group_by(tpm_frame, Samples)
  tpm_aboveFilt <- summarize(tpm_frame, tpmFilter=sum(tpmFilter))
}

aboveTPM10 <- aboveTPM(TPMs_tidy, threshold=10)
aboveTPM20 <- aboveTPM(TPMs_tidy, threshold=20)


aboveTPM10_plot <- ggplot(aboveTPM10, aes(x=Samples, y=tpmFilter, fill=Samples))+geom_col()+ggtitle("TPM above 10")+JF_theme+xlab("")+theme(legend.position = "none")+theme(axis.text.x = element_text(angle = 45, hjust=1))+ylab("# genes above threshold")
aboveTPM20_plot <- ggplot(aboveTPM20, aes(x=Samples, y=tpmFilter, fill=Samples))+geom_col()+ggtitle("TPM above 20")+JF_theme+theme(legend.position = "none")+theme(axis.text.x = element_text(angle = 45, hjust=1))+ylab("# genes above threshold")+xlab("")

aboveTPM10_plot
aboveTPM20_plot
```

###Plot the distribtion of rawCounts in each sample:
```{r plotRawDistros, echo=FALSE}
rawDensity <- ggplot(featureCounts_tidy,aes(log10(counts+1),fill=Samples ))+geom_histogram(alpha=0.3,position="identity",bins=100)+JF_theme+theme(legend.position = c(0.75,0.75))+ggtitle("rawCounts_density")
rawDensity
```


###Plot the number of genes in each sample above rawCounts=10, 20, 40
```{r aboveRawThreshold, echo=FALSE}
aboveRaw <- function(FC_tidy, threshold=10) {
  #take in a featureCounts table and a raw threshold, 
  raw_frame <- data.frame(FC_tidy, rawFilter=FC_tidy$counts >= threshold)
  raw_frame <- group_by(raw_frame, Samples)
  raw_aboveFilt <- summarize(raw_frame, rawFilter=sum(rawFilter))
}

aboveRaw10 <- aboveRaw(featureCounts_tidy, threshold=10)
aboveRaw20 <- aboveRaw(featureCounts_tidy, threshold=20)
aboveRaw40 <- aboveRaw(featureCounts_tidy, threshold=40)

aboveRaw10_plot <- ggplot(aboveRaw10, aes(x=Samples, y=rawFilter, fill=Samples))+geom_col()+ggtitle("Raw above 10")+JF_theme+theme(legend.position = "none")+theme(axis.text.x = element_text(angle = 45, hjust=1))+ylab("# genes above threshold")+xlab("")
aboveRaw20_plot <- ggplot(aboveRaw20, aes(x=Samples, y=rawFilter, fill=Samples))+geom_col()+ggtitle("Raw above 20")+JF_theme+theme(legend.position = "none")+theme(axis.text.x = element_text(angle = 45, hjust=1))+ylab("# genes above threshold")+xlab("")
aboveRaw40_plot <- ggplot(aboveRaw40, aes(x=Samples, y=rawFilter, fill=Samples))+geom_col()+ggtitle("Raw above 40")+JF_theme+theme(legend.position = "none")+theme(axis.text.x = element_text(angle = 45, hjust=1))+ylab("# genes above threshold")+xlab("")

aboveRaw10_plot
aboveRaw20_plot
aboveRaw40_plot
```

###Code to filter a TPM-table either for genes where 1) all samples above a TPM-based filter (filterTPMs)
###Or 2) genes above a rawCount threshold (defult=10) in a given fraction of samples (default=1/4).
```{r filter_functions}
filterTPMs <- function(TPMs, threshold=10) {
  TPMs_filt <- filter_if(TPMs, is.double, all_vars(. > threshold))
}

filterRaw <- function(FC_tidy, threshold=10, frac_samples=1/4, experimentTypes, TPMs) {
  raw_frame <- data.frame(FC_tidy, rawFilter=FC_tidy$counts >= threshold)
  #First, filter for all samples where a gene passed the raw filter. Then, count the number of instances
  #Finally, filter for Genes where the number of samples exceeds the cutoff (where gene is above threshold in n/2-1 samples). Then filter the rows of the TPM matrix for gene IDs matching the genes passing the filter.
  
  filt_frame <- filter(raw_frame, rawFilter==TRUE)
  filt_frame <- filt_frame %>%
    group_by(Geneid) %>%
    add_count(Geneid)
  
  # raw_frame <- group_by(raw_frame, Geneid) %>%
  #   add_count(rawFilter)
  sampNames=experimentTypes$samps
  numSamps <- length(sampNames)
  sampCutoff <- frac_samples * numSamps
  GenesOverFilt <- filter(filt_frame, n >= sampCutoff)
  TPMs_filt <- filter(TPMs, Geneid %in% GenesOverFilt$Geneid)
}

```

###Calculate and plot Pearson's R correlation coefficient for all genes above a certain raw threshold (here, 10).
###In a certain fraction of samples (here, 1/4 of samples)
###Leave mRNA counts and date in the report, remove for the figure version for clarity.
```{r TPMCorrelationHeatmap}
TPMs_filtered <- filterRaw(FC_tidy=featureCounts_tidy,threshold=10, frac_samples=1/4, experimentTypes = experimentTypes, TPMs=TPMs)
numGenesAboveThreshold <- dim(TPMs_filtered)[1]
numGenesAboveThreshold

cormat <- cor(dplyr::select((TPMs_filtered), -Geneid))
experimentTypesPlus_mRNACounts <- merge(experimentTypes, total_mRNA_counts, by="Samples")
anno <- dplyr::select(experimentTypesPlus_mRNACounts, -samps)
rownames(anno) <- experimentTypesPlus_mRNACounts$samps
TPM_cor_pheatmap <- pheatmap(cormat, display_numbers = T, color=colorRampPalette(brewer.pal(n=9, name="YlOrRd"))(100), annotation_col=dplyr::select(anno, -Samples), treeheight_row = 10, treeheight_col = 10, show_colnames = F, labels_col = anno$Samples, show_rownames = F, fontsize=7)
TPM_cor_pheatmap

anno <- dplyr::select(experimentTypesPlus_mRNACounts, -samps, -date, -mRNA_counts)
rownames(anno) <- experimentTypesPlus_mRNACounts$samps
TPM_cor_pheatmap_simple <- pheatmap(cormat, display_numbers = T, color=colorRampPalette(brewer.pal(n=9, name="YlOrRd"))(100), annotation_col=dplyr::select(anno, -Samples), treeheight_row = 10, treeheight_col = 10, show_colnames = F, labels_col = anno$Samples, show_rownames = F, fontsize=7)
TPM_cor_pheatmap_simple


anno <- dplyr::select(experimentTypesPlus_mRNACounts, -samps, -date, -mRNA_counts, -genotype)
rownames(anno) <- experimentTypesPlus_mRNACounts$samps
TPM_cor_pheatmap_simpler <- pheatmap(cormat, display_numbers = F, color=colorRampPalette(brewer.pal(n=9, name="YlOrRd"))(100), annotation_col=dplyr::select(anno, -Samples,-cells,-rep), treeheight_row = 10, treeheight_col = 10, show_colnames = F, labels_col = anno$Samples, show_rownames = F, fontsize=7)
TPM_cor_pheatmap_simpler


```

#Filter out samples with fewer than 150000 mRNA-aligned reads (rounded to nearest 1000)
```{r TPM_heatmap_and_PCA_sample_filtering}
read_threshold=149500
featureCounts_tidy_sampFilt <- merge(featureCounts_tidy, total_mRNA_counts, by="Samples") %>%
  filter(mRNA_counts > read_threshold )

experimentTypes_sampFilt <- filter(experimentTypesPlus_mRNACounts, mRNA_counts > read_threshold)

featureCounts_tidy_sampFilt_forTPMcalc <- dplyr::select(featureCounts_tidy_sampFilt, -mRNA_counts)
TPMs_sampFilt <- calcTPM(featureCounts_tidy_sampFilt_forTPMcalc)

TPMs_sampFilt <- filterRaw(FC_tidy=featureCounts_tidy_sampFilt,threshold=20, frac_samples=1/5, experimentTypes = experimentTypes_sampFilt, TPMs=TPMs_sampFilt)
numGenesAboveThreshold <- dim(TPMs_sampFilt)[1]
numGenesAboveThreshold

cormat_sampFilt <- cor(dplyr::select((TPMs_sampFilt), -Geneid))

anno <- dplyr::select(experimentTypes_sampFilt, -samps, -date, -mRNA_counts, -genotype,-rep)
rownames(anno) <- experimentTypes_sampFilt$samps
TPM_cor_pheatmap_sampFilt <- pheatmap(cormat_sampFilt, display_numbers = T, color=colorRampPalette(brewer.pal(n=9, name="YlOrRd"))(100), annotation_col=dplyr::select(anno, -Samples), treeheight_row = 10, treeheight_col = 10, show_colnames = F, labels_row = anno$Samples, show_rownames = F, fontsize=7)
TPM_cor_pheatmap_sampFilt


TPMs_sampFilt.pca <- prcomp(t(dplyr::select((TPMs_sampFilt), -Geneid)))
TPMs_sampFilt.pca_X_tidy <- data.frame(samps=rownames(TPMs_sampFilt.pca$x), TPMs_sampFilt.pca$x)
TPMs_sampFilt.pca_X_tidy <- merge(TPMs_sampFilt.pca_X_tidy, experimentTypes_sampFilt, by="samps" )

TPMs_sampFilt.pca.var <- TPMs_sampFilt.pca$sdev^2
TPMs_sampFilt.pca.var <- round(TPMs_sampFilt.pca.var/sum(TPMs_sampFilt.pca.var)*100,1)

TPMs_sampFilt.PC1v2_plot <- ggplot(TPMs_sampFilt.pca_X_tidy, aes(x=PC1, y=PC2, color=subtype, shape=exp))+geom_point()+xlab(paste("PC1- ",TPMs_sampFilt.pca.var[1],"%", sep=""))+ylab(paste("PC2- ",TPMs_sampFilt.pca.var[2],"%", sep=""))
TPMs_sampFilt.PC1v2_plot
TPMs_sampFilt.PC2v3_plot <- ggplot(TPMs_sampFilt.pca_X_tidy, aes(x=PC2, y=PC3, color=subtype, shape=exp))+geom_point()+xlab(paste("PC2- ",TPMs_sampFilt.pca.var[2],"%", sep=""))+ylab(paste("PC3- ",TPMs_sampFilt.pca.var[3],"%", sep=""))
TPMs_sampFilt.PC2v3_plot

TPMs_sampFilt_tidy <- gather(TPMs_sampFilt, "samps","TPM",-Geneid)
TPMs_sampFilt_tidy <- merge(TPMs_sampFilt_tidy, experimentTypes_sampFilt, by="samps")

TPMs_sampFilt_vioPlot <- ggplot(TPMs_sampFilt_tidy, aes(x=Samples, y=log10(TPM+1), fill=subtype))+geom_violin()+facet_wrap(~exp, scales="free_x") 
TPMs_sampFilt_vioPlot
```

###Make the counts matrix for DESeq2.
```{r makeCountsMatrixForDESeq2}

DESeq2_counts_matrix <- featureCounts_tidy_sampFilt %>%
    dplyr::select(samps, Geneid, counts) %>%
    spread("samps","counts") %>%
    filter(Geneid %in% TPMs_sampFilt$Geneid)

#write DESeq2 counts matrix as output:
write.csv(DESeq2_counts_matrix, file="CPN_SCPN_DESeq2_counts_matrix.csv", quote=F, row.names=F)

rownames(DESeq2_counts_matrix) <- DESeq2_counts_matrix$Geneid
DESeq2_counts_matrix <- dplyr::select(DESeq2_counts_matrix, -Geneid)
colData <- dplyr::select(experimentTypes_sampFilt, samps, exp,subtype)
colData <- dplyr::arrange(colData, samps)

rownames(colData) <- colData$samps
colData <- dplyr::select(colData, -samps)

runDESeq_bySubtype <- function(FC) {
  DESeq2_counts_matrix <- FC %>%
    dplyr::select(samps, Geneid, counts) %>%
    spread("samps","counts")
  rownames(DESeq2_counts_matrix) <- DESeq2_counts_matrix$Geneid
  DESeq2_counts_matrix <- dplyr::select(DESeq2_counts_matrix, -Geneid)
  
  # colData <- filter(metaData, samps %in% FC$samps & region %in% regions & stage %in% stages & strain %in% strains) %>% dplyr::select(samps, stage, strain, region)
  colData <- FC %>%
    group_by(samps, exp, subtype) %>%
    summarize(n=n()) %>%
    dplyr::select(-n) %>%
    dplyr::arrange(samps)
  colData <- data.frame(colData)
rownames(colData) <- colData$samps
colData <- dplyr::select(colData, -samps)
colData <- droplevels(colData)

dds <- DESeqDataSetFromMatrix(DESeq2_counts_matrix, colData = colData, design=~subtype)
dds <- DESeq(dds)
}

test <- runDESeq_bySubtype(featureCounts_tidy_sampFilt)

```

###Run DESeq2, including an exp:subtype interaction term. The genes with this 
###interaction term significant are the ones differentially translated. 
###However, this includes exclusively translated genes, and buffered genes.
###Need to run DESeq2 separately on the AF (RNA) and RP (Ribo-Seq) data sets
###To be able to parse out translational regulation from buffering.
###Update 4/21/21: running one full model: ~exp+subtype+exp:subtype
###The main effect of exp is translation efficiency (TE) in CPN
###The main effect of exp+exp:subtype interaction is SCPN TE
###The main effect of subtype is the RP/AF ratio in CPN
###The main effect of subtype+exp:subtype interaction is RP/AF in SCPN
###Awkwardly renaming the results objects to match previous names
###Using genes that are significant in the subtype main effect as "RNA"
###and genes significant in the exp comparison as "Ribo". Significant in both
###is "RNA+Ribo"
###Write out the Geneids and log2FC for pre-ranked GSEA analysis
```{r DESeq2}
sig=0.05
TE_sig=0.1
dds <- DESeqDataSetFromMatrix(countData = DESeq2_counts_matrix, colData = colData, design=~exp+subtype+subtype:exp)
dds$exp <- factor(dds$exp, levels=c("AF","RP"))
dds$subtype <- factor(dds$subtype, levels=c("CPN","SCPN"))
dds <- DESeq(dds)
res_diffTE <- results(dds, alpha=TE_sig, name="expRP.subtypeSCPN", independentFiltering =T)


#main effect of exp, RP vs AF in CPN
res_TE_CPN <- results(dds, contrast=c("exp", "RP","AF" ), alpha=sig)
#exp effect for SCPN
res_TE_SCPN <- results(dds, contrast=list(c("exp_RP_vs_AF", "expRP.subtypeSCPN" )), alpha=sig)
#effect of subtype in AF, RP
res_subtype_AF <- results(dds, contrast=c("subtype","SCPN","CPN"), alpha=sig)
res_subtype_RP <- results(dds, contrast=list(c("subtype_SCPN_vs_CPN","expRP.subtypeSCPN")), alpha=sig)

res_AF_df <- data.frame(res_subtype_AF, contrast="mRNA", Geneid=rownames(res_subtype_AF)) %>% mutate(mRNA_sig=case_when(padj < sig & log2FoldChange < 0 ~ "lower", padj < sig & log2FoldChange > 0 ~ "higher", padj > sig ~ "not significant"))
res_RP_df <- data.frame(res_subtype_RP, contrast="RPF", Geneid=rownames(res_subtype_RP)) %>% mutate(RP_sig=case_when(padj < sig & log2FoldChange < 0 ~ "lower", padj < sig & log2FoldChange > 0 ~ "higher", padj > sig ~ "not significant"))
res_diffTE_df <- data.frame(res_diffTE, contrast="subtypeTE", Geneid=rownames(res_diffTE)) %>% mutate(diffTE_sig=case_when(padj < TE_sig & log2FoldChange < 0 ~ "lower", padj < TE_sig & log2FoldChange > 0 ~ "higher", padj > TE_sig ~ "not significant"))
res_TE_CPN_df <- data.frame(res_TE_CPN, contrast="TE_CPN", Geneid=rownames(res_TE_CPN)) %>% mutate(TE_CPN_sig=case_when(padj < sig & log2FoldChange < 0 ~ "lower", padj < sig & log2FoldChange > 0 ~ "higher", padj > sig ~ "not significant"))
res_TE_SCPN_df <- data.frame(res_TE_SCPN, contrast="TE_SCPN", Geneid=rownames(res_TE_SCPN)) %>% mutate(TE_SCPN_sig=case_when(padj < sig & log2FoldChange < 0 ~ "lower", padj < sig & log2FoldChange > 0 ~ "higher", padj > sig ~ "not significant"))

mRNA_MAplot <- ggplot(filter(res_AF_df, is.na(mRNA_sig)==F), aes(y=log2FoldChange,x=log10(baseMean), color=mRNA_sig))+JF_theme+geom_point(aes=0.5, size=0.2)+scale_color_manual(values=c("red","blue","grey"))+ylim(-6,6)+theme(legend.position = c(1 , 0.8), legend.key.size = unit(0.5,"line"))
RPF_MAplot <- ggplot(filter(res_RP_df, is.na(RP_sig)==F) , aes(y=log2FoldChange,x=log10(baseMean), color=RP_sig))+JF_theme+geom_point(aes=0.5, size=0.2)+scale_color_manual(values=c("red","blue","grey"))+ylim(-6,6)+theme(legend.position = c(1, 0.8), legend.key.size = unit(0.5,"line"))
TE_MAplot <- ggplot(filter(res_diffTE_df, is.na(diffTE_sig)==F) , aes(y=log2FoldChange,x=log10(baseMean), color=diffTE_sig))+JF_theme+geom_point(aes=0.5, size=0.2)+scale_color_manual(values=c("red","blue","grey"))+ylim(-6,6)+theme(legend.position = c(1, 0.8), legend.key.size = unit(0.5,"line"))

subtypeComparison_res_df <- bind_rows(res_AF_df, res_RP_df, res_diffTE_df, res_TE_CPN_df, res_TE_SCPN_df) 
subtypeComparison_res_df  <- subtypeComparison_res_df %>% left_join(mouse_gene_symbols_wGeneid, by="Geneid")
RPvsAF_df <- subtypeComparison_res_df %>% filter(contrast %in% c("mRNA","RPF") & is.na(padj)==F) %>% dplyr::select(Geneid, log2FoldChange, contrast) %>% spread(key="contrast",value=log2FoldChange) %>% left_join(res_diffTE_df, by="Geneid") %>% dplyr::select(Geneid, mRNA, RPF, diffTE_sig)
RPvsAF_scatterplot <- ggplot(RPvsAF_df, aes(x=mRNA,y=RPF))+JF_theme+geom_point(alpha=0.5, size=0.2)+geom_hline(yintercept=0)+geom_vline(xintercept=0)+xlab("mRNA log2FoldChange (SCPN/CPN)")+ylab("RPF log2FoldChange (SCPN/CPN)")+ggtitle(paste("Pearson's R=",round(cor(RPvsAF_df$mRNA,RPvsAF_df$RPF,method="pearson"),3)))+theme(legend.position = c(0.75, 0.75))
RPvsAF_scatterplot 

TE_CPNvsSCPN_df <- subtypeComparison_res_df %>% filter(contrast %in% c("TE_CPN","TE_SCPN") & is.na(padj)==F) %>% dplyr::select(Geneid, log2FoldChange, contrast) %>% spread(key="contrast",value=log2FoldChange) %>% left_join(res_diffTE_df, by="Geneid") %>% dplyr::select(Geneid, TE_CPN, TE_SCPN, diffTE_sig) %>% left_join(dplyr::select(res_TE_SCPN_df, Geneid, TE_SCPN_sig), by="Geneid") %>% left_join(dplyr::select(res_TE_CPN_df, Geneid, TE_CPN_sig), by="Geneid")

TE_CPNvsSCPN_scatterplot <- ggplot(TE_CPNvsSCPN_df, aes(x=TE_CPN,y=TE_SCPN, color=diffTE_sig))+JF_theme+geom_point(alpha=1, size=0.2)+geom_hline(yintercept=0)+geom_vline(xintercept=0)+scale_color_manual(values=c("red","blue","grey","grey"))+xlab("TE in CPN")+ylab("TE in SCPN")+ggtitle(paste("Pearson's R=",round(cor(TE_CPNvsSCPN_df$TE_CPN,TE_CPNvsSCPN_df$TE_SCPN,method="pearson"),3)))+theme(legend.position = c(0.85, 0.3), legend.key.size = unit(0.5,"line"))
TE_CPNvsSCPN_scatterplot

TE_CPNvsSCPN_scatterplot_TE_CPN_sig_colored <- ggplot(TE_CPNvsSCPN_df, aes(x=TE_CPN,y=TE_SCPN, color=TE_CPN_sig))+JF_theme+geom_point(alpha=0.5, size=0.2)+geom_hline(yintercept=0)+geom_vline(xintercept=0)+scale_color_manual(values=c("red","blue","black","grey"))+xlab("TE in CPN")+ylab("TE in SCPN")+ggtitle("T.E. in SCPN vs. T.E. in CPN, skewed TE in CPN highlighted ")+theme(legend.position = c(0.85, 0.3), legend.key.size = unit(0.5,"line"))
TE_CPNvsSCPN_scatterplot_TE_CPN_sig_colored 

TE_CPNvsSCPN_scatterplot_TE_SCPN_sig_colored <- ggplot(TE_CPNvsSCPN_df, aes(x=TE_CPN,y=TE_SCPN, color=TE_SCPN_sig))+JF_theme+geom_point(alpha=0.5, size=0.2)+geom_hline(yintercept=0)+geom_vline(xintercept=0)+scale_color_manual(values=c("red","blue","black","grey"))+xlab("TE in CPN")+ylab("TE in SCPN")+ggtitle("T.E. in SCPN vs. T.E. in CPN, skewed TE in CPN highlighted ")+theme(legend.position = c(0.85, 0.3), legend.key.size = unit(0.5,"line"))
TE_CPNvsSCPN_scatterplot_TE_SCPN_sig_colored 

log2FC_histo_plot <- filter(subtypeComparison_res_df, (contrast %in% c("TE_SCPN","TE_CPN")==F)) %>%
  ggplot(aes(x=log2FoldChange, fill=contrast))+JF_theme+geom_histogram(alpha=0.5, bins=400, position="identity")+ylab("gene count")
log2FC_histo_plot

log2FC_ecdf_plot <- filter(subtypeComparison_res_df, (contrast %in% c("TE_SCPN","TE_CPN")==F)) %>%
  ggplot(aes(x=log2FoldChange, color=contrast))+JF_theme+stat_ecdf()+ylab("cumlative_fraction")
log2FC_ecdf_plot

```

###Are AF and RP dispersions different?
###Answer: yes. dispersion correlation is 0.2857. Have to draw them from different distributions. 

```{r AFvsRP_DESeq2}

AF_dds <- runDESeq_bySubtype(filter(featureCounts_tidy_sampFilt, exp=="AF"))
RP_dds <- runDESeq_bySubtype(filter(featureCounts_tidy_sampFilt, exp=="RP"))

AF_fullDEseq <- data.frame(mcols(AF_dds), exp="AF", Geneid=rownames(mcols(AF_dds))) %>% filter(allZero==F)
RP_fullDEseq <- data.frame(mcols(RP_dds),exp="RP", Geneid=rownames(mcols(RP_dds))) %>% filter(allZero==F)
AF_fullDEseq <- filter(AF_fullDEseq, (Geneid %in% RP_fullDEseq$Geneid)==T)
RP_fullDEseq <- filter(RP_fullDEseq, (Geneid %in% AF_fullDEseq$Geneid)==T)

AF_RP_full_DEseq <- bind_rows(RP_fullDEseq, AF_fullDEseq)  %>% filter(Geneid %in% subtypeComparison_res_df$Geneid)

AF_RP_distGeneEst <- AF_RP_full_DEseq %>% dplyr::select(Geneid, dispGeneEst, exp) %>% spread(key=exp, value=dispGeneEst)
AFvsRP_dispGeneEstPlot <- ggplot(AF_RP_distGeneEst, aes(x=AF,y=RP))+JF_theme+geom_point()+xlim(c(0,2.5))+ylim(c(0,2.5))

AF_RP_dispersions <- AF_RP_full_DEseq %>% dplyr::select(Geneid, dispersion, exp) %>% spread(key=exp, value=dispersion)
AFvsRP_dispersionPlot <- ggplot(AF_RP_dispersions, aes(x=AF,y=RP))+JF_theme+geom_point()

```

###Make AF and RP sim data sets using the estimated dispersion parameters.
###Deep knowledge here: baseMean represents the mean counts per sample after normalizing for libSize!
###Also size=1/dispersion, where dispersion is the DESeq2 dispersion. 
###Got 6/9949 diff exp genes at padj < 0.1 for simulated RP data, and 4/9949 for AF.
###Nice flat pval distros too for n1 vs n2 comparison. 
###DiffTE seems to have nice null distros in the sim data. Thousands of diff AF vs RP genes, but just 5 diff n1 vs n2 genes, and 5 diffTE expRP.conditionn2 genes.
###Setting the baseMean of AF and RP for sim to be baseMean (AF+RP)/2 is even better. 0 diffTE genes, 3 diff RPvsAF genes, 3 condition n1v2 genes. Totally null. 

```{r sim_setup}
AF_fullDEseq_filt <- AF_fullDEseq %>% filter(Geneid %in% subtypeComparison_res_df$Geneid)
RP_fullDEseq_filt <- RP_fullDEseq %>% filter(Geneid %in% subtypeComparison_res_df$Geneid)

drawFromNbinom <- function(baseMean, dispersion,Geneid, n=8) {
  #returnList=list("bM"=baseMean,"disp"=dispersion)
  draws <- rnbinom(n=n, size=1/dispersion, mu=baseMean)
  return(c(Geneid, draws))
}

simFrom_fullDESeq <- function(fullDESeq, seed=1182, cond_n=4, expType="AF") {
  set.seed(seed)
  sim <- mapply(drawFromNbinom, fullDESeq$baseMean, fullDESeq$dispersion, fullDESeq$Geneid, cond_n*2)
  sim <- t(sim)
  simSampNames <- c(paste(paste(expType,"S",seq(1:cond_n),sep=""),"n1",sep="_")
    ,paste(paste(expType, "S",seq(1:cond_n),sep=""),"n2",sep="_")) 
  colnames(sim) <- c("Geneid",simSampNames)
  rownames(sim) <- sim[,1]
  sim <- matrix(as.numeric(sim[,-1]), ncol=ncol(sim[,-1]), dimnames=list(rownames(sim),colnames(sim[,-1])))
  sim <- data.frame(sim)
}

DESeq2_RNAonly_runme <- function(countsData) {
  colData <- data.frame(condition=str_extract(colnames(countsData), regex("n.")))
  #return(colData)
  rownames(colData) <- colnames(countsData)
  dds <- DESeqDataSetFromMatrix(countData = countsData, colData = colData, design=~condition)
  dds <- DESeq(dds)
  res <- data.frame(results(dds))
}
sig_threshold <- 0.1

###set the baseMean in both AF and RP to be the same
common_baseMeans <- (AF_fullDEseq_filt$baseMean+RP_fullDEseq_filt$baseMean)/2
AF_fullDE_commonMeans <- data.frame(Geneid=AF_fullDEseq_filt$Geneid, baseMean=common_baseMeans, dispersion=AF_fullDEseq_filt$dispersion )
RP_fullDE_commonMeans <- data.frame(Geneid=RP_fullDEseq_filt$Geneid, baseMean=common_baseMeans, dispersion=RP_fullDEseq_filt$dispersion)


AF_sim <- simFrom_fullDESeq(fullDESeq=AF_fullDE_commonMeans, cond_n=4)
AF_sim_DEseq_res  <- DESeq2_RNAonly_runme(AF_sim)
AF_sim2 <- simFrom_fullDESeq(fullDESeq=AF_fullDE_commonMeans, cond_n=4, seed=81523)
RP_sim <- simFrom_fullDESeq(fullDESeq=RP_fullDE_commonMeans, cond_n=4, exp="RP")
RP_sim_DEseq_res  <- DESeq2_RNAonly_runme(RP_sim)
RP_sim2 <- simFrom_fullDESeq(fullDESeq=RP_fullDE_commonMeans, cond_n=4, exp="RP", seed=81523)

# AF_sim <- simFrom_fullDESeq(fullDESeq=AF_fullDEseq_filt, cond_n=4)
# AF_sim_DEseq_res  <- DESeq2_RNAonly_runme(AF_sim)
# RP_sim <- simFrom_fullDESeq(fullDESeq=RP_fullDEseq_filt, cond_n=4, exp="RP")
# RP_sim_DEseq_res  <- DESeq2_RNAonly_runme(RP_sim)

AF_RP_concat_sim <- cbind(AF_sim, RP_sim)

run_diffTE <- function(sim) {
  #take in a sim with both labeled AF and RP samples, and run DESeq2,
  #testing for both differences between samples and interaction term.
  #assumes sample names begin with {exp}S_n{condition}
  colDataSim <- data.frame(condition=str_extract(colnames(sim), regex("n.")),
                        exp=str_extract(colnames(sim), regex("[^S]*")))
  rownames(colDataSim) <- colnames(sim)
  dds <- DESeqDataSetFromMatrix(countData = sim, colData = colDataSim, design=~exp*condition)
  dds <- DESeq(dds)
}

AF_RP_sim_dds <- run_diffTE(AF_RP_concat_sim)
```
###fit injected fold change with a gamma distro, shape=2, rate=1.5 (might be a bit heavy tailed).
###Test with AF_sim and RP_sim was good (baseMeans from each)! FDR <0.1 158 n1vs2 DE genes. 129 TP, 71 FN, 29 FP. 82% precision, 65% recall. 65% sensitivity, 99% specificity. 
###With common means, got 151 TP, 41 FN, 27 FP. 75.5% sensitivity, 99% specific. 
```{r transcriptional_change_injection}
mRNA_lFC_distributions <- ggplot(subtypeComparison_res_df %>% filter(contrast=="mRNA"), aes(x=log2FoldChange, color=mRNA_sig))+JF_theme+geom_density()
RP_lFC_distributions <- ggplot(subtypeComparison_res_df %>% filter(contrast=="RPF"), aes(x=log2FoldChange, color=RP_sig))+JF_theme+geom_density()

mRNA_lFC_sig_means <- subtypeComparison_res_df %>% filter(contrast=="mRNA") %>%
  group_by(mRNA_sig) %>% summarize(meanLFC=mean(log2FoldChange),medianLFC=median(log2FoldChange))
RPF_lFC_sig_means <- subtypeComparison_res_df %>% filter(contrast=="RPF") %>%
  group_by(RP_sig) %>% summarize(meanLFC=mean(log2FoldChange), medianLFC=median(log2FoldChange))

inject_transcriptional_change <- function(AF_sim, RP_sim, n_up=100, n_down=100, gamma_shape=2, gamma_rate=1.5, minFC=0.25, seed=62189) {
  AF <- data.frame(AF_sim, Geneid=rownames(AF_sim))
  RP <- data.frame(RP_sim, Geneid=rownames(RP_sim))
  AF <- AF %>% gather(key="sample",value="counts", -Geneid) %>% mutate(condition=str_extract(sample, regex("n.")))
  RP <- RP %>% gather(key="sample",value="counts", -Geneid) %>% mutate(condition=str_extract(sample, regex("n.")))
  
  mRNA_TruePos_Genes <- sample(rownames(AF_sim), size=(n_up+n_down) )
  downGenes <- mRNA_TruePos_Genes[1:n_down]
  upGenes <- mRNA_TruePos_Genes[(n_down+1):(n_up+n_down)]
  downFCs <- 2^(rgamma(n=n_down,shape=gamma_shape, rate=gamma_rate)+minFC)
  upFCs <- 2^(rgamma(n=n_down,shape=gamma_shape, rate=gamma_rate)+minFC)
  #for genes which go up, multiply n2 by FC. For genes that go down, multiply n1 by FC.
  up <- data.frame(Geneid=upGenes, FCs=upFCs, TP="up")
  down <- data.frame(Geneid=downGenes, FCs=downFCs, TP="down")
  mRNA_FCs <- bind_rows(up, down)
  
  AF <- AF %>% left_join(mRNA_FCs, by="Geneid") %>% mutate(FCs=case_when(
   TP=="up" & condition=="n2" ~ FCs,
   TP=="up" & condition=="n1" ~ 1,
   TP=="down" & condition=="n2" ~ 1,
   TP=="down" & condition=="n1" ~ FCs,
   is.na(TP)==T ~ 1
  )) %>% mutate(new_counts=round(counts*FCs,0))
  RP <- RP %>% left_join(mRNA_FCs, by="Geneid") %>% mutate(FCs=case_when(
   TP=="up" & condition=="n2" ~ FCs,
   TP=="up" & condition=="n1" ~ 1,
   TP=="down" & condition=="n2" ~ 1,
   TP=="down" & condition=="n1" ~ FCs,
   is.na(TP)==T ~ 1
  )) %>% mutate(new_counts=round(counts*FCs,0))
  returnList <- list("AF"=AF,"RP"=RP)
  #return(list(mRNA_TruePos_Genes, downGenes, upGenes))
}

AFandRP_sim_mRNA_injection <- inject_transcriptional_change(AF_sim, RP_sim, n_up=100, n_down=100)

injection2DESeq_res <- function(injection_list) {
  #Take in injected, simulated AF and RP data, run DESeq2. Return the dds and results_df for all 3 comparisons (condition, exp, interaction)
  AF <- injection_list$AF
  RP <- injection_list$RP
  AF_TruePos <- AF %>% dplyr::select(Geneid, condition, TP, FCs) %>%
    filter((TP=="down" & condition=="n1") | (TP=="up" & condition=="n2") | (TP=="up_buffered" & condition=="n2") | (TP=="down_buffered" & condition=="n1") | (TP=="up_intensified" & condition=="n2") | (TP=="down_intensified" & condition=="n1")) %>% distinct()
  RP_TruePos <- RP %>% dplyr::select(Geneid, condition, TP, FCs) %>% filter((TP=="diffTE_down" & condition=="n1") | (TP=="diffTE_up" & condition=="n2") | (TP=="up_buffered" & condition=="n2") | (TP=="down_buffered" & condition=="n1") | (TP=="up_intensified" & condition=="n2") | (TP=="down_intensified" & condition=="n1")  ) %>% distinct()
  #return(list(AF_TruePos, RP_TruePos))
  AF_simulation <- AF %>% dplyr::select(Geneid, sample, new_counts) %>%
    spread(key="sample",value="new_counts") %>% column_to_rownames("Geneid")
  RP_simulation <- RP %>% dplyr::select(Geneid, sample, new_counts) %>%
    spread(key="sample",value="new_counts") %>% column_to_rownames("Geneid")
  AFandRP_simulation <- cbind(AF_simulation, RP_simulation)
  AFandRP_dds <- run_diffTE(AFandRP_simulation)
  AFvsRP_res <- data.frame(results(AFandRP_dds, name="exp_RP_vs_AF"), contrast="AFvsRP", Geneid=rownames(AFandRP_simulation)) %>% left_join(AF_TruePos, by="Geneid")
  n1vs2_res <- data.frame(results(AFandRP_dds, name="condition_n2_vs_n1"), contrast="n1v2", Geneid=rownames(AFandRP_simulation)) %>% left_join(AF_TruePos, by="Geneid")
  TE_res <- data.frame(results(AFandRP_dds, name="expRP.conditionn2"), contrast="diffTE", Geneid=rownames(AFandRP_simulation)) %>% left_join(RP_TruePos, by="Geneid")
  full_res <- bind_rows(AFvsRP_res, n1vs2_res, TE_res) 
  return(list("full_res"=full_res, "dds"=AFandRP_dds))
}
AFandRP_mRNA_injection_results <- injection2DESeq_res(AFandRP_sim_mRNA_injection)
AFandRP_mRNA_injection_sig_OR_TP <- filter(AFandRP_mRNA_injection_results$full_res, contrast=="n1v2") %>% filter(padj < sig_threshold | is.na(TP)==F)
```
```{r ROC_analysis}

res2pROC <- function(res_df, contrastOfInterest) {
  #Take in a res_df, pull out the contrast of interest. Make the roc.
  res <- res_df %>% filter(contrast==contrastOfInterest) %>% arrange(padj) %>% mutate(response=case_when(
    padj > sig_threshold ~ 0,
    padj < sig_threshold ~ 1
  ), 
  predictor=case_when(
    is.na(TP)==T ~ 0,
    is.na(TP)==F ~ 1
  ))
  theRoc <- roc(response=res$predictor, predictor=res$padj)
}
AFandRP_mRNA_injection_roc <- res2pROC(res_df=AFandRP_mRNA_injection_results$full_res, contrastOfInterest = "n1v2")

precisionAndRecall <- function(res, theContrast, sig=0.1, TP_terms) {
  #Take in results from a diffTE_injection+DESeq2 run, and a list of TP_terms to consider as true positives, and calculate precision and recall. 
  res_filt <-filter(res, contrast==theContrast) %>% mutate(call=case_when(
    (TP %in% TP_terms)==T & padj < sig ~ "TP",
    (TP %in% TP_terms)==T & (padj > sig | is.na(padj)==T) ~ "FN",
    (TP %in% TP_terms)==F & padj < sig ~ "FP",
    (TP %in% TP_terms)==F & padj > sig | (padj > sig | is.na(padj)==T) ~ "TN"
  )) %>% filter(is.na(call)==F)
  confusion_matrix <- res_filt %>% group_by(call) %>% summarize(n=n()) 
  precision=confusion_matrix[confusion_matrix$call=="TP",]$n/(confusion_matrix[confusion_matrix$call=="TP",]$n+confusion_matrix[confusion_matrix$call=="FP",]$n)
  recall=confusion_matrix[confusion_matrix$call=="TP",]$n/(confusion_matrix[confusion_matrix$call=="TP",]$n+confusion_matrix[confusion_matrix$call=="FN",]$n)
  return(list("precision"=precision, "recall"=recall, "confusion_matrix"=confusion_matrix))
}


PR_curve <- function(res, theContrast, TP_terms) {
  #Take in results from a diffTE_injection+DESeq2 run, and a list of TP_terms to consider as true positives, and calculate precision and recall. 
  res_filt <-filter(res, contrast==theContrast) %>% arrange(padj) %>%
    mutate(TruePos=case_when((TP %in% TP_terms)==T ~ 1, (TP %in% TP_terms)==F~ 0),
           FalsePos=case_when( (TP %in% TP_terms)==T ~ 0, (TP %in% TP_terms)==F~ 1))
    numTruePos <- sum(res_filt$TruePos, na.rm=T)
    res_filt <- res_filt %>% mutate(recall=cumsum(TruePos)/numTruePos, precision=cumsum(TruePos)/(cumsum(TruePos)+cumsum(FalsePos)))
    PR_curve <- ggplot(res_filt, aes(x=recall, y=precision))+JF_theme+geom_line()
}

#test <- PR_curve(sim_mRNA4000_TE40_b40_i40_res$full_res, theContrast="diffTE", TP_terms=c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
```

###Even without buffering or intensification, DESeq2 doesn't do great. 
###Precision ~50-65%, recall=~20% for all conditions. 
###Using a narrower gamma function for TE fold changes, but with a higher minimum.
###When I pass these narrower FCs in for diffExp, DESeq does ok (90% precision, 58% recall, AUC=0.92)

```{r diffTE_injection}

inject_diffTE <- function(AF_sim, RP_sim, n_up=400, n_down=400, gamma_shape=1, gamma_rate=1.5, minFC=0.25, seed=62189, diffTE=40, buffered=0, intensified=0,TEgamma_shape=1,TEgamma_rate=2.5, TEminFC=1) {
  set.seed(seed)
  AF <- data.frame(AF_sim, Geneid=rownames(AF_sim))
  RP <- data.frame(RP_sim, Geneid=rownames(RP_sim))
  AF <- AF %>% gather(key="sample",value="counts", -Geneid) %>% mutate(condition=str_extract(sample, regex("n.")))
  RP <- RP %>% gather(key="sample",value="counts", -Geneid) %>% mutate(condition=str_extract(sample, regex("n.")))
#return(RP)
Geneids <- data.frame(Geneid=rownames(AF_sim))
 mRNA_TruePos_Genes <- sample(Geneids$Geneid, size=(n_up+n_down) )
 mRNA_TrueNeg_Genes <- filter(Geneids, (Geneid %in% mRNA_TruePos_Genes)==F)$Geneid
 #return(list("TP"=mRNA_TruePos_Genes,"TN"=mRNA_TrueNeg_Genes))
  downGenes <- mRNA_TruePos_Genes[1:n_down]
  upGenes <- mRNA_TruePos_Genes[(n_down+1):(n_up+n_down)]
  downFCs <- 2^(rgamma(n=n_down,shape=gamma_shape, rate=gamma_rate)+minFC)
  upFCs <- 2^(rgamma(n=n_down,shape=gamma_shape, rate=gamma_rate)+minFC)
  #for genes which go up, multiply n2 by FC. For genes that go down, multiply n1 by FC.
  up <- data.frame(Geneid=upGenes, FCs=upFCs, TP="up")
  down <- data.frame(Geneid=downGenes, FCs=downFCs, TP="down")
  mRNA_FCs <- bind_rows(up, down)

###select the diffTE genes from the TrueNeg genes. diffTE_num is just diffTE/2... want half up and half down
diffTE_num <- round(diffTE/2)
diffTE_genes <- sample(mRNA_TrueNeg_Genes, size=diffTE)
diffTE_downGenes <-diffTE_genes[1:diffTE_num ]
diffTE_upGenes <- diffTE_genes[(diffTE_num+1):(diffTE)]
diffTE_downFCs <- 2^(rgamma(n=diffTE_num,shape=TEgamma_shape, rate=TEgamma_rate)+TEminFC)
diffTE_upFCs <- 2^(rgamma(n=diffTE_num,shape=TEgamma_shape, rate=TEgamma_rate)+TEminFC)
diffTE_up <- data.frame(Geneid=diffTE_upGenes, FCs=diffTE_upFCs, TP="diffTE_up")
diffTE_down <- data.frame(Geneid=diffTE_downGenes, FCs=diffTE_downFCs, TP="diffTE_down")

##add in genes to be intensified or buffered
buffedIntensed_Geneids <- sample(mRNA_TruePos_Genes, size=(buffered+intensified))
buffed_genes <- buffedIntensed_Geneids[1:buffered]
intensed_genes <- buffedIntensed_Geneids[(buffered+1):(buffered+intensified)]
# buffed_FCs <- 2^(rgamma(n=buffered,shape=gamma_shape, rate=gamma_rate)+minFC)
# intensed_FCs <- 2^(rgamma(n=buffered,shape=gamma_shape, rate=gamma_rate)+minFC)

#label which genes in the mRNA_FCs are buffered (their fold change stays the same in the AF/mRNA data)

mRNA_FCs <- mRNA_FCs %>%
  mutate(TP=case_when(
    Geneid %in% buffed_genes & TP=="up" ~ "up_buffered",
    Geneid %in% buffed_genes & TP=="down" ~ "down_buffered",
    Geneid %in% intensed_genes & TP=="up" ~ "up_intensified",
    Geneid %in% intensed_genes & TP=="down" ~ "down_intensified",
    (Geneid %in% buffed_genes)==F & (Geneid %in% intensed_genes)==F ~ TP
  ))

#return(mRNA_FCs)
#multiply the mRNA fold changes by the buffed or intensed fold changes in the RP samples. 
#Draw a random point from the gamma distribution to get the buffering/inte sifying fold change. 
#up_buffered means mRNA goes up, TE goes down, so RP=FC/gamma in n2 (denominator). down buffered means mRNA goes down, so RP=FC/gamma in n1. 
#intensified up means RNA and RP both go up in n2, intensified 2 means both RNA and RP are higher in n1 than n2. 


RP_FCs <- mRNA_FCs

RP_FCs <- bind_rows(RP_FCs, diffTE_up, diffTE_down)
mRNA_FCs <- bind_rows(mRNA_FCs, diffTE_up, diffTE_down)
  AF <- AF %>% left_join(mRNA_FCs, by="Geneid") %>% mutate(FCs=case_when(
   TP=="up" & condition=="n2" ~ FCs,
   TP=="up" & condition=="n1" ~ 1,
   TP=="down" & condition=="n2" ~ 1,
   TP=="down" & condition=="n1" ~ FCs,
   TP=="up_buffered" & condition=="n2" ~ FCs,
   TP=="up_buffered" & condition=="n1" ~ 1,
   TP=="down_buffered" & condition=="n2" ~ 1,
   TP=="down_buffered" & condition=="n1" ~ FCs,
   TP=="up_intensified" & condition=="n2" ~ FCs,
   TP=="up_intensified" & condition=="n1" ~ 1,
   TP=="down_intensified" & condition=="n2" ~ 1,
   TP=="down_intensified" & condition=="n1" ~ FCs,
   is.na(TP)==T ~ 1,
   (TP %in% c("diffTE_up","diffTE_down"))==T ~ 1
  )) %>% mutate(new_counts=round(counts*FCs,0))
  RP <- RP %>% left_join(RP_FCs, by="Geneid") %>% mutate(FCs=case_when(
   TP=="diffTE_up" & condition=="n2" ~ FCs,
   TP=="diffTE_up" & condition=="n1" ~ 1,
   TP=="diffTE_down" & condition=="n2" ~ 1,
   TP=="diffTE_down" & condition=="n1" ~ FCs,
   TP=="up" & condition=="n2" ~ FCs,
   TP=="up" & condition=="n1" ~ 1,
   TP=="down" & condition=="n2" ~ 1,
   TP=="down" & condition=="n1" ~ FCs,
   TP=="up_buffered" & condition=="n1" ~ 1,
   TP=="up_buffered" & condition=="n2" ~ FCs/(2^(rgamma(n=1,shape=TEgamma_shape, rate=TEgamma_rate)+TEminFC)),
   TP=="down_buffered" & condition=="n1" ~ FCs/(2^(rgamma(n=1,shape=TEgamma_shape, rate=TEgamma_rate)+TEminFC)),
   TP=="down_buffered" & condition=="n2" ~ 1,
   TP=="up_intensified" & condition=="n1" ~ 1 ,
   TP=="up_intensified" & condition=="n2" ~ FCs*(2^(rgamma(n=1,shape=TEgamma_shape, rate=TEgamma_rate)+TEminFC)),
   TP=="down_intensified" & condition=="n1" ~ FCs*(2^(rgamma(n=1,shape=TEgamma_shape, rate=TEgamma_rate)+TEminFC)),
   TP=="down_intensified" & condition=="n2" ~ 1,
   is.na(TP)==T ~ 1
  )) %>% mutate(new_counts=round(counts*FCs,0))
  returnList <- list("AF"=AF,"RP"=RP)
}

#Buffering+Intensification, 800 exp genes, 120 total diff genes
sim_mRNA800_TE40_b40_i40 <- inject_diffTE(AF_sim=AF_sim, RP_sim=RP_sim, diffTE=40, buffered=40, intensified=40)
sim_mRNA800_TE40_b40_i40_res <- injection2DESeq_res(sim_mRNA800_TE40_b40_i40)
sim_mRNA800_TE40_b40_i40_PR <- precisionAndRecall(sim_mRNA800_TE40_b40_i40_res$full_res, theContrast="diffTE", TP_terms=c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim_mRNA800_TE40_b40_i40_PR_curve <- PR_curve(sim_mRNA800_TE40_b40_i40_res$full_res, theContrast="diffTE", TP_terms=c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim_mRNA800_TE40_b40_i40_roc <- res2pROC(sim_mRNA800_TE40_b40_i40_res$full_res, contrastOfInterest = "diffTE")

sim2_mRNA800_TE40_b40_i40 <- inject_diffTE(AF_sim=AF_sim2, RP_sim=RP_sim2, diffTE=40, buffered=40, intensified=40, seed=81523)
sim2_mRNA800_TE40_b40_i40_res <- injection2DESeq_res(sim2_mRNA800_TE40_b40_i40)
sim2_mRNA800_TE40_b40_i40_PR <- precisionAndRecall(sim2_mRNA800_TE40_b40_i40_res$full_res, theContrast="diffTE", TP_terms=c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim2_mRNA800_TE40_b40_i40_PR_curve <- PR_curve(sim2_mRNA800_TE40_b40_i40_res$full_res, theContrast="diffTE", TP_terms=c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim2_mRNA800_TE40_b40_i40_roc <- res2pROC(sim2_mRNA800_TE40_b40_i40_res$full_res, contrastOfInterest = "diffTE")

#Buffering+Intensification, 4000 exp genes, 120 total diff genes
sim_mRNA4000_TE40_b40_i40 <- inject_diffTE(AF_sim=AF_sim, RP_sim=RP_sim, diffTE=40, buffered=40, intensified=40, n_up=2000, n_down=2000)
sim_mRNA4000_TE40_b40_i40_res <- injection2DESeq_res(sim_mRNA4000_TE40_b40_i40)
sim_mRNA4000_TE40_b40_i40_PR <- precisionAndRecall(sim_mRNA4000_TE40_b40_i40_res$full_res, theContrast="diffTE", TP_terms=c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim_mRNA4000_TE40_b40_i40_PR_curve <- PR_curve(sim_mRNA4000_TE40_b40_i40_res$full_res, theContrast="diffTE", TP_terms=c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim_mRNA4000_TE40_b40_i40_roc <- res2pROC(sim_mRNA4000_TE40_b40_i40_res$full_res, contrastOfInterest = "diffTE")

sim2_mRNA4000_TE40_b40_i40 <- inject_diffTE(AF_sim=AF_sim2, RP_sim=RP_sim2, diffTE=40, buffered=40, intensified=40, n_up=2000, n_down=2000, seed=81523)
sim2_mRNA4000_TE40_b40_i40_res <- injection2DESeq_res(sim2_mRNA4000_TE40_b40_i40)
sim2_mRNA4000_TE40_b40_i40_PR <- precisionAndRecall(sim2_mRNA4000_TE40_b40_i40_res$full_res, theContrast="diffTE", TP_terms=c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim2_mRNA4000_TE40_b40_i40_PR_curve <- PR_curve(sim2_mRNA4000_TE40_b40_i40_res$full_res, theContrast="diffTE", TP_terms=c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim2_mRNA4000_TE40_b40_i40_roc <- res2pROC(sim2_mRNA4000_TE40_b40_i40_res$full_res, contrastOfInterest = "diffTE")

# 4000 exp genes, 10 total diff genes
sim_mRNA4000_TE10_b0_i0 <- inject_diffTE(AF_sim=AF_sim, RP_sim=RP_sim, diffTE=10, buffered=0, intensified=0, n_up=2000, n_down=2000)
sim_mRNA4000_TE10_b0_i0_res <- injection2DESeq_res(sim_mRNA4000_TE10_b0_i0)


#No buffering or intensification, 800 diffExp, 40 diffTE
sim_mRNA800_TE40_b0_i0 <- inject_diffTE(AF_sim=AF_sim, RP_sim=RP_sim)
sim_mRNA800_TE40_b0_i0_res <- injection2DESeq_res(sim_mRNA800_TE40_b0_i0)
sim_mRNA800_TE40_b0_i0_PR <- precisionAndRecall(sim_mRNA800_TE40_b0_i0_res$full_res, theContrast="diffTE", TP_terms=c("diffTE_down","diffTE_up"))
sim_mRNA800_TE40_b0_i0_roc <- res2pROC(res_df = sim_mRNA800_TE40_b0_i0_res$full_res, contrastOfInterest = "diffTE")

###See how well difExp works with the narrower gamma distro used for diffTE
###supplied to generate mRNA fold changes.

sim_mRNA800_TE40_b0_i0_gr1_gs2.5 <- inject_diffTE(AF_sim=AF_sim, RP_sim=RP_sim, gamma_shape = 1, gamma_rate=2.5, minFC=0.5)
sim_mRNA800_TE40_b0_i0_gr1_gs2.5_res <- injection2DESeq_res(sim_mRNA800_TE40_b0_i0_gr1_gs2.5)
sim_mRNA800_TE40_b0_i0_gr1_gs2.5_mRNA_PR <- precisionAndRecall(sim_mRNA800_TE40_b0_i0_gr1_gs2.5_res$full_res, theContrast = "n1v2", TP_terms = c("up","down"))
sim_mRNA800_TE40_b0_i0_gr1_gs2.5_mRNA_PR_curve <- PR_curve(sim_mRNA800_TE40_b0_i0_gr1_gs2.5_res$full_res, theContrast = "n1v2", TP_terms = c("up","down"))
sim_mRNA800_TE40_b0_i0_gr1_gs2.5_mRNA_pROC <- res2pROC(sim_mRNA800_TE40_b0_i0_gr1_gs2.5_res$full_res, contrast="n1v2")

#no diffTE or Exp
sim_mRNA200_TE10_b0_i0<- inject_diffTE(AF_sim=AF_sim, RP_sim=RP_sim, diffTE=10, buffered=0, intensified=0, n_up=100, n_down=100)
sim_mRNA0_TE0_b0_i0_res <- injection2DESeq_res(sim_mRNA200_TE10_b0_i0)
#sim_mRNA0_TE0_b0_i0_mRNA_PR <- precisionAndRecall(sim_mRNA0_TE0_b0_i0$full_res, theContrast = "n1v2", TP_terms = c("up","down"))
#sim_mRNA0_TE0_b0_i0 <- res2pROC(sim_mRNA0_TE0_b0_i0$full_res, contrast="n1v2")

plot_diffTE_from_injection <- function(res) {
  res_filt <- res %>% filter(contrast=="n1v2" | contrast=="diffTE") 
    MAplot <- ggplot(res_filt, aes(x=log10(baseMean),y=log2FoldChange,color=is.na(TP)))+JF_theme+facet_wrap(~contrast)+geom_point(size=0.2, alpha=0.5)+theme(legend.position = "top")+scale_color_manual(values=c("red","grey"))
}

sim_mRNA4000_TE40_b40_i40_injection_MAplots <- plot_diffTE_from_injection(sim_mRNA4000_TE40_b40_i40_res$full_res)
sim2_mRNA4000_TE40_b40_i40_injection_MAplots <- plot_diffTE_from_injection(sim2_mRNA4000_TE40_b40_i40_res$full_res)

sim_mRNA800_TE40_b40_i40_injection_MAplots <- plot_diffTE_from_injection(sim_mRNA800_TE40_b40_i40_res$full_res)+ggtitle("sim1, mRNA800, \nTE 40, buffered 40, intensified 40")
sim2_mRNA800_TE40_b40_i40_injection_MAplots <- plot_diffTE_from_injection(sim2_mRNA800_TE40_b40_i40_res$full_res)+ggtitle("sim2, mRNA4000, \nTE 40, buffered 40, intensified 40")

sim_mRNA4000_TE40_b40_i40_injection_MAplots <- plot_diffTE_from_injection(sim_mRNA4000_TE40_b40_i40_res$full_res)+ggtitle("sim1, mRNA4000, \nTE 40, buffered 40, intensified 40")
sim2_mRNA4000_TE40_b40_i40_injection_MAplots <- plot_diffTE_from_injection(sim2_mRNA4000_TE40_b40_i40_res$full_res)+ggtitle("sim2, mRNA4000, \nTE 40, buffered 40, intensified 40")
```

```{r xtail_on_sim}

injection2xtail <- function(injection_list, bins=10000, ci=0.90, outName) {
  AF <- injection_list$AF
  RP <- injection_list$RP
  AF_TruePos <- AF %>% dplyr::select(Geneid, condition, TP, FCs) %>%
    filter((TP=="down" & condition=="n1") | (TP=="up" & condition=="n2") | (TP=="up_buffered" & condition=="n2") | (TP=="down_buffered" & condition=="n1") | (TP=="up_intensified" & condition=="n2") | (TP=="down_intensified" & condition=="n1")) %>% distinct()
  RP_TruePos <- RP %>% dplyr::select(Geneid, condition, TP, FCs) %>% filter((TP=="diffTE_down" & condition=="n1") | (TP=="diffTE_up" & condition=="n2") | (TP=="up_buffered" & condition=="n2") | (TP=="down_buffered" & condition=="n1") | (TP=="up_intensified" & condition=="n2") | (TP=="down_intensified" & condition=="n1")  ) %>% distinct()
  #return(list(AF_TruePos, RP_TruePos))
  AF_simulation <- AF %>% dplyr::select(Geneid, sample, new_counts) %>%
    spread(key="sample",value="new_counts") %>% column_to_rownames("Geneid")
  RP_simulation <- RP %>% dplyr::select(Geneid, sample, new_counts) %>%
    spread(key="sample",value="new_counts") %>% column_to_rownames("Geneid")
   AF_TruePos <- AF %>% dplyr::select(Geneid, condition, TP, FCs) %>%
    filter((TP=="down" & condition=="n1") | (TP=="up" & condition=="n2") | (TP=="up_buffered" & condition=="n2") | (TP=="down_buffered" & condition=="n1") | (TP=="up_intensified" & condition=="n2") | (TP=="down_intensified" & condition=="n1")) %>% distinct()
  RP_TruePos <- RP %>% dplyr::select(Geneid, condition, TP, FCs) %>% filter((TP=="diffTE_down" & condition=="n1") | (TP=="diffTE_up" & condition=="n2") | (TP=="up_buffered" & condition=="n2") | (TP=="down_buffered" & condition=="n1") | (TP=="up_intensified" & condition=="n2") | (TP=="down_intensified" & condition=="n1")  ) %>% distinct()
  #return(list(AF_TruePos, RP_TruePos))
  AF_simulation <- AF %>% dplyr::select(Geneid, sample, new_counts) %>%
    spread(key="sample",value="new_counts") %>% column_to_rownames("Geneid")
  RP_simulation <- RP %>% dplyr::select(Geneid, sample, new_counts) %>%
    spread(key="sample",value="new_counts") %>% column_to_rownames("Geneid")
  condition <- str_extract(colnames(AF_simulation), regex("n."))
  xtail_res <- xtail(mrna=AF_simulation,rpf=RP_simulation, condition = condition, bins=bins, ci=ci)
  xtail_table <- data.frame(resultsTable(xtail_res), contrast="diffTE")
  xtail_table <- data.frame(xtail_table, Geneid=rownames(xtail_table)) %>% left_join(RP_TruePos, by="Geneid") %>% mutate(padj=pvalue.adjust, pvalue=pvalue_final)
  write.csv(xtail_table, file=outName, quote=F)
  return(xtail_table)
}

#Xtail is slow. Very slow. If the file exists, don't run it again.
if (file.exists("sim_mRNA800_TE40_b40_i40_xtail.csv")==F) {
sim_mRNA800_TE40_b40_i40_xtail <- injection2xtail(sim_mRNA800_TE40_b40_i40, outName = "sim_mRNA800_TE40_b40_i40_xtail.csv")
} else {
  sim_mRNA800_TE40_b40_i40_xtail <- read.csv("sim_mRNA800_TE40_b40_i40_xtail.csv")
}
if (file.exists("sim_mRNA4000_TE40_b40_i40_xtail.csv")==F) {
sim_mRNA4000_TE40_b40_i40_xtail <- injection2xtail(sim_mRNA4000_TE40_b40_i40, outName = "sim_mRNA4000_TE40_b40_i40_xtail.csv")
} else {
  sim_mRNA4000_TE40_b40_i40_xtail <- read.csv("sim_mRNA4000_TE40_b40_i40_xtail.csv")
}

sim_mRNA800_TE40_b40_i40_xtail <- sim_mRNA800_TE40_b40_i40_xtail %>% mutate(pvalue=pvalue_final)
sim_mRNA800_TE40_b40_i40_xtail_PR <- precisionAndRecall(sim_mRNA800_TE40_b40_i40_xtail, theContrast = "diffTE", TP_terms=c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim_mRNA800_TE40_b40_i40_xtail_PR_curve <- PR_curve(sim_mRNA800_TE40_b40_i40_xtail, theContrast = "diffTE", TP_terms=c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim_mRNA800_TE40_b40_i40_xtail_roc <- res2pROC(sim_mRNA800_TE40_b40_i40_xtail, contrast="diffTE")

sim_mRNA4000_TE40_b40_i40_xtail <- sim_mRNA4000_TE40_b40_i40_xtail %>% mutate(pvalue=pvalue_final)
sim_mRNA4000_TE40_b40_i40_xtail_PR <- precisionAndRecall(sim_mRNA4000_TE40_b40_i40_xtail, theContrast = "diffTE", TP_terms=c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim_mRNA4000_TE40_b40_i40_xtail_PR_curve <- PR_curve(sim_mRNA4000_TE40_b40_i40_xtail, theContrast = "diffTE", TP_terms=c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim_mRNA4000_TE40_b40_i40_xtail_roc <- res2pROC(sim_mRNA4000_TE40_b40_i40_xtail, contrast="diffTE")

#Xtail is slow. Very slow. If the file exists, don't run it again.
if (file.exists("sim2_mRNA800_TE40_b40_i40_xtail.csv")==F) {
sim2_mRNA800_TE40_b40_i40_xtail <- injection2xtail(sim2_mRNA800_TE40_b40_i40, outName = "sim2_mRNA800_TE40_b40_i40_xtail.csv")
} else {
  sim2_mRNA800_TE40_b40_i40_xtail <- read.csv("sim2_mRNA800_TE40_b40_i40_xtail.csv")
}
if (file.exists("sim2_mRNA4000_TE40_b40_i40_xtail.csv")==F) {
sim2_mRNA4000_TE40_b40_i40_xtail <- injection2xtail(sim2_mRNA4000_TE40_b40_i40, outName = "sim2_mRNA4000_TE40_b40_i40_xtail.csv")
} else {
  sim2_mRNA4000_TE40_b40_i40_xtail <- read.csv("sim2_mRNA4000_TE40_b40_i40_xtail.csv")
}

sim2_mRNA800_TE40_b40_i40_xtail <- sim2_mRNA800_TE40_b40_i40_xtail %>% mutate(pvalue=pvalue_final)
sim2_mRNA800_TE40_b40_i40_xtail_PR <- precisionAndRecall(sim2_mRNA800_TE40_b40_i40_xtail, theContrast = "diffTE", TP_terms=c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim2_mRNA800_TE40_b40_i40_xtail_PR_curve <- PR_curve(sim2_mRNA800_TE40_b40_i40_xtail, theContrast = "diffTE", TP_terms=c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim2_mRNA800_TE40_b40_i40_xtail_roc <- res2pROC(sim2_mRNA800_TE40_b40_i40_xtail, contrast="diffTE")

sim2_mRNA4000_TE40_b40_i40_xtail <- sim2_mRNA4000_TE40_b40_i40_xtail %>% mutate(pvalue=pvalue_final)
sim2_mRNA4000_TE40_b40_i40_xtail_PR <- precisionAndRecall(sim2_mRNA4000_TE40_b40_i40_xtail, theContrast = "diffTE", TP_terms=c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim2_mRNA4000_TE40_b40_i40_xtail_PR_curve <- PR_curve(sim2_mRNA4000_TE40_b40_i40_xtail, theContrast = "diffTE", TP_terms=c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim2_mRNA4000_TE40_b40_i40_xtail_roc <- res2pROC(sim2_mRNA4000_TE40_b40_i40_xtail, contrast="diffTE")

```

```{r anota2seq}
injection2anota2seq <- function(injection_list) {
    AF <- injection_list$AF
  RP <- injection_list$RP
  AF_TruePos <- AF %>% dplyr::select(Geneid, condition, TP, FCs) %>%
    filter((TP=="down" & condition=="n1") | (TP=="up" & condition=="n2") | (TP=="up_buffered" & condition=="n2") | (TP=="down_buffered" & condition=="n1") | (TP=="up_intensified" & condition=="n2") | (TP=="down_intensified" & condition=="n1")) %>% distinct()
  RP_TruePos <- RP %>% dplyr::select(Geneid, condition, TP, FCs) %>% filter((TP=="diffTE_down" & condition=="n1") | (TP=="diffTE_up" & condition=="n2") | (TP=="up_buffered" & condition=="n2") | (TP=="down_buffered" & condition=="n1") | (TP=="up_intensified" & condition=="n2") | (TP=="down_intensified" & condition=="n1")  ) %>% distinct()
  #return(list(AF_TruePos, RP_TruePos))
  AF_simulation <- AF %>% dplyr::select(Geneid, sample, new_counts) %>%
    spread(key="sample",value="new_counts") %>% column_to_rownames("Geneid")
  RP_simulation <- RP %>% dplyr::select(Geneid, sample, new_counts) %>%
    spread(key="sample",value="new_counts") %>% column_to_rownames("Geneid")
   AF_TruePos <- AF %>% dplyr::select(Geneid, condition, TP, FCs) %>%
    filter((TP=="down" & condition=="n1") | (TP=="up" & condition=="n2") | (TP=="up_buffered" & condition=="n2") | (TP=="down_buffered" & condition=="n1") | (TP=="up_intensified" & condition=="n2") | (TP=="down_intensified" & condition=="n1")) %>% distinct()
  RP_TruePos <- RP %>% dplyr::select(Geneid, condition, TP, FCs) %>% filter((TP=="diffTE_down" & condition=="n1") | (TP=="diffTE_up" & condition=="n2") | (TP=="up_buffered" & condition=="n2") | (TP=="down_buffered" & condition=="n1") | (TP=="up_intensified" & condition=="n2") | (TP=="down_intensified" & condition=="n1")  ) %>% distinct()
  #return(list(AF_TruePos, RP_TruePos))
  AF_simulation <- AF %>% dplyr::select(Geneid, sample, new_counts) %>%
    spread(key="sample",value="new_counts") %>% column_to_rownames("Geneid")
  RP_simulation <- RP %>% dplyr::select(Geneid, sample, new_counts) %>%
    spread(key="sample",value="new_counts") %>% column_to_rownames("Geneid")
  phenoVec <- str_extract(colnames(AF_simulation), regex("n."))
  anotaDataSet <- anota2seqDataSetFromMatrix(
    dataP=RP_simulation,
    dataT = AF_simulation,
    phenoVec = phenoVec,
    dataType="RNAseq",
    normalize= T, transformation = "TMM-log2"
  )
  #Set more realistic thresholds here. Bigger slopes allowed, smaller padj
  anotaDataSet <- anota2seqRun(anotaDataSet, thresholds=list(
    "minSlopeTranslation"=-5,
    "maxSlopeTranslation"=5,
    "minSlopeBuffering"=-5,
    "maxSlopeBuffering"=5,
    "maxPAdj"=0.1
  ))
  return(anotaDataSet)
}

sim_mRNA200_TE10_b0_i0_anota <- injection2anota2seq(sim_mRNA200_TE10_b0_i0)
sim_mRNA800_TE40_b40_i40_anota <- injection2anota2seq(sim_mRNA800_TE40_b40_i40)
sim_mRNA4000_TE40_b40_i40_anota <- injection2anota2seq(sim_mRNA4000_TE40_b40_i40)
sim_mRNA4000_TE10_b0_i0_anota <- injection2anota2seq(sim_mRNA4000_TE10_b0_i0)
sim2_mRNA800_TE40_b40_i40_anota <- injection2anota2seq(sim2_mRNA800_TE40_b40_i40)
sim2_mRNA4000_TE40_b40_i40_anota <- injection2anota2seq(sim2_mRNA4000_TE40_b40_i40)


anota2res2df <- function(anota2res, contrast="diffTE", injection_list) {
  #pull out the table from anota2 results
  RP <- injection_list$RP
  RP_TruePos <- RP %>% dplyr::select(Geneid, condition, TP, FCs) %>%   filter((TP=="diffTE_down" & condition=="n1") | (TP=="diffTE_up" & condition=="n2") | (TP=="up_buffered" & condition=="n2") | (TP=="down_buffered" & condition=="n1") | (TP=="up_intensified" & condition=="n2") | (TP=="down_intensified" & condition=="n1")  ) %>% distinct()
  res_df <- anota2seqGetOutput(anota2res, output="singleDf", selContrast = 1)
  res_df <- data.frame(res_df, contrast="diffTE")
  res_df <- res_df %>% mutate(Geneid=identifier, padj=translation.apvRvmPAdj, obs_lFC=translation.apvEff) %>% left_join(RP_TruePos, by="Geneid") %>% mutate(exp_lFC=log2(FCs))
}

sim_mRNA4000_TE40_b40_i40_anota_res <- anota2res2df(anota2res=sim_mRNA4000_TE40_b40_i40_anota, injection_list=sim_mRNA4000_TE40_b40_i40)
sim_mRNA4000_TE40_b40_i40_anota_PR <- precisionAndRecall(sim_mRNA4000_TE40_b40_i40_anota_res, theContrast = "diffTE", TP_terms = c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim_mRNA4000_TE40_b40_i40_anota_PR_curve <- PR_curve(sim_mRNA4000_TE40_b40_i40_anota_res, theContrast = "diffTE", TP_terms = c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim_mRNA4000_TE40_b40_i40_anota_roc <- res2pROC(sim_mRNA4000_TE40_b40_i40_anota_res, contrastOfInterest = "diffTE") 

sim_mRNA800_TE40_b40_i40_anota_res <- anota2res2df(anota2res=sim_mRNA800_TE40_b40_i40_anota, injection_list=sim_mRNA800_TE40_b40_i40)
sim_mRNA800_TE40_b40_i40_anota_PR <- precisionAndRecall(sim_mRNA800_TE40_b40_i40_anota_res, theContrast = "diffTE", TP_terms = c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim_mRNA800_TE40_b40_i40_anota_PR_curve <- PR_curve(sim_mRNA800_TE40_b40_i40_anota_res, theContrast = "diffTE", TP_terms = c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim_mRNA800_TE40_b40_i40_anota_roc <- res2pROC(sim_mRNA800_TE40_b40_i40_anota_res, contrastOfInterest = "diffTE") 

sim2_mRNA4000_TE40_b40_i40_anota_res <- anota2res2df(anota2res=sim2_mRNA4000_TE40_b40_i40_anota, injection_list=sim2_mRNA4000_TE40_b40_i40)
sim2_mRNA4000_TE40_b40_i40_anota_PR <- precisionAndRecall(sim2_mRNA4000_TE40_b40_i40_anota_res, theContrast = "diffTE", TP_terms = c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim2_mRNA4000_TE40_b40_i40_anota_PR_curve <- PR_curve(sim2_mRNA4000_TE40_b40_i40_anota_res, theContrast = "diffTE", TP_terms = c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim2_mRNA4000_TE40_b40_i40_anota_roc <- res2pROC(sim2_mRNA4000_TE40_b40_i40_anota_res, contrastOfInterest = "diffTE") 

sim2_mRNA800_TE40_b40_i40_anota_res <- anota2res2df(anota2res=sim2_mRNA800_TE40_b40_i40_anota, injection_list=sim2_mRNA800_TE40_b40_i40)
sim2_mRNA800_TE40_b40_i40_anota_PR <- precisionAndRecall(sim2_mRNA800_TE40_b40_i40_anota_res, theContrast = "diffTE", TP_terms = c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim2_mRNA800_TE40_b40_i40_anota_PR_curve <- PR_curve(sim2_mRNA800_TE40_b40_i40_anota_res, theContrast = "diffTE", TP_terms = c("diffTE_down","diffTE_up","up_buffered","down_buffered","up_intensified","down_intensified"))
sim2_mRNA800_TE40_b40_i40_anota_roc <- res2pROC(sim2_mRNA800_TE40_b40_i40_anota_res, contrastOfInterest = "diffTE") 
```

```{r compareROCsAndConfusions}

mRNA4000_40TE_40b_40i_ROC_list=list(
  "DESeq2"=sim_mRNA4000_TE40_b40_i40_roc,
  "xtail"=sim_mRNA4000_TE40_b40_i40_xtail_roc,
  "anota"=sim_mRNA4000_TE40_b40_i40_anota_roc)
mRNA800_40TE_40b_40i_ROC_list=list(
  "DESeq2"=sim_mRNA800_TE40_b40_i40_roc,
  "xtail"=sim_mRNA800_TE40_b40_i40_xtail_roc,
  "anota"=sim_mRNA800_TE40_b40_i40_anota_roc)

mRNA4000_40TE_40b_40i_PR_list=list(
  "DESeq2"=sim_mRNA4000_TE40_b40_i40_PR_curve,
  "xtail"=sim_mRNA4000_TE40_b40_i40_xtail_PR_curve,
  "anota2seq"=sim_mRNA4000_TE40_b40_i40_anota_PR_curve
)
mRNA800_40TE_40b_40i_PR_list=list(
  "DESeq2"=sim_mRNA800_TE40_b40_i40_PR_curve,
  "xtail"=sim_mRNA800_TE40_b40_i40_xtail_PR_curve,
  "anota2seq"=sim_mRNA800_TE40_b40_i40_anota_PR_curve
)

sim2_mRNA4000_40TE_40b_40i_PR_list=list(
  "DESeq2"=sim2_mRNA4000_TE40_b40_i40_PR_curve,
  "xtail"=sim2_mRNA4000_TE40_b40_i40_xtail_PR_curve,
  "anota2seq"=sim2_mRNA4000_TE40_b40_i40_anota_PR_curve
)
sim2_mRNA800_40TE_40b_40i_PR_list=list(
  "DESeq2"=sim2_mRNA800_TE40_b40_i40_PR_curve,
  "xtail"=sim2_mRNA800_TE40_b40_i40_xtail_PR_curve,
  "anota2seq"=sim2_mRNA800_TE40_b40_i40_anota_PR_curve
)

compare_ROCs <- function(ROClist) {
  roc_dfs <- lapply(ROClist, function(x){
    ROC <- x
    sensitivities <- ROC$sensitivities
    specificities <- ROC$specificities
    auc=ROC$auc
    df <- data.frame(sensitivity=sensitivities, specificity=specificities)
  }) %>% bind_rows(.id="ROC_name")
}

compare_PR_curves <- function(PRlist) {
  PR_df <- lapply(PRlist, function(x){
   PR <- x$data
  df <- data.frame(precision=PR$precision,recall=PR$recall)
  }) %>% bind_rows(.id="PR_name")
  PR_comp_plot <- ggplot(PR_df, aes(x=recall,y=precision, color=PR_name))+JF_theme+geom_line()
}

mRNA4000_40TE_40b_40i_ROC_df <- compare_ROCs(mRNA4000_40TE_40b_40i_ROC_list)
mRNA4000_40TE_40b_40i_PR_comp_plot <- compare_PR_curves(mRNA4000_40TE_40b_40i_PR_list)+ggtitle("sim1, mRNA4000, \nTE 40, buffered 40, intensified 40 \nP vs R")
mRNA4000_40TE_40b_40i_ROC_plot <- ggplot(mRNA4000_40TE_40b_40i_ROC_df, aes(x=1-specificity, y=sensitivity, color=ROC_name))+JF_theme+geom_line()+xlab("False Positive Rate")+ylab("True Positive Rate")+geom_abline(slope=1, intercept=0)
mRNA800_40TE_40b_40i_ROC_df <- compare_ROCs(mRNA800_40TE_40b_40i_ROC_list)
mRNA800_40TE_40b_40i_PR_comp_plot <- compare_PR_curves(mRNA800_40TE_40b_40i_PR_list)+ggtitle("sim1, mRNA4000, \nTE 40, buffered 40, intensified 40 \nP vs R")
mRNA800_40TE_40b_40i_ROC_plot <- ggplot(mRNA800_40TE_40b_40i_ROC_df, aes(x=1-specificity, y=sensitivity, color=ROC_name))+JF_theme+geom_line()+xlab("False Positive Rate")+ylab("True Positive Rate")+geom_abline(slope=1, intercept=0)

sim2_mRNA800_40TE_40b_40i_PR_comp_plot <- compare_PR_curves(sim2_mRNA800_40TE_40b_40i_PR_list)+ggtitle("sim2, mRNA800, \nTE 40, buffered 40, intensified 40 \nP vs R")
sim2_mRNA4000_40TE_40b_40i_PR_comp_plot <- compare_PR_curves(sim2_mRNA4000_40TE_40b_40i_PR_list)+ggtitle("sim2, mRNA4000, \nTE 40, buffered 40, intensified 40 \nP vs R")

```

```{r simulation_results_fig}
sim_results_fig <- (sim_mRNA4000_TE40_b40_i40_injection_MAplots + sim2_mRNA4000_TE40_b40_i40_injection_MAplots) / ((mRNA4000_40TE_40b_40i_PR_comp_plot | sim2_mRNA4000_40TE_40b_40i_PR_comp_plot)+plot_layout(guides="collect")) /
(sim_mRNA800_TE40_b40_i40_injection_MAplots + sim2_mRNA800_TE40_b40_i40_injection_MAplots) / ((mRNA800_40TE_40b_40i_PR_comp_plot | sim2_mRNA800_40TE_40b_40i_PR_comp_plot)+plot_layout(guides="collect")) +theme(text=element_text("ArialMT")) + plot_annotation(tag_levels = list(c("A","B","C","D","E","F","G","H"))) & theme(plot.tag=element_text(size=12, face="bold"))


pdfName=paste(outDir, "sim_res_fig.pdf", sep="/")
pdf(file=pdfName, width=6.5, height=9)
plot(sim_results_fig)
dev.off()

```


```{r enrichGO}
rna_universe <- subtypeComparison_res_df %>%
 dplyr::select(Geneid) %>% distinct()
rna_universe <- rna_universe$Geneid


run_enrichGO <- function(genesOfInterest, gene_uni=rna_universe, min_count=15) {
  GO_res <- enrichGO(gene=genesOfInterest, universe=gene_uni,  OrgDb = "org.Mm.eg.db", keyType = "ENSEMBL", ont = "ALL", pAdjustMethod = "BH", readable = T, pool = F,) %>% data.frame() %>% filter(Count > min_count) %>% separate(GeneRatio, into = c("Significant", "denom")) %>%  mutate(GeneRatio = as.numeric(Significant) / as.numeric(denom)) %>% separate(BgRatio, into=c("notSignificant","denom")) %>% mutate(BgRatio=as.numeric(notSignificant)/as.numeric(denom))
}

run_enrichGO_raw_output <- function(genesOfInterest, gene_uni=rna_universe, min_count=15) {
  #same as run_enrichGO, but does not convert the output to a dataframe. 
  GO_res <- enrichGO(gene=genesOfInterest, universe=gene_uni,  OrgDb = "org.Mm.eg.db", keyType = "ENSEMBL", ont = "ALL", pAdjustMethod = "BH", readable = T, pool = F,)
}

SCPNvsCPN_TE_higher <- (subtypeComparison_res_df %>% filter(diffTE_sig == "higher"))$Geneid
SCPNvsCPN_TE_lower <- (subtypeComparison_res_df %>% filter(diffTE_sig == "lower"))$Geneid
diffTE_genes <- union(SCPNvsCPN_TE_higher, SCPNvsCPN_TE_lower)
CPN_TE_high <- (subtypeComparison_res_df %>% filter(TE_CPN_sig == "higher"))$Geneid
SCPN_TE_high <- (subtypeComparison_res_df %>% filter(TE_SCPN_sig == "higher"))$Geneid
CPN_TE_low <- (subtypeComparison_res_df %>% filter(TE_CPN_sig == "lower"))$Geneid
SCPN_TE_low <- (subtypeComparison_res_df %>% filter(TE_SCPN_sig == "lower"))$Geneid
CPN_TE_notSig <- (subtypeComparison_res_df %>% filter(contrast=="TE_CPN" & (TE_CPN_sig %in% c("higher","lower"))==F))$Geneid
SCPN_TE_notSig <- (subtypeComparison_res_df %>% filter(contrast=="TE_SCPN" & (TE_SCPN_sig %in% c("higher","lower"))==F))$Geneid


SCPN_mRNA <- (subtypeComparison_res_df %>% filter(mRNA_sig == "higher"))$Geneid
CPN_mRNA <- (subtypeComparison_res_df %>% filter(mRNA_sig == "lower"))$Geneid
SCPN_RPF <- (subtypeComparison_res_df %>% filter(RP_sig == "higher"))$Geneid
CPN_RPF <- (subtypeComparison_res_df %>% filter(RP_sig == "lower"))$Geneid

CPN_SCPN_TE_upset_list <- list("CPN_high"=CPN_TE_high, "CPN_low"=CPN_TE_low, "SCPN_high"=SCPN_TE_high, "SCPN_low"=SCPN_TE_low)
CPN_SCPN_TE_upset <- upset(fromList(CPN_SCPN_TE_upset_list), order.by="freq", text.scale=2.67)
CPN_SCPN_TE_upset 

SCPN_CPN_AF_RP_upset_list <- list("SCPN_mRNA"=SCPN_mRNA, "CPN_mRNA"=CPN_mRNA, "SCPN_RPF"=SCPN_RPF, "CPN_RPF"=CPN_RPF)
SCPN_CPN_AF_RP_upset <- upset(fromList(SCPN_CPN_AF_RP_upset_list), order.by="freq")
SCPN_CPN_AF_RP_upset 

diffTE_upset_list <- list("SCPN_TE"=SCPNvsCPN_TE_higher, "CPN_TE"=SCPNvsCPN_TE_lower, "SCPN_mRNA"=SCPN_mRNA, "CPN_mRNA"=CPN_mRNA, "SCPN_RPF"=SCPN_RPF, "CPN_RPF"=CPN_RPF )
diffTE_upset <- upset(fromList(diffTE_upset_list), keep.order = T, nsets=length(diffTE_upset_list), text.scale = 2)
diffTE_upset 
diffTE_focus_list <- list(list("SCPN_TE","SCPN_mRNA"),list("SCPN_TE","SCPN_RPF"),
                          list("SCPN_TE","CPN_mRNA"),list("SCPN_TE","CPN_RPF"),
                          list("CPN_TE","CPN_mRNA"),list("CPN_TE","CPN_RPF"),
                          list("CPN_TE","SCPN_mRNA"),list("CPN_TE","SCPN_RPF"),
                          list("CPN_TE","SCPN_mRNA","CPN_RPF"), list("SCPN_TE","SCPN_RPF","SCPN_mRNA"),list("CPN_TE","CPN_RPF","CPN_mRNA"), list("SCPN_TE","CPN_mRNA","SCPN_RPF"))
diffTE_upset_TE_focus <- upset(fromList(diffTE_upset_list), order.by = "freq",intersections=diffTE_focus_list, text.scale=2)
diffTE_upset_TE_focus 



subtypeComp_GO_list <- list("SCPN_vsCPN_TEhigher"=SCPNvsCPN_TE_higher,"SCPN_vsCPN_TElower"=SCPNvsCPN_TE_lower,
                            "CPN_TE_high"=CPN_TE_high, "SCPN_TE_high"=SCPN_TE_high,
                            "CPN_TE_low"=CPN_TE_low, "SCPN_TE_low"=SCPN_TE_low,
                            "diffTE"=diffTE_genes,
                            "SCPN_mRNA"=SCPN_mRNA, "CPN_mRNA"=CPN_mRNA, 
                            "SCPN_RPF"=SCPN_RPF, "CPN_RPF"=CPN_RPF)

ego_subtypes <- lapply(subtypeComp_GO_list, run_enrichGO)
ego_subtypes_raw <- lapply(subtypeComp_GO_list, run_enrichGO_raw_output)

names(ego_subtypes ) <- names(ego_subtypes)

plot_ego <- function(ego, numToPlot=5, onto=c("BP","CC","MF"), title="") {
  ego_filt <- ego %>% filter(ONTOLOGY %in% onto) %>% group_by(ONTOLOGY) %>% slice_min(n=numToPlot, order_by=qvalue)
  ego_filt$Description <- factor(ego_filt$Description, levels=ego_filt$Description)
  ego_plot <- ggplot(ego_filt, aes(x=GeneRatio,y=Description, color=-log10(qvalue),size=Count))+JF_theme+geom_point()+facet_wrap(~ONTOLOGY, nrow=3, scales="free_y")+ggtitle(title)
}

ego_subtypes_plots <- lapply(ego_subtypes, plot_ego)
ego_subtypes_plots_BP_CC <- lapply(ego_subtypes, plot_ego, onto=c("BP","CC"))

ego_subtypes_barplots <- lapply(ego_subtypes_raw, barplot, showCategory=5, font.size=7, label_format=16)

TE_CPN_higher_barplot <- ego_subtypes_barplots$CPN_TE_high + ggtitle("TE CPN, skew higher")+JF_theme+theme(legend.position = "top")
TE_CPN_higher_barplot
TE_SCPN_higher_barplot <- ego_subtypes_barplots$SCPN_TE_high + ggtitle("TE SCPN, skew higher")+JF_theme+theme(legend.position = "top")
TE_SCPN_higher_barplot
TE_CPN_lower_barplot <- ego_subtypes_barplots$CPN_TE_low + ggtitle("TE CPN, skew lower")+JF_theme+theme(legend.position = "top")
TE_CPN_lower_barplot
TE_SCPN_lower_barplot <- ego_subtypes_barplots$SCPN_TE_low + ggtitle("TE SCPN, skew lower")+JF_theme+theme(legend.position = "top")
TE_SCPN_lower_barplot
###Tried treeplot, ran into dependency issues.


```

```{r subtypesFigures_layout}

###Need to manually put in the upset plots as these are not ggplots. 


subtypesComp_Figure_upsetSpacers <- as.ggplot(TPM_cor_pheatmap_simpler) / (RPvsAF_scatterplot | mRNA_MAplot | RPF_MAplot ) / (TE_MAplot + TE_CPNvsSCPN_scatterplot) /(plot_spacer()+plot_spacer())+theme(text=element_text("ArialMT")) + plot_annotation(tag_levels = list(c("A","B","C","D","E","F","G","H"))) & theme(plot.tag=element_text(size=12, face="bold"))+theme(legend.key.size=unit(0.5,"line"))




pdfName=paste(outDir, "SubtypeComparison_fig.pdf", sep="/")
pdf(file=pdfName, width=6.5, height=9)
plot(subtypesComp_Figure_upsetSpacers)
dev.off()



withinSubtypeTE <- (plot_spacer()| TE_CPNvsSCPN_scatterplot_TE_CPN_sig_colored | TE_CPNvsSCPN_scatterplot_TE_SCPN_sig_colored) /(TE_CPN_higher_barplot + TE_CPN_lower_barplot) / (TE_SCPN_higher_barplot + TE_SCPN_lower_barplot) / (plot_spacer() | plot_spacer()) /(plot_spacer() | plot_spacer()) + plot_annotation(tag_levels = list(c("A","B","C","D","E","F","G","H","I","J","K"))) & theme(plot.tag=element_text(size=12, face="bold"))+theme(legend.key.size=unit(0.5,"line"))

pdfName=paste(outDir, "withinSubtypeTE_fig.pdf", sep="/")
pdf(file=pdfName, width=6.5, height=9)
plot(withinSubtypeTE)
dev.off()

```

```{r motif_analyis}
#Take the longest UTR out of the ones listed, so there is just one sequence for each Geneid
#Also filter for Length > 10; I had this in the old code from 2021 might be a min length filter for STREME
five_utr_longest.bed <- read.table("five_prime_utr.bed", header=T) %>% mutate(Length=end-start) %>% group_by(Geneid) %>% slice_max(n=1, order_by=Length) %>% filter(Length > 10)
three_utr_longest.bed <- read.table("three_prime_utr.bed", header=T) %>% mutate(Length=end-start) %>% group_by(Geneid) %>% slice_max(n=1, order_by=Length) %>% filter(Length > 10)

CPNandSCPN_TE_high <- intersect(SCPN_TE_high, CPN_TE_high)
CPNandSCPN_TE_low <- intersect(SCPN_TE_low, CPN_TE_low)

CPNandSCPN_TE_high_longest_5UTR.bed <- five_utr_longest.bed %>% filter(Geneid %in% CPNandSCPN_TE_high)
CPNandSCPN_TE_low_longest_5UTR.bed <- five_utr_longest.bed %>% filter(Geneid %in% CPNandSCPN_TE_low )
CPNandSCPN_TE_high_longest_3UTR.bed <- three_utr_longest.bed %>% filter(Geneid %in% CPNandSCPN_TE_high )
CPNandSCPN_TE_low_longest_3UTR.bed <- three_utr_longest.bed %>% filter(Geneid %in% CPNandSCPN_TE_low )

writeBedFile <- function(bedFile) {
  bedOut<- data.frame(chr=bedFile$chr, start=bedFile$start, end=bedFile$end, name=bedFile$Geneid, score=bedFile$score, strand=bedFile$strand)
  outname=deparse(substitute(bedFile))
  write.table(bedOut, file=outname, row.names = F, col.names = F, quote=F, sep="\t")
}

longest_5UTR.bed <- five_utr_longest.bed
longest_3UTR.bed <- three_utr_longest.bed
writeBedFile(longest_5UTR.bed)
writeBedFile(longest_3UTR.bed)

writeBedFile(CPNandSCPN_TE_high_longest_5UTR.bed)
writeBedFile(CPNandSCPN_TE_low_longest_5UTR.bed)
writeBedFile(CPNandSCPN_TE_high_longest_3UTR.bed)
writeBedFile(CPNandSCPN_TE_low_longest_3UTR.bed)

five_utr_bed <- read.table("five_prime_utr.bed", header=T) %>% mutate(Length=end-start) %>% group_by(Geneid)  %>% filter(Length > 10)
three_utr_bed <- read.table("three_prime_utr.bed", header=T) %>% mutate(Length=end-start) %>% group_by(Geneid) %>% filter(Length > 10)


CPNandSCPN_TE_high_5UTR.bed <- five_utr_bed %>% filter(Geneid %in% CPNandSCPN_TE_high)
CPNandSCPN_TE_low_5UTR.bed <- five_utr_bed %>% filter(Geneid %in% CPNandSCPN_TE_low )
CPNandSCPN_TE_high_3UTR.bed <- three_utr_bed %>% filter(Geneid %in% CPNandSCPN_TE_high )
CPNandSCPN_TE_low_3UTR.bed <- three_utr_bed %>% filter(Geneid %in% CPNandSCPN_TE_low )

writeBedFile(CPNandSCPN_TE_high__5UTR.bed)
writeBedFile(CPNandSCPN_TE_low__5UTR.bed)
writeBedFile(CPNandSCPN_TE_high__3UTR.bed)
writeBedFile(CPNandSCPN_TE_low__3UTR.bed)
```

#run transite if UTR fastqs are extracted
```{r transite_analysis}

five_utr.fasta <- "extracted_utrs/5UTR_longest.fasta"
three_utr.fasta <- "extracted_utrs/3UTR_longest.fasta"

if (file.exists(five_utr.fasta)==T & file.exists(three_utr.fasta)==T) {
  five_utr_seqs <- readDNAStringSet(five_utr.fasta)
  three_utr_seqs <- readDNAStringSet(three_utr.fasta)
  five_utr_seq_df <- data.frame(seq=five_utr_seqs, Geneid=str_extract(names(five_utr_seqs), regex("ENSMUSG[0-9]+")), seqtype="5'-UTR") %>% mutate(seq=gsub("T","U", seq))
  three_utr_seq_df <- data.frame(seq=three_utr_seqs, Geneid=str_extract(names(three_utr_seqs), regex("ENSMUSG[0-9]+")), seqtype="3'-UTR") %>% mutate(seq=gsub("T","U", seq))
  
res2transite_input <- function(res, seq_df) {
  TS_input <- res %>% left_join(seq_df, by="Geneid") %>% mutate(refSeq=Geneid, value=log2FoldChange) %>% dplyr::select(refSeq, value, seqtype, seq) %>% filter(is.na(seq)==F & grepl("N",seq)==F) %>% arrange(desc(value))
}

TE_CPN_transite_input_5UTR <- res2transite_input(res=res_TE_CPN_df, seq_df = five_utr_seq_df)
TE_CPN_transite_5UTR <- run_matrix_spma(sorted_transcript_sequences=TE_CPN_transite_input_5UTR$seq, sorted_transcript_values=TE_CPN_transite_input_5UTR$value, transcript_values_label = "T.E. CPN", cache=F, n_cores=8 )
TE_CPN_transite_input_3UTR <- res2transite_input(res=res_TE_CPN_df, seq_df = three_utr_seq_df)
TE_CPN_transite_3UTR <- run_matrix_spma(sorted_transcript_sequences=TE_CPN_transite_input_3UTR$seq, sorted_transcript_values=TE_CPN_transite_input_3UTR$value, transcript_values_label = "T.E. CPN", cache=F, n_cores=8 )
TE_SCPN_transite_input_3UTR <- res2transite_input(res=res_TE_SCPN_df, seq_df = three_utr_seq_df)
TE_SCPN_transite_3UTR <- run_matrix_spma(sorted_transcript_sequences=TE_SCPN_transite_input_3UTR$seq, sorted_transcript_values=TE_SCPN_transite_input_3UTR$value, transcript_values_label = "T.E. SCPN", cache=F, n_cores=8 )


  # TE_CPN_transite_input_5UTR <- res_TE_CPN_df %>% left_join(five_utr_seq_df, by="Geneid") %>% mutate(refSeq=Geneid, value=log2FoldChange) %>% dplyr::select(refSeq, value, seqtype, seq)
  # 
  
  
  
  
} else {
  warning("run extractUTRsFromBed in order to run transite")
}


```



```{r uORFs}
CPN_ORFs_file="CPN_RiboCode/RiboCode_ORFs_collapsed.txt"
SCPN_ORFs_file="SCPN_RiboCode/RiboCode_ORFs_collapsed.txt"

CPN_ORFs_res_collapsed <- read.table(file=CPN_ORFs_file, sep="\t", header=T) %>% mutate(subtype="CPN")
SCPN_ORFs_res_collapsed <- read.table(file=SCPN_ORFs_file, sep="\t", header=T) %>% mutate(subtype="SCPN")
ORF_res <- bind_rows(CPN_ORFs_res_collapsed, SCPN_ORFs_res_collapsed)

#label novel uORFs mapping to pseudogenes
novelORF_in_pseudogene <- ORF_res$ORF_type=="novel" & (ORF_res$transcript_type %in% c("polymorphic_pseudogene","processed_pseudogene","pseudogene","transcribed_processed_pseudogene","transcribed_unitary_pseudogene","transcribed_unprocessed_pseudogene","translated_processed_pseudogene","unprocessed_pseudogene"))==T
ORF_res[novelORF_in_pseudogene==T,]$ORF_type <- "pseudogene" 

#plot the counts of ORF types by subtype
ORF_type_counts <- ORF_res %>% group_by(subtype, ORF_type) %>% summarize(ORF_count=n())

uORF_res <- filter(ORF_res, ORF_type=="uORF" | ORF_type=="Overlap_uORF")

ORF_types_plot <- ggplot(ORF_type_counts, aes(x=ORF_type, y=ORF_count, fill=ORF_type, label=ORF_count))+JF_theme+geom_col()+geom_text()+facet_wrap(~subtype)+theme(axis.text.x = element_text(angle = 45, hjust=1))+theme(legend.position = "none")

uORF_shared <- intersect(filter(uORF_res, ORF_type=="uORF" & subtype=="CPN")$ORF_ID, filter(uORF_res, ORF_type=="uORF" & subtype=="SCPN")$ORF_ID)
Overlap_uORF_shared <-  intersect(filter(uORF_res, ORF_type=="Overlap_uORF" & subtype=="CPN")$ORF_ID, filter(uORF_res, ORF_type=="Overlap_uORF" & subtype=="SCPN")$ORF_ID)
uORF_shared_Geneids <- filter(uORF_res, ORF_ID %in% uORF_shared)
uORF_shared_Geneids <- unique(uORF_shared_Geneids$gene_id)
Overlap_uORF_shared_Geneids <- filter(uORF_res, ORF_ID %in% Overlap_uORF_shared)
Overlap_uORF_shared_Geneids <- unique(Overlap_uORF_shared_Geneids$gene_id)


uORF_res_shared <- uORF_res %>% mutate(shared_status=case_when(
  ORF_type == "uORF" & ORF_ID %in% uORF_shared ~ "shared",
  ORF_type == "Overlap_uORF" & ORF_ID %in% Overlap_uORF_shared ~ "shared",
  ORF_type == "uORF" & (ORF_ID %in% uORF_shared)==F & subtype=="CPN" ~ "CPN only",
  ORF_type == "uORF" & (ORF_ID %in% uORF_shared)==F & subtype=="SCPN" ~ "SCPN only",
  ORF_type == "Overlap_uORF" & (ORF_ID %in% Overlap_uORF_shared)==F & subtype=="CPN" ~ "CPN only",
  ORF_type == "Overlap_uORF" & (ORF_ID %in% uORF_shared)==F & subtype=="SCPN" ~ "SCPN only"
  )) %>% group_by(ORF_type, shared_status) %>% summarize(count=n())
#divide shared counts by 2 to avoid double count. 
uORF_res_shared[uORF_res_shared$shared_status=="shared",]$count <- uORF_res_shared[uORF_res_shared$shared_status=="shared",]$count/2

uORF_shared_col_plot <- ggplot(uORF_res_shared, aes(x=ORF_type, y=count, fill=shared_status))+geom_col(position="stack")+JF_theme+theme(axis.text.x = element_text(angle = 45, hjust=1))
###Overlap between CPN and SCPN is a bit lower than I expected. Tried re-runing RiboCode with a more stringent pval (0.01 vs 0.05)
###Still low. I wonder a bit if this might be because a decent % of uORFs are in genes that are differentially expressed between subtypes.

uORF_ID_list <- list("CPN_uORF"=filter(ORF_res, ORF_type=="uORF" & subtype=="CPN")$ORF_ID,
                     "CPN_Overlap_uORF"=filter(ORF_res, ORF_type=="Overlap_uORF" & subtype=="CPN")$ORF_ID,
                     "SCPN_uORF"=filter(ORF_res, ORF_type=="uORF" & subtype=="SCPN")$ORF_ID,
                     "SCPN_Overlap_uORF"=filter(ORF_res, ORF_type=="Overlap_uORF" & subtype=="SCPN")$ORF_ID)
uORF_ID_upset <- upset(fromList(uORF_ID_list),order.by = "freq")
uORF_ID_upset
###Hypothesis wrong. Even among genes that are not diff exp, only about 33% of uORFs are shared. 
uORF_ID_notdiffExp_list <- list("CPN_uORF"=filter(ORF_res, ORF_type=="uORF" & subtype=="CPN" & (gene_id %in% c(subtypeComp_GO_list$CPN_mRNA, subtypeComp_GO_list$SCPN_mRNA))==F)$ORF_ID,
                     "CPN_Overlap_uORF"=filter(ORF_res, ORF_type=="Overlap_uORF" & subtype=="CPN" & (gene_id %in% c(subtypeComp_GO_list$CPN_mRNA, subtypeComp_GO_list$SCPN_mRNA))==F)$ORF_ID,
                     "SCPN_uORF"=filter(ORF_res, ORF_type=="uORF" & subtype=="SCPN" & (gene_id %in% c(subtypeComp_GO_list$CPN_mRNA, subtypeComp_GO_list$SCPN_mRNA))==F)$ORF_ID,
                     "SCPN_Overlap_uORF"=filter(ORF_res, ORF_type=="Overlap_uORF" & subtype=="SCPN" & (gene_id %in% c(subtypeComp_GO_list$CPN_mRNA, subtypeComp_GO_list$SCPN_mRNA))==F)$ORF_ID)
uORF_ID_notdiffExp_upset <- upset(fromList(uORF_ID_notdiffExp_list),order.by = "freq")
uORF_ID_notdiffExp_upset



uORF_transcriptID_list <- list("CPN_uORF"=filter(ORF_res, ORF_type=="uORF" & subtype=="CPN")$transcript_id,
                     "CPN_Overlap_uORF"=filter(ORF_res, ORF_type=="Overlap_uORF" & subtype=="CPN")$transcript_id,
                     "SCPN_uORF"=filter(ORF_res, ORF_type=="uORF" & subtype=="SCPN")$transcript_id,
                     "SCPN_Overlap_uORF"=filter(ORF_res, ORF_type=="Overlap_uORF" & subtype=="SCPN")$transcript_id)
uORF_transcriptID_upset <- upset(fromList(uORF_transcriptID_list))
uORF_transcriptID_upset

uORF_geneID_list <- list("CPN_uORF"=filter(ORF_res, ORF_type=="uORF" & subtype=="CPN")$gene_id,
                     "CPN_Overlap_uORF"=filter(ORF_res, ORF_type=="Overlap_uORF" & subtype=="CPN")$gene_id,
                     "SCPN_uORF"=filter(ORF_res, ORF_type=="uORF" & subtype=="SCPN")$gene_id,
                     "SCPN_Overlap_uORF"=filter(ORF_res, ORF_type=="Overlap_uORF" & subtype=="SCPN")$gene_id)
uORF_geneID_upset <- upset(fromList(uORF_geneID_list))
uORF_geneID_upset

uORF_GO_list <- list("CPN_uORFs" = filter(ORF_res, (ORF_type=="uORF"| ORF_type=="Overlap_uORF" ) & subtype=="CPN")$gene_id,
                     "SCPN_uORFs"= filter(ORF_res, (ORF_type=="uORF"| ORF_type=="Overlap_uORF" ) & subtype=="SCPN")$gene_id,
                     "uORF_shared"=unique(filter(ORF_res, (ORF_ID %in% uORF_shared)==T | (ORF_ID %in% Overlap_uORF_shared)==T)$gene_id))
uORF_ego_raw <- lapply(uORF_GO_list, run_enrichGO_raw_output)

###See the strong enrichment for "synapse and structure" genes seen previously.
uORF_ego_barplot <- lapply(uORF_ego_raw, barplot, showCategory=5,font.size=7, label_format=24)
uORF_ego_barplot$CPN_uORFs <- uORF_ego_barplot$CPN_uORFs+theme(legend.position = "top")+ggtitle("top GO terms, CPN")+JF_theme
uORF_ego_barplot$SCPN_uORFs <- uORF_ego_barplot$SCPN_uORFs+theme(legend.position = "top")+ggtitle("top GO terms uORFs, SCPN")+JF_theme
uORF_ego_barplot$uORF_shared <- uORF_ego_barplot$uORF_shared+ggtitle("top GO terms CPN, SCPN shared uORFs")+JF_theme



subtypeComparison_res_df_plus_uORFs <- subtypeComparison_res_df %>%
  mutate(uORF_status=case_when(
    Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="uORF")$gene_id ~ "CPN_uORF",
    Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="Overlap_uORF")$gene_id ~ "CPN_Overlap_uORF",
    Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="uORF")$gene_id ~ "SCPN_uORF",
    Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="Overlap_uORF")$gene_id ~ "SCPN_Overlap_uORF"
  )) %>%
  mutate(uORF_status_intersect=case_when(
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="uORF")$gene_id==T) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="uORF")$gene_id==F) ~ "CPN_uORF",
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="uORF")$gene_id==F) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="uORF")$gene_id==T) ~ "SCPN_uORF",
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="uORF")$gene_id==T) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="uORF")$gene_id==T) ~ "SCPN&CPN_uORF",
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="Overlap_uORF")$gene_id==T) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="Overlap_uORF")$gene_id==F) ~ "CPN_Overlap_uORF",
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="Overlap_uORF")$gene_id==F) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="Overlap_uORF")$gene_id==T) ~ "SCPN_Overlap_uORF",
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="Overlap_uORF")$gene_id==T) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="Overlap_uORF")$gene_id==T) ~ "SCPN&CPN_Overlap_uORF",
))

res_TE_CPN_df <- res_TE_CPN_df %>% mutate(uORF_status=case_when(
  Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="uORF")$gene_id ~ "CPN_uORF",
  Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="Overlap_uORF")$gene_id ~ "CPN_Overlap_uORF",
  T ~ "no_uORF"
)) %>% mutate(uORF_shared=case_when(
  (Geneid %in% uORF_shared_Geneids)==T ~ "uORF",
  (Geneid %in% Overlap_uORF_shared_Geneids)==T ~ "Overlap_uORF",
  T ~ "no_shared_uORF"
))
res_TE_SCPN_df <- res_TE_SCPN_df %>% mutate(uORF_status=case_when(
  Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="uORF")$gene_id ~ "SCPN_uORF",
    Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="Overlap_uORF")$gene_id ~ "SCPN_Overlap_uORF",
  T ~ "no_uORF"
)) %>% mutate(uORF_shared=case_when(
  (Geneid %in% uORF_shared_Geneids)==T ~ "uORF",
  (Geneid %in% Overlap_uORF_shared_Geneids)==T ~ "Overlap_uORF",
  T ~ "no_shared_uORF"
))
res_diffTE_df <- res_diffTE_df %>% mutate(uORF_status=case_when(
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="uORF")$gene_id==T) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="uORF")$gene_id==F) ~ "CPN_uORF",
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="uORF")$gene_id==F) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="uORF")$gene_id==T) ~ "SCPN_uORF",
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="uORF")$gene_id==T) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="uORF")$gene_id==T) ~ "SCPN&CPN_uORF",
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="Overlap_uORF")$gene_id==T) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="Overlap_uORF")$gene_id==F) ~ "CPN_Overlap_uORF",
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="Overlap_uORF")$gene_id==F) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="Overlap_uORF")$gene_id==T) ~ "SCPN_Overlap_uORF",
  (Geneid %in% filter(uORF_res, subtype=="CPN" & ORF_type=="Overlap_uORF")$gene_id==T) & (Geneid %in% filter(uORF_res, subtype=="SCPN" & ORF_type=="Overlap_uORF")$gene_id==T) ~ "SCPN&CPN_Overlap_uORF",
))


TE_CPN_uORF_plot <- ggplot(res_TE_CPN_df, aes(x=log2FoldChange, color=uORF_status))+JF_theme+stat_ecdf(size=0.1)+facet_wrap(~contrast)+xlim(c(-2,2))+ylab("cumulative_fraction")+ggtitle("T.E. by uORF status, CPN")+xlab("T.E.")+theme(legend.position="top")+geom_hline(yintercept=0.5,size=0.2)+geom_vline(xintercept=0, size=0.2)
TE_CPN_uORF_plot

meanAndTest <- function(x, y) {
  theTest <- wilcox.test(x,y)
  mean_x <- mean(x)
  mean_y <- mean(y)
  returnList <- list("test"=theTest, "mean_x"=mean_x,"mean_y"=mean_y)
}

TE_CPN_noVsuORF.test <- meanAndTest(filter(res_TE_CPN_df, uORF_status=="CPN_uORF" & uORF_shared=="uORF")$log2FoldChange, filter(res_TE_CPN_df,  uORF_shared=="no_shared_uORF")$log2FoldChange)

TE_SCPN_noVsuORF.test <- meanAndTest(filter(res_TE_SCPN_df, uORF_status=="SCPN_uORF" & uORF_shared=="uORF")$log2FoldChange, filter(res_TE_SCPN_df,  uORF_shared=="no_shared_uORF")$log2FoldChange)

TE_CPN_noVsOverlap_uORF.test <- meanAndTest(filter(res_TE_CPN_df, uORF_status=="CPN_Overlap_uORF" & uORF_shared=="Overlap_uORF")$log2FoldChange, filter(res_TE_CPN_df,  uORF_shared=="no_shared_uORF")$log2FoldChange)

TE_SCPN_noVsuOverlap_ORF.test <- meanAndTest(filter(res_TE_SCPN_df, uORF_status=="SCPN_Overlap_uORF" & uORF_shared=="Overlap_uORF")$log2FoldChange, filter(res_TE_SCPN_df,  uORF_shared=="no_shared_uORF")$log2FoldChange)



TE_CPN_shared_uORF_plot <- ggplot(res_TE_CPN_df, aes(x=log2FoldChange, color=uORF_shared))+JF_theme+stat_ecdf(size=0.1)+facet_wrap(~contrast)+xlim(c(-2,2))+ylab("cumulative_fraction")+ggtitle("T.E. by uORF status, CPN")+xlab("T.E.")+theme(legend.position="top")+geom_hline(yintercept=0.5,size=0.2)+geom_vline(xintercept=0, size=0.2)
TE_CPN_shared_uORF_plot

TE_SCPN_uORF_plot <- ggplot(res_TE_SCPN_df, aes(x=log2FoldChange, color=uORF_status))+JF_theme+stat_ecdf(size=0.1)+facet_wrap(~contrast)+xlim(c(-2,2))+ylab("cumulative_fraction")+ggtitle("T.E. by uORF status, SCPN")+xlab("T.E.")+theme(legend.position="top")+geom_hline(yintercept=0.5,size=0.2)+geom_vline(xintercept=0, size=0.2)
TE_SCPN_uORF_plot

TE_SCPN_shared_uORF_plot <- ggplot(res_TE_SCPN_df, aes(x=log2FoldChange, color=uORF_shared))+JF_theme+stat_ecdf(size=0.1)+facet_wrap(~contrast)+xlim(c(-2,2))+ylab("cumulative_fraction")+ggtitle("T.E. by uORF status, SCPN")+xlab("T.E.")+theme(legend.position="top")+geom_hline(yintercept=0.5,size=0.2)+geom_vline(xintercept=0, size=0.2)
TE_SCPN_shared_uORF_plot

TE_distros_for_GO_ID <- function(CPN_df, SCPN_df, GO_IDs, allGO=ensembl_to_allGO, GO_name) {
  GO_ID_genes <- filter(allGO, go_id %in% GO_IDs)
  CPN_df <- CPN_df %>% mutate(gene_set=case_when(
    (Geneid %in% GO_ID_genes$ensembl_id)==T ~ GO_name,
    (Geneid %in% GO_ID_genes$ensembl_id)==F ~ paste("not",GO_name, sep="_")
  )) %>% mutate(uORF_simple=case_when(
    uORF_shared == "uORF" ~ "uORF",
    uORF_shared == "Overlap_uORF" ~ "uORF",
    T ~ "no_uORF"
  ))
  SCPN_df <- SCPN_df %>% mutate(gene_set=case_when(
    (Geneid %in% GO_ID_genes$ensembl_id)==T ~ GO_name,
    (Geneid %in% GO_ID_genes$ensembl_id)==F ~ paste("not",GO_name, sep="_")
  )) %>% mutate(uORF_simple=case_when(
    uORF_shared == "uORF" ~ "uORF",
    uORF_shared == "Overlap_uORF" ~ "uORF",
    T ~ "no_uORF"
  ))
  CPN_GO_plot <- ggplot(CPN_df, aes(x=log2FoldChange, color=interaction( uORF_simple,gene_set)))+JF_theme+stat_ecdf(size=0.1)+xlim(c(-2,2))+ylab("cumulative_fraction")+ggtitle("T.E. by uORF status, CPN")+xlab("T.E.")+geom_hline(yintercept=0.5, size=0.2)+geom_vline(xintercept=0, size=0.2)+theme(legend.position = "top")
  SCPN_GO_plot <- ggplot(SCPN_df, aes(x=log2FoldChange, color=interaction( uORF_simple,gene_set)))+JF_theme+stat_ecdf(size=0.1)+xlim(c(-2,2))+ylab("cumulative_fraction")+ggtitle("T.E. by uORF status, SCPN")+xlab("T.E.")+geom_hline(yintercept=0.5, size=0.2)+geom_vline(xintercept=0, size=0.2)+theme(legend.position = "top")
  CPN_filt <- filter(CPN_df, gene_set==GO_name)
  SCPN_filt <- filter(SCPN_df, gene_set==GO_name)
  CPN_test <- meanAndTest(filter(CPN_filt, uORF_simple =="uORF")$log2FoldChange,filter(CPN_filt, uORF_simple =="no_uORF")$log2FoldChange)
   SCPN_test <- meanAndTest(filter(SCPN_filt, uORF_simple =="uORF")$log2FoldChange,filter(SCPN_filt, uORF_simple =="no_uORF")$log2FoldChange)
  return_list <- list("CPN_plot"=CPN_GO_plot, "SCPN_plot"=SCPN_GO_plot, "CPN_test"=CPN_test,"SCPN_test"=SCPN_test)
  return(return_list)
}

synapse_organization_TE_plots <- TE_distros_for_GO_ID(CPN_df = res_TE_CPN_df, SCPN_df = res_TE_SCPN_df, GO_IDs="GO:0050808", GO_name="Synapse_Org")
#Can get a full res with uORF and GO-term status by pulling the data out from the ggplots
synapse_organization_genes_res <- synapse_organization_TE_plots$CPN_plot$data %>% left_join(mouse_gene_symbols_wGeneid, by="Geneid") %>% filter(gene_set=="Synapse_Org")


axonogenesis_TE_plots <- TE_distros_for_GO_ID(CPN_df = res_TE_CPN_df, SCPN_df = res_TE_SCPN_df, GO_IDs="GO:0007409", GO_name="Axonogenesis")
axon_development_TE_plots <-  TE_distros_for_GO_ID(CPN_df = res_TE_CPN_df, SCPN_df = res_TE_SCPN_df, GO_IDs="GO:0061564", GO_name="Axon dev.")
axon_development_genes_res <- axon_development_TE_plots$CPN_plot$data %>% left_join(mouse_gene_symbols_wGeneid, by="Geneid") %>% filter(gene_set=="Axon dev.")

mRNA_processing_TE_plots <- TE_distros_for_GO_ID(CPN_df = res_TE_CPN_df, SCPN_df = res_TE_SCPN_df, GO_IDs="GO:0006397", GO_name="mRNA processing")
RNP_biogenesis_TE_plots <-  TE_distros_for_GO_ID(CPN_df = res_TE_CPN_df, SCPN_df = res_TE_SCPN_df, GO_IDs="GO:0022613", GO_name="ribonucleoprotein complex biogenesis")
synapse_assembly_TE_plots <- TE_distros_for_GO_ID(CPN_df = res_TE_CPN_df, SCPN_df = res_TE_SCPN_df, GO_IDs="GO:0007416", GO_name="synapse assembly")

res_TE_CPN_df <- res_TE_CPN_df %>% mutate(TE_quintile=ntile(log2FoldChange,n=5))
res_TE_SCPN_df <- res_TE_SCPN_df %>% mutate(TE_quintile=ntile(log2FoldChange,n=5))

TE_CPN_uORF_byQuntile_plot <- ggplot(res_TE_CPN_df, aes(x=log2FoldChange, color=uORF_status))+JF_theme+stat_ecdf(size=0.1)+facet_wrap(~TE_quintile,scales="free_x")+ylab("cumulative_fraction")+ggtitle("T.E. by uORF status, CPN")+xlab("T.E.")+theme(legend.position="top")+geom_hline(yintercept=0.5)+geom_vline(xintercept=0)+geom_hline(yintercept=0.5,size=0.2)+geom_vline(xintercept=0, size=0.2)


```

```{r uORF_examples}

TPMs_filtered_tidy <- TPMs_filtered %>% gather(key="samps",value="TPM",-Geneid) %>% left_join(experimentTypesPlus_mRNACounts, by="samps") %>% left_join(mouse_gene_symbols_wGeneid, by="Geneid") %>% mutate(exp=factor(exp),subtype=factor(subtype))


plotGeneOfInterestCounts_TE_linear <- function(geneCounts, se=F) {
  geneFacetPlot <- ggplot(geneCounts, aes(x=exp, y=TPM, color=factor(subtype)))+JF_theme+geom_point(alpha=0.5)+geom_smooth(method="lm", aes(x=as.numeric(exp),y=TPM), se=se)+facet_wrap(~external_gene_name, scales="free_y", nrow=1)+scale_color_manual(values=c("orange","blue"))+theme(strip.text = element_text(face="bold", size=7, margin=margin(c(0,0,0,0, "cm"))))
}

Nrxn_TE_plots <- plotGeneOfInterestCounts_TE_linear(filter(TPMs_filtered_tidy, external_gene_name %in% c("Nrxn1","Nrxn2")))
Nlgn_TE_plots <- plotGeneOfInterestCounts_TE_linear(filter(TPMs_filtered_tidy, external_gene_name %in% c("Nlgn1","Nlgn2","Nlgn3")))
Sema_TE_plots <- plotGeneOfInterestCounts_TE_linear(filter(TPMs_filtered_tidy, external_gene_name %in% c("Sema4c","Sema4g","Sema6c","Sema6d"))) 

psiteFrameORFPlotting <- function(res, gene, ORF, specific_transcript=NULL){
  #function to take in a gene and an ORF type, and plot the number of Psites in each frame over each of the ORFs of the indicated type
  res_filt <- res %>% 
    filter(gene_name == gene & ORF_type == ORF ) %>%
    mutate(dist_to_CDS = ORF_tstart - as.numeric(annotated_tstart)) %>%
     mutate(tx_ID=transcript_id) %>%
    dplyr::select(gene_name, ORF_type, ORF_tstart, Psites_sum_frame0, Psites_sum_frame1, Psites_sum_frame2, ORF_length, annotated_tstart, tx_ID, dist_to_CDS, subtype) %>%
    gather(key="Frame", value="Counts",-gene_name,-ORF_type, -ORF_tstart,-ORF_length,-annotated_tstart,-tx_ID,-dist_to_CDS, -subtype) %>%
    mutate(Frame=replace(Frame, Frame== "Psites_sum_frame0", 0)) %>%
    mutate(Frame=replace(Frame, Frame== "Psites_sum_frame1", 1)) %>%
    mutate(Frame=replace(Frame, Frame== "Psites_sum_frame2", 2))

if (is.null(specific_transcript)==T) {
  psiteFrameCounts_plot <- ggplot(res_filt, aes(x=Frame,y=Counts,fill=Frame))+geom_col()+facet_wrap(~subtype+dist_to_CDS)+theme(strip.text = element_text(face="bold", size=7, margin=margin(c(0,0,0,0, "cm"))),  legend.position = "none")+ggtitle(paste(gene, ORF, sep=" "))
}
else {
  psiteFrameCounts_plot <- ggplot(filter(res_filt, tx_ID==specific_transcript), aes(x=Frame,y=Counts,fill=Frame))+geom_col()+facet_wrap(~subtype+dist_to_CDS)+theme(strip.text = element_text(face="bold", size=7, margin=margin(c(0,0,0,0, "cm"))),  legend.position = "none")+ggtitle(paste(gene, ORF, sep=" "))
}
  
}

uORF_res_sharedOnly <- filter(uORF_res, ORF_ID %in% uORF_shared | ORF_ID %in% Overlap_uORF_shared)

Nrxn1beta_psiteFrames_plot <- psiteFrameORFPlotting(uORF_res_sharedOnly,"Nrxn1","uORF", specific_transcript="ENSMUST00000159778")+JF_theme
Nrxn1beta_psiteFrames_plot

Nrxn1alpha_psiteFrames_plot <- psiteFrameORFPlotting(uORF_res_sharedOnly,"Nrxn1","uORF", specific_transcript="ENSMUST00000160844")+JF_theme
Nrxn1alpha_psiteFrames_plot

Nrxn2_psiteFrames_plot <- psiteFrameORFPlotting(uORF_res_sharedOnly,"Nrxn2","uORF")+JF_theme
Nrxn2_psiteFrames_plot

Nlgn2_psiteFrames_plot <- psiteFrameORFPlotting(uORF_res_sharedOnly,"Nlgn2","uORF")+JF_theme
Nlgn2_psiteFrames_plot

Nlgn3_psiteFrames_plot <- psiteFrameORFPlotting(uORF_res_sharedOnly,"Nlgn3","uORF")+JF_theme
Nlgn3_psiteFrames_plot

Sema4g_psiteFrames_plot <- psiteFrameORFPlotting(uORF_res_sharedOnly,"Sema4g","uORF")+JF_theme
Sema4g_psiteFrames_plot

Sema4c_psiteFrames_plot <- psiteFrameORFPlotting(uORF_res_sharedOnly,"Sema4c","uORF")+JF_theme
Sema4c_psiteFrames_plot

Sema6c_psiteFrames_plot <- psiteFrameORFPlotting(uORF_res_sharedOnly,"Sema6c","uORF")+JF_theme
Sema6c_psiteFrames_plot

Sema6d_psiteFrames_plot <- psiteFrameORFPlotting(uORF_res_sharedOnly,"Sema6d","uORF")+JF_theme
Sema6d_psiteFrames_plot


```


```{r uORFs_figure}
uORFs_figure <- ORF_types_plot / (TE_CPN_uORF_plot + TE_SCPN_uORF_plot) / (uORF_ego_barplot$CPN_uORFs + uORF_ego_barplot$SCPN_uORFs) / (synapse_assembly_TE_plots$CPN_plot + synapse_assembly_TE_plots$SCPN_plot) + plot_annotation(tag_levels = list(c("A","B","C","D","E","F","G","H"))) & theme(plot.tag=element_text(size=12, face="bold"))+theme(legend.key.size=unit(0.5,"line"))
pdfName=paste(outDir, "uORF_fig.pdf", sep="/")
pdf(file=pdfName, width=6.5, height=9)
plot(uORFs_figure)
dev.off()
```
```{r uORFs_alt_figure}
uORFs_layout <-
  'AAAAABBB
   CCCCDDDD
   #EEEEEEE
   FFFFGGGG
   HHHHIIII
  '
uORF_alt_figure <-  ORF_types_plot+uORF_shared_col_plot + TE_CPN_shared_uORF_plot + TE_SCPN_shared_uORF_plot + uORF_ego_barplot$uORF_shared + synapse_organization_TE_plots$CPN_plot+ synapse_organization_TE_plots$SCPN_plot + axon_development_TE_plots$CPN_plot+axon_development_TE_plots$SCPN_plot+plot_layout(design=uORFs_layout)+plot_annotation(tag_levels = list(c("A","B","C","D","E","F","G","H","I"))) & theme(plot.tag=element_text(size=12, face="bold"))+theme(legend.key.size=unit(0.5,"line"))

# uORF_alt_figure <-  (ORF_types_plot+uORF_shared_col_plot) / (TE_CPN_shared_uORF_plot + TE_SCPN_shared_uORF_plot) / (uORF_ego_barplot$uORF_shared) / (synapse_organization_TE_plots$CPN_plot+ synapse_organization_TE_plots$SCPN_plot) /(axon_development_TE_plots$CPN_plot+axon_development_TE_plots$SCPN_plot)+plot_layout(design=uORFs_layout)+plot_annotation(tag_levels = list(c("A","B","C","D","E","F","G","H","I"))) & theme(plot.tag=element_text(size=12, face="bold"))+theme(legend.key.size=unit(0.5,"line"))

pdfName=paste(outDir, "uORF_alt_fig.pdf", sep="/")
pdf(file=pdfName, width=6.5, height=9)
plot(uORF_alt_figure)
dev.off()

```

```{r uORF_example_fig}
uORF_example_fig <- (Sema_TE_plots) / (Sema4c_psiteFrames_plot | Sema4g_psiteFrames_plot | Sema6c_psiteFrames_plot | Sema6d_psiteFrames_plot) / (Nlgn_TE_plots) / (Nlgn2_psiteFrames_plot+Nlgn3_psiteFrames_plot) / (Nrxn_TE_plots) / (Nrxn1alpha_psiteFrames_plot+Nrxn1beta_psiteFrames_plot) / (Nrxn2_psiteFrames_plot+plot_spacer())+plot_layout(heights=c(1,2,1,2,1,2,2))+ plot_annotation(tag_levels = list(c("A","B","C","D","E","F","G","H","I","J","K"))) & theme(plot.tag=element_text(size=12, face="bold"))

pdfName=paste(outDir, "uORF_example_fig.pdf", sep="/")
pdf(file=pdfName, width=6.5, height=9)
plot(uORF_example_fig)
dev.off()

```


###Run RiboWaltz to evaluate 3-nt periodicity in the data sets. 
###Requires a directory of transcriptome-aligned BamFiles and a gtf
```{r bamToList, echo=FALSE}
#Try loading in all the files and see how it goes. After a more careful
#reading of the RiboWaltz vignette, it looks like it will process samples separately.
#make a named vector mapping bamFile names (no path or extension needed) to descriptiveNames
RP_samps <- filter(experimentTypes, exp=="RP")
RP_samps <- data.frame(RP_samps)
RP_descriptiveNames <- as.character(RP_samps$Samples)
names(RP_descriptiveNames) <- as.character(RP_samps$samps)
write.csv(file="test337.txt", RP_descriptiveNames)
annotation_dt <- create_annotation(gtfpath = gtf, dataSource = "GRCm38.95",organism="Mus musculus")
reads_list <- bamtolist(bamfolder = bamFiles, annotation = annotation_dt, transcript_align = FALSE, name_samples = RP_descriptiveNames) 
```

```{r lengthFilter, echo=FALSE}
#From last run, need to set length_filter_mode="custom" for "bad" or
#RNA-seq libraries to get scored. 
filtered_list <- riboWaltz::length_filter(data = reads_list, length_filter_mode = "custom", length_range = c(27:33))
```

```{r psite, echo=FALSE}
#perform p-site analysis for all samples, save data to a sub-folder of the transcriptome alignment.
dir.create(outDir)
psite_dir <- paste(outDir,'/psite', sep="")
dir.create(psite_dir)
psite <- psite(data=filtered_list, plot=T, plot_dir = psite_dir, plot_format = "pdf")
```

###Make plots that stratify by frame:
####At first, I plotted using reads_list. This makes plots across all reads lengths
####Plotting using the length-filtered reads (filtered_list) makes much nicer graphs.
```{r stratify, echo=FALSE}
reads_psite_list <- psite_info(filtered_list,psite)
frames_stratified <- frame_psite_length(reads_psite_list, region="all",cl=90)
strat_dir <- paste(outDir,"/frame_counts",sep="")
dir.create(strat_dir)
#setwd(strat_dir)
pdfName <- paste(strat_dir, "Frame_counts_all_samples.pdf", sep="/" )
frames_stratified$plot
pdf(pdfName)
frames_stratified$plot
dev.off()
#The plot looks squished if I try plotting all of them on the same plot. Try looping over the sampleNames and making individual plots for each one. Update: If I only include filtered reads, the graphs aren't too squished and are readable
sampleNames=unique(psite$sample)
for(sampleName in sampleNames) {
  frame_stratified=frame_psite_length(reads_psite_list,  sample=sampleName, region="all",cl=90)
  pdfName=paste(sampleName,".pdf",sep="")
  pdf(file=pdfName)
  plot(frame_stratified$plot)
  dev.off()
}
```

###P-site heatmaps (both directly from RiboWaltz, and by re-plotting the RiboWaltz data for a nicer-looking plot) for all samples in current set. 
###Currently using scale factors as 1/CDS_reads:
```{r psite_heatmap, echo=FALSE}
heat_dir <- paste(outDir,"/heatmaps",sep="")
dir.create(heat_dir)
#setwd(heat_dir)
#calculate scale factors as the inverse of CDS_counts
scale_factors <- filter(featureCounts_tidy, exp=="RP") %>%
  group_by(samps) %>%
  summarize(CDS_reads=sum(counts)) %>%
  mutate(scale_factors=1/CDS_reads) %>%
  dplyr::select(scale_factors)
scale_factors <- c(t(scale_factors))
names(scale_factors) <- sampleNames
#Build up the list of sample names
sampNamesList <- list()
for (sampName in sampleNames) {
  sampNamesList[[sampName]] <- c(sampName)
}
psite_heatmap <- metaheatmap_psite(reads_psite_list, annotation=annotation_dt,log=F,utr5l=15, cdsl=30, utr3l=15, sample=sampNamesList, scale_factors=scale_factors)
psite_heatmap$plot
pdfName=paste(heat_dir, "psite_heatmap.pdf",sep="/")
pdf(file=pdfName, width=11, height=8.5)
plot(psite_heatmap$plot)
dev.off()
#Example plotting function that looks nicer than a heatmap:
psite_signal_plot <- ggplot(psite_heatmap$dt, aes(x=distance,y=reads,color=sample))+geom_line()+JF_theme+facet_wrap(~reg)
psite_signal_plot
psite_dt <- psite_heatmap$dt
psite_dt$reg <- plyr::revalue(psite_dt$reg, c("Distance from start (nt)"="from start (nt):","Distance from stop (nt)"="from stop (nt):"))
psite_signal_heat <- ggplot(psite_dt, aes(x=distance,fill=reads,y=sample))+geom_tile()+facet_wrap(~reg)+scale_fill_gradient(low="white", high="black")+JF_theme+xlab("distance from start or stop (nt)")+ylab("")+xlab("")
psite_signal_heat
#look at how scaling is done. Try row normalization of raw read counts to compute frequencies within the regions of interest
psite_heatmap_raw <- metaheatmap_psite(reads_psite_list, annotation=annotation_dt,log=F,utr5l=15, cdsl=30, utr3l=15, sample=sampNamesList)
psite_dt_raw <- psite_heatmap_raw$dt
psite_dt_raw$reg <- plyr::revalue(psite_dt_raw$reg, c("Distance from start (nt)"="from start (nt):","Distance from stop (nt)"="from stop (nt):"))
psite_region_sums <- psite_dt_raw %>%
  group_by(sample, reg) %>%
  summarize(region_sum=sum(reads))
psite_dt_raw <- merge(psite_dt_raw, psite_region_sums, by=c("sample","reg"))
psite_dt_raw <- psite_dt_raw %>%
  mutate(freq=reads/region_sum) %>% 
  mutate(Samples=sample)
titration_metadata <- dplyr::select(experimentTypes, Samples, exp, date, cells) %>% filter(exp=="RP")
psite_dt_raw <- merge(psite_dt_raw, titration_metadata, by="Samples")
psite_dt_raw <- psite_dt_raw %>%
  mutate(Samples= fct_reorder(Samples,as.numeric(cells), .desc=F))
#psite_dt_raw <- merge(psite_dt_raw, dplyr::filter(experimentTypes, Samples, exp_date), by="Samples")
psite_freqHeat <- ggplot(psite_dt_raw, aes(x=distance,fill=freq,y=sample))+geom_tile()+facet_wrap(~reg)+scale_fill_gradient(low="white", high="black")+JF_theme+xlab("distance from start or stop (nt)")+ylab("")+xlab("")+theme(legend.position="bottom")
psite_freqHeat
pdfName=paste(heat_dir, "psite_freq_heatmap.pdf",sep="/")
pdf(file=pdfName, width=11, height=8.5)
plot(psite_freqHeat)
dev.off()
```

###Next, calculate the percent of P-sites within each reading frame for each sample.
```{r framePsite, echo=FALSE}
frame_dir <- paste(outDir,"/frame_psite_percentage",sep="")
dir.create(frame_dir)
#setwd(frame_dir)
frame_psites <- frame_psite(reads_psite_list, region="all")
frame_psites$plot
pdfName <- paste(frame_dir, "Frame_percentages_all_samples.pdf", sep="/")
pdf(pdfName)
plot(frame_psites$plot)
dev.off()
sampleNames=unique(psite$sample)
for(sampleName in sampleNames) {
  aFramePsite=frame_psite(reads_psite_list, sample=sampleName, region="all")
  pdfName=paste(sampleName,".pdf",sep="")
  pdf(file=pdfName)
  plot(aFramePsite$plot)
  dev.off()
}
```

###And make P-site frequency plots for comparison across samples:
```{r psitePlotting, echo=FALSE}
plotFramePsites <- function(samples, frame_psites, sampNames=samples) {
  #function to take in a list of samples, and frame_psites Ribowaltz object    #Plots the percentage of reads over CDSs in each frame, grouped by samples.
  FPs <- frame_psites$dt
  FPsToPlot <- filter(FPs, sample %in% samples & region == "CDS" )
  FPplot <- ggplot(data=FPsToPlot, aes(x=sample,y=percentage,fill=as.factor(frame)))
  FPplot <- FPplot+geom_bar(stat="identity", position=position_dodge())+theme_minimal(base_size=18)+scale_fill_discrete(name="frame")+ylim(0,100)+scale_x_discrete(labels=sampNames)+JF_theme
  FPplot <- FPplot+theme(axis.text.x = element_text(angle = 45, hjust=1))+xlab("")+ylab("%P-sites in each frame")
                           
}
#setwd(frame_dir)
psiteFreq<- plotFramePsites(sampleNames,frame_psites)
pdfName=paste(frame_dir, "psite_freq.pdf",sep="/")
pdf(file=pdfName, width=11, height=8.5)
plot(psiteFreq)
dev.off()
psiteFreq
```

###Plots for reviewer #3's questions about P-site frequency broken down by TE status.
```{r Reviewer3_chunk}

CPN_TE_high_tx <- ensmusg2ensmust %>% filter(Geneid %in% CPN_TE_high)
SCPN_TE_high_tx <- ensmusg2ensmust %>% filter(Geneid %in% SCPN_TE_high)
CPN_TE_low_tx <- ensmusg2ensmust %>% filter(Geneid %in% CPN_TE_low)
SCPN_TE_low_tx <- ensmusg2ensmust %>% filter(Geneid %in% SCPN_TE_low)
CPN_TE_notSig_tx <- ensmusg2ensmust %>% filter( (Geneid %in% CPN_TE_high)==F & (Geneid %in% CPN_TE_low)==F)
SCPN_TE_notSig_tx <- ensmusg2ensmust %>% filter( (Geneid %in% SCPN_TE_high)==F & (Geneid %in% SCPN_TE_low)==F)

transcripts_by_TE_list <- list("CPN_TE_high"=CPN_TE_high_tx$ensembl_transcript_id, "CPN_TE_low"=CPN_TE_low_tx$ensembl_transcript_id, "SCPN_TE_high"=SCPN_TE_high_tx$ensembl_transcript_id, "SCPN_TE_low"=SCPN_TE_low_tx$ensembl_transcript_id, "CPN_TE_noSig"=CPN_TE_notSig_tx$ensembl_transcript_id, "SCPN_TE_noSig"=SCPN_TE_notSig_tx$ensembl_transcript_id )

split_psites_by_TE_list <- function(transcript_list, RPL=reads_psite_list) {
  split_RPL <- lapply(RPL, function(x) {
    filter(x, transcript %in% transcript_list)
  })
}

splitPs <- lapply(transcripts_by_TE_list, split_psites_by_TE_list)

splitP_frame_psites <- lapply(splitPs, frame_psite, region="all")
splitP_frame_psite_bySample_plots <- lapply(splitP_frame_psites, plotFramePsites, samples=sampleNames)
splitP_frame_psite_bySample_plots$CPN_TE_high <- splitP_frame_psite_bySample_plots$CPN_TE_high+ggtitle("CPN_TE_high")+theme(legend.position="top")
splitP_frame_psite_bySample_plots$SCPN_TE_high <- splitP_frame_psite_bySample_plots$SCPN_TE_high+ggtitle("SCPN_TE_high")+theme(legend.position="top")
splitP_frame_psite_bySample_plots$CPN_TE_low <- splitP_frame_psite_bySample_plots$CPN_TE_low+ggtitle("CPN_TE_low")+theme(legend.position="top")
splitP_frame_psite_bySample_plots$SCPN_TE_low <- splitP_frame_psite_bySample_plots$SCPN_TE_low+ggtitle("SCPN_TE_low")+theme(legend.position="top")
splitP_frame_psite_bySample_plots$CPN_TE_noSig <- splitP_frame_psite_bySample_plots$CPN_TE_noSig+ggtitle("SCPN_TE_noSig")+theme(legend.position="top")
splitP_frame_psite_bySample_plots$SCPN_TE_noSig <- splitP_frame_psite_bySample_plots$SCPN_TE_noSig+ggtitle("SCPN_TE_noSig")+theme(legend.position="top")

split_Reads <- lapply(transcripts_by_TE_list, split_psites_by_TE_list, RPL=reads_list)

cds_distro_from_reads_list <- function(RL=reads_list) {
  RL_length_distros <- lapply(RL, function(x) {
  RL_lengths <- data.frame(x) %>% group_by(length) %>% summarize(length_counts_bySample=n())
  })
  RL_lengths <- bind_rows(RL_length_distros, .id="sample") %>% mutate(subtype=case_when(regexpr('^CPN', sample)==T~"CPN",regexpr('^SCPN', sample)==T~"SCPN"))
  lengths_bySubtype <- RL_lengths %>% group_by(subtype, length) %>% summarize(length_counts_bySubtype=sum(length_counts_bySample))
  readSums_bySubtype <- lengths_bySubtype %>% group_by(subtype) %>% summarize(readSums=sum(length_counts_bySubtype))
  lengths_bySubtype <- lengths_bySubtype %>% left_join(readSums_bySubtype, by="subtype") %>% mutate(freq_bySubtype=length_counts_bySubtype/readSums)
}

split_Read_distros <- lapply(split_Reads, cds_distro_from_reads_list)  

split_Read_distros <- bind_rows(split_Read_distros, .id="TE_status")

CPN_length_distro_by_TE_plot <- ggplot(filter(split_Read_distros, subtype=="CPN" & regexpr("^CPN",TE_status)==T), aes(x=length,y=freq_bySubtype, color=TE_status))+JF_theme+geom_line()+theme(legend.position="top")+ggtitle("read length distro over CDS by CPN T.E. status")+xlim(c(20,40))
CPN_length_distro_by_TE_plot
SCPN_length_distro_by_TE_plot <- ggplot(filter(split_Read_distros, subtype=="SCPN" & regexpr("^SCPN",TE_status)==T), aes(x=length,y=freq_bySubtype, color=TE_status))+JF_theme+geom_line()+theme(legend.position="top")+ggtitle("read length distro over CDS by SCPN T.E. status")+xlim(c(20,40))
SCPN_length_distro_by_TE_plot 

#moveTo new Rmd
loadFC2tidyCounts <- function(experimentType, CDS, five_utr, three_utr) {
sampNames=experimentType$samps
#write.csv(file="test.txt", sampNames)
CDS <- read.table(file=CDS, header= T, sep="\t", stringsAsFactors = F)
Five_prime <- read.table(file=five_utr, header= T, sep="\t", stringsAsFactors = F)
Three_prime <- read.table(file=three_utr, header= T, sep="\t", stringsAsFactors = F)
filteredCDS <- CDS %>% 
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)
filtered5p <- Five_prime %>%
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)
filtered3p <- Three_prime %>%
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)

maxSampCol <- 6+length(sampNames)
mRNA_counts <- filteredCDS[,7:maxSampCol]+filtered5p[,7:maxSampCol]+filtered3p[,7:maxSampCol]
CDS_counts <- filteredCDS[,7:maxSampCol]
FiveP_counts <- filtered5p[,7:maxSampCol]
ThreeP_counts <- filtered3p[,7:maxSampCol]
colnames(CDS_counts) <- sampNames
colnames(FiveP_counts) <- sampNames
colnames(ThreeP_counts) <- sampNames
CDS_counts <- CDS_counts %>% mutate(Geneid=filteredCDS$Geneid, Length=filteredCDS$Length, region="CDS")
FiveP_counts <-FiveP_counts %>% mutate(Geneid=filteredCDS$Geneid, Length=filtered5p$Length, region="5' UTR")
ThreeP_counts <-ThreeP_counts %>% mutate(Geneid=filteredCDS$Geneid, Length=filtered3p$Length, region="3' UTR")

All_counts <- rbind(CDS_counts, FiveP_counts, ThreeP_counts)
All_counts_tidy <- gather(All_counts, "samps","counts", -Geneid, -Length, -region)
All_counts_tidy <- merge(All_counts_tidy, experimentType, by="samps")
}

All_counts_tidy <- loadFC2tidyCounts(experimentTypes, CDS, five_utr, three_utr)

makeRegionCov <- function(ACT, sampNames=descriptiveNames) {
  rpk <- ACT %>%
  group_by(samps, region, exp) %>%
  summarize(totCounts=sum(counts),totLength=sum(Length)) %>%
  mutate(rpk=totCounts/(totLength/1000))

million_mapped <- ACT %>%
  group_by(samps) %>%
  summarize(mRNA_mapped_mil=sum(counts)/1e6)

rpkm <- merge(rpk, million_mapped, by="samps") %>%
  mutate(rpkm=rpk/mRNA_mapped_mil)

rpkm$region <- factor(rpkm$region, levels=c("5' UTR","CDS","3' UTR"))
rpkm <- merge(rpkm, sampNames, by="samps")
# rpkm_byRegion <- rpkm_byRegion %>%
#   mutate(Samples=sampNames)
}

test <- makeRegionCov(ACT=All_counts_tidy)

geneids_by_TE_list <- list("CPN_TE_high"=CPN_TE_high, "CPN_TE_low"=CPN_TE_low, "CPN_TE_noSig"=CPN_TE_notSig,
                           "SCPN_TE_high"=SCPN_TE_high, "SCPN_TE_low"=SCPN_TE_low, "SCPN_TE_noSig"=SCPN_TE_notSig)
split_ACT <- lapply(geneids_by_TE_list, function(x, ACT=All_counts_tidy) {
  filter(ACT, Geneid %in% x)
})
split_ACT_regionCovs <- lapply(split_ACT, makeRegionCov)
split_ACT_regionCovs <- split_ACT_regionCovs %>% bind_rows(.id="TE_status") %>% mutate(subtype=case_when(regexpr('^CPN', descriptiveNames)==T~"CPN",regexpr('^SCPN', descriptiveNames)==T~"SCPN"))

CPN_regionCov_by_TE_status <- ggplot(filter(split_ACT_regionCovs, subtype=="CPN" & exp=="RP" & regexpr("^CPN",TE_status)==T), aes(x=TE_status, y=rpkm, fill=region))+JF_theme+geom_col(position="dodge")+theme(legend.position = "top")+facet_wrap(~descriptiveNames, nrow=1)+ggtitle("Coverage by region by T.E. status, CPN")+theme(axis.text.x = element_text(angle = 45, hjust=1))
CPN_regionCov_by_TE_status 
SCPN_regionCov_by_TE_status <- ggplot(filter(split_ACT_regionCovs, subtype=="SCPN" & exp=="RP" & regexpr("^SCPN",TE_status)==T), aes(x=TE_status, y=rpkm, fill=region))+JF_theme+geom_col(position="dodge")+theme(legend.position = "top")+facet_wrap(~descriptiveNames, nrow=1)+ggtitle("Coverage by region by T.E. status, SCPN")+theme(axis.text.x = element_text(angle = 45, hjust=1))
SCPN_regionCov_by_TE_status

TE_riboWaltz_QC_fig <- (CPN_length_distro_by_TE_plot | SCPN_length_distro_by_TE_plot) / (splitP_frame_psite_bySample_plots$CPN_TE_high + splitP_frame_psite_bySample_plots$CPN_TE_noSig + splitP_frame_psite_bySample_plots$CPN_TE_low) / (splitP_frame_psite_bySample_plots$SCPN_TE_high + splitP_frame_psite_bySample_plots$SCPN_TE_noSig + splitP_frame_psite_bySample_plots$SCPN_TE_low) / (CPN_regionCov_by_TE_status | SCPN_regionCov_by_TE_status ) + plot_annotation(tag_levels = list(c("A","B","C","D","E","F","G","H","I","J"))) & theme(plot.tag=element_text(size=12, face="bold"))+theme(legend.key.size=unit(0.5,"line"))

pdfName=paste(outDir, "TE_riboWaltz_QC_fig.pdf", sep="/")
pdf(file=pdfName, width=6.5, height=9)
plot(TE_riboWaltz_QC_fig)
dev.off()

```
```{r sesh}
sessionInfo()

```

