---
author: "John E Froberg"
date: Sys.Date()
title: "CountsPlotting"
output: html_document
---

```{r setup}
knitr::opts_chunk$set(echo=FALSE, message=FALSE)


# #variables that are normally passed in from the Snakemake run:
# experimentTypeFile="experimentTypesFull.csv"
# lengthCountsFile="dedup_lengthDistroTidy.txt"
# CDS="dedupBams/featureCounts_CDS_summary.txt"
# three_utr="dedupBams/featureCounts_three_prime_utr_summary.txt"
# five_utr="dedupBams/featureCounts_five_prime_utr_summary.txt"
# gtf="Mus_musculus.GRCm38.95_chrNamed_headFix.gtf"
# bamFiles="dedup_RPbams"
# outDir="test"

JF_theme <- theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "black"), plot.title = element_text(size = 7, face = "bold"),legend.title=element_text(size=7), legend.text=element_text(size=7), axis.text = element_text(size=7), axis.title.x = element_text(size=7),axis.title.y=element_text(size=7))
```


###Load in the metadata from the experiment types file. 
###Compute "descriptive names" by dropping the "samps" and "date" fields and pasting together.
###Save descriptiveNames as Upper Case "Samples" in experimentTypes.
```{r loadMetadata, echo=FALSE}
experimentTypes=read.csv(experimentTypeFile)
 descriptiveNames <- experimentTypes %>%
  unite(col="descriptiveNames", -samps, -date,-exp, sep=".")
 descriptiveNames <- select(descriptiveNames, samps, descriptiveNames)
 experimentTypes$Samples <- descriptiveNames$descriptiveNames
 experimentTypes
```

###Convert length counts to relative frequencies, add in the metadata
```{r countsToFreqwMeta}
lengthCounts <- read.csv(lengthCountsFile)
totalCounts <- lengthCounts %>%
  group_by(samps,anno) %>%
  summarize(sum=sum(count))

lengthDistro <- merge(lengthCounts, totalCounts, by=c("samps", "anno"))
lengthDistro <- lengthDistro %>%
  mutate(freq=count/sum)

#metadata <- read.csv("metadata.csv")
lengthDistro <- merge(lengthDistro, experimentTypes, by="samps")

```
###Find the length with max freq for all samples
```{r maxFreq}
maxFreq <- lengthDistro %>%
  group_by(samps, anno) %>%
  summarize(max=max(freq))

max <- merge(maxFreq, lengthDistro, by=c("samps","anno"))
maxLengthDistro <- filter(max, freq==max)
maxLengthCDS_RPs <- filter(maxLengthDistro, exp=="RP" & anno=="CDS")
```


###Plot Freq vs length for the CDS with various groupings of the samples
```{r freqPlotting}
CDS_riboProf <- ggplot(filter(lengthDistro, exp=="RP",anno=="CDS"), aes(x=length, y=freq, color=Samples))+geom_line(size=1)+JF_theme+xlim(20,40)+xlab("length (nt)")+theme(axis.title.x = element_text(vjust=0))

CDS_riboProf_byDate <- ggplot(filter(lengthDistro, exp=="RP",anno=="CDS"), aes(x=length, y=freq, color=Samples))+geom_line(size=1)+facet_wrap(~date)+xlim(20,40)+JF_theme

CDS_riboProf_byDate_noxlim <- ggplot(filter(lengthDistro, exp=="RP",anno=="CDS"), aes(x=length, y=freq, color=descriptiveNames))+geom_line(size=1)+facet_wrap(~date)+ylim(0,0.5)+JF_theme

if ( "AF" %in% levels(lengthDistro$exp) ) {
CDS_AF_byDate <-ggplot(filter(lengthDistro, exp=="AF",anno=="CDS"), aes(x=length, y=freq, color=Samples))+geom_line(size=1)+facet_wrap(~date)+ylim(0,0.5)+JF_theme
CDS_AF_byDate
}


allAnno_RiboProf_bySample <- ggplot(filter(lengthDistro, exp=="RP", anno %in% c("CDS","three_prime_utr","snoRNA")), aes(x=length, y=freq, color=anno))+geom_line(size=1)+facet_wrap(~Samples, nrow=ceiling(nlevels(factor(lengthDistro$Samples))/6))+xlim(20,40)+JF_theme+xlab("length (nt)")

CDS_riboProf
CDS_riboProf_byDate
allAnno_RiboProf_bySample
```



###Load in the featureCounts. Use only CDS reads for RiboProf, 3'+5'UTR+CDS reads for AlkFrag libraries
```{r loadFeatureCounts, echo=FALSE}
loadFC <- function(experimentType, CDS, five_utr, three_utr) {
sampNames=experimentType$samps
CDS <- read.table(file=CDS, header= T, sep="\t", stringsAsFactors = F)
Five_prime <- read.table(file=five_utr, header= T, sep="\t", stringsAsFactors = F)
Three_prime <- read.table(file=three_utr, header= T, sep="\t", stringsAsFactors = F)
filteredCDS <- CDS %>% 
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)
filtered5p <- Five_prime %>%
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)
filtered3p <- Three_prime %>%
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)

#move to new Rmd
maxSampCol <- 6+length(sampNames)
mRNA_counts <- filteredCDS[,7:maxSampCol]+filtered5p[,7:maxSampCol]+filtered3p[,7:maxSampCol]
CDS_counts <- filteredCDS[,7:maxSampCol]
#

colnames(mRNA_counts) <- sampNames
colnames(CDS_counts) <- sampNames
mRNA_counts <- mRNA_counts %>%
  mutate(Geneid=filteredCDS$Geneid, Length=filteredCDS$Length)
CDS_counts <- CDS_counts %>%
  mutate(Geneid=filteredCDS$Geneid, Length=filteredCDS$Length)
mRNA_counts_tidy <- gather(mRNA_counts,"samps","counts", -Geneid, -Length)
CDS_counts_tidy <- gather(CDS_counts,"samps","counts", -Geneid, -Length)
mRNA_counts_tidy <- merge(mRNA_counts_tidy, experimentType, by="samps")
CDS_counts_tidy <- merge(CDS_counts_tidy, experimentType, by="samps")

raw_counts_tidy <- rbind(filter(mRNA_counts_tidy, exp=="AF"),filter(CDS_counts_tidy, exp=="RP"))
}


featureCounts_tidy <- loadFC(experimentTypes, CDS, five_utr, three_utr)
total_mRNA_counts <- featureCounts_tidy %>%
  group_by(Samples) %>%
  summarize(mRNA_counts=sum(counts))
total_mRNA_counts
total_mRNA_counts_plot <- ggplot(total_mRNA_counts, aes(x=Samples, y=mRNA_counts, fill=Samples))+geom_bar(stat="identity", position="dodge")+JF_theme+theme(axis.text.x = element_text(angle = 45, hjust=1))+xlab("")+geom_hline(yintercept = 400000)
total_mRNA_counts_plot
```

###Calculate the and plot reads per kilobase per million mapped (to mRNA) for the ribosome profiling experiments
```{r regionCovCalc}
#moveTo new Rmd
loadFCw_regions <- function(experimentType, CDS, five_utr, three_utr) {
sampNames=experimentType$samps
CDS <- read.table(file=CDS, header= T, sep="\t", stringsAsFactors = F)
Five_prime <- read.table(file=five_utr, header= T, sep="\t", stringsAsFactors = F)
Three_prime <- read.table(file=three_utr, header= T, sep="\t", stringsAsFactors = F)
filteredCDS <- CDS %>% 
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)
filtered5p <- Five_prime %>%
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)
filtered3p <- Three_prime %>%
  filter(Geneid %in% Five_prime$Geneid & Geneid %in% Three_prime$Geneid)

maxSampCol <- 6+length(sampNames)
mRNA_counts <- filteredCDS[,7:maxSampCol]+filtered5p[,7:maxSampCol]+filtered3p[,7:maxSampCol]
CDS_counts <- filteredCDS[,7:maxSampCol]
FiveP_counts <- filtered5p[,7:maxSampCol]
ThreeP_counts <- filtered3p[,7:maxSampCol]
colnames(CDS_counts) <- sampNames
colnames(FiveP_counts) <- sampNames
colnames(ThreeP_counts) <- sampNames
CDS_counts <- CDS_counts %>% mutate(Geneid=filteredCDS$Geneid, Length=filteredCDS$Length, region="CDS")
FiveP_counts <-FiveP_counts %>% mutate(Geneid=filteredCDS$Geneid, Length=filtered5p$Length, region="5' UTR")
ThreeP_counts <-ThreeP_counts %>% mutate(Geneid=filteredCDS$Geneid, Length=filtered3p$Length, region="3' UTR")

All_counts <- rbind(CDS_counts, FiveP_counts, ThreeP_counts)
All_counts_tidy <- gather(All_counts, "samps","counts", -Geneid, -Length, -region)
All_counts_tidy <- merge(All_counts_tidy, experimentType, by="samps")

rpk <- All_counts_tidy %>%
  group_by(samps, region, exp) %>%
  summarize(totCounts=sum(counts),totLength=sum(Length)) %>%
  mutate(rpk=totCounts/(totLength/1000))

million_mapped <- All_counts_tidy %>%
  group_by(samps) %>%
  summarize(mRNA_mapped_mil=sum(counts)/1e6)

rpkm <- merge(rpk, million_mapped, by="samps") %>%
  mutate(rpkm=rpk/mRNA_mapped_mil)
#
}

rpkm_byRegion <- loadFCw_regions(experimentTypes, CDS, five_utr, three_utr)
rpkm_byRegion$region <- factor(rpkm_byRegion$region, levels=c("5' UTR","CDS","3' UTR"))
rpkm_byRegion <- merge(rpkm_byRegion, descriptiveNames, by="samps")
rpkm_byRegion <- rpkm_byRegion %>%
  rename(Samples=descriptiveNames)
rpkm_byRegion_plot <- ggplot(filter(rpkm_byRegion, exp=="RP"), aes(y=rpkm,x=Samples, fill=region))+facet_wrap(~exp)+geom_bar(stat="identity", position="dodge")+JF_theme+theme(axis.text.x = element_text(angle = 45, hjust=1))+xlab("")
rpkm_byRegion_plot
#
```


###Calculate TPMs from the raw counts
```{r calculateTPM, echo=FALSE}

calcTPM <- function(RC_tidy) {
  RC <- RC_tidy %>% 
    select(samps, Geneid, Length, counts) %>% #not robust to different types of metadata
    spread("samps","counts")
  Geneid=RC$Geneid
  lengths=RC$Length
  rawCounts = RC[,3:ncol(RC)] #rawCounts start with 7th column
  
  x <- rawCounts/lengths
  tpm.mat <- t( t(x) * 1e6 / colSums(x) )
  TPM_out <- data.frame(Geneid, tpm.mat)
  colnames(TPM_out) <- c("Geneid",colnames(rawCounts))
  TPM_out
}
TPMs <- calcTPM(featureCounts_tidy)
```

###Now plot the log10TPM+1 distributions
```{r plotTPM_distros, echo=FALSE}

tpmDensity <- function(TPMs,TPM_title="TPM_density") {
  #takes in a tidy table of tpms and returns a density plot
 JF_theme <- theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),panel.background = element_blank(), axis.line = element_line(colour = "black"), plot.title = element_text(face = "bold"))
  tpmPlot <- ggplot(TPMs,aes(log10TPM,fill=Samples ))+geom_histogram(alpha=0.3,position="identity",bins=40)+JF_theme+theme(legend.position = c(0.75,0.75))+ggtitle(TPM_title)
}


TPMs_tidy <- gather(TPMs, "samps","TPM",-Geneid)
TPMs_tidy <- TPMs_tidy %>%
  mutate(log10TPM=log10(TPM+1))
TPMs_tidy <- merge(TPMs_tidy, descriptiveNames, by="samps")
TPMs_tidy <- TPMs_tidy %>%
  rename(Samples=descriptiveNames)

TPMdensity <- tpmDensity(TPMs_tidy)
TPMdensity <- TPMdensity + ylab("# of genes")
TPMdensity
```

###Plot the number of genes above TPM threshold=10 and threshold=20
```{r aboveTPMthreshold, echo=FALSE}
aboveTPM <- function(TPM_tidy, threshold=10) {
  #take in a tpm table and a tpm threshold, 
  tpm_frame <- data.frame(TPM_tidy, tpmFilter=TPM_tidy$TPM >= threshold)
  tpm_frame <- group_by(tpm_frame, Samples)
  tpm_aboveFilt <- summarize(tpm_frame, tpmFilter=sum(tpmFilter))
}

aboveTPM10 <- aboveTPM(TPMs_tidy, threshold=10)
aboveTPM20 <- aboveTPM(TPMs_tidy, threshold=20)


aboveTPM10_plot <- ggplot(aboveTPM10, aes(x=Samples, y=tpmFilter, fill=Samples))+geom_col()+ggtitle("TPM above 10")+JF_theme+xlab("")+theme(legend.position = "none")+theme(axis.text.x = element_text(angle = 45, hjust=1))+ylab("# genes above threshold")
aboveTPM20_plot <- ggplot(aboveTPM20, aes(x=Samples, y=tpmFilter, fill=Samples))+geom_col()+ggtitle("TPM above 20")+JF_theme+theme(legend.position = "none")+theme(axis.text.x = element_text(angle = 45, hjust=1))+ylab("# genes above threshold")+xlab("")

aboveTPM10_plot
aboveTPM20_plot
```


###Plot the number of genes in each sample above rawCounts=10, 20, 40
```{r aboveRawThreshold, echo=FALSE}
aboveRaw <- function(FC_tidy, threshold=10) {
  #take in a featureCounts table and a raw threshold, 
  raw_frame <- data.frame(FC_tidy, rawFilter=FC_tidy$counts >= threshold)
  raw_frame <- group_by(raw_frame, Samples)
  raw_aboveFilt <- summarize(raw_frame, rawFilter=sum(rawFilter))
}

aboveRaw10 <- aboveRaw(featureCounts_tidy, threshold=10)
aboveRaw20 <- aboveRaw(featureCounts_tidy, threshold=20)
aboveRaw40 <- aboveRaw(featureCounts_tidy, threshold=40)

aboveRaw10_plot <- ggplot(aboveRaw10, aes(x=Samples, y=rawFilter, fill=Samples))+geom_col()+ggtitle("Raw above 10")+JF_theme+theme(legend.position = "none")+theme(axis.text.x = element_text(angle = 45, hjust=1))+ylab("# genes above threshold")+xlab("")
aboveRaw20_plot <- ggplot(aboveRaw20, aes(x=Samples, y=rawFilter, fill=Samples))+geom_col()+ggtitle("Raw above 20")+JF_theme+theme(legend.position = "none")+theme(axis.text.x = element_text(angle = 45, hjust=1))+ylab("# genes above threshold")+xlab("")
aboveRaw40_plot <- ggplot(aboveRaw40, aes(x=Samples, y=rawFilter, fill=Samples))+geom_col()+ggtitle("Raw above 40")+JF_theme+theme(legend.position = "none")+theme(axis.text.x = element_text(angle = 45, hjust=1))+ylab("# genes above threshold")+xlab("")

aboveRaw10_plot
aboveRaw20_plot
aboveRaw40_plot
```

###Calculate and plot Pearson's R correlation coefficient for all genes above a certain TPM filtering condition. 
###Leave mRNA counts and date in the report, remove for the figure version for clarity.
```{r TPMCorrelationHeatmap}
filterTPMs <- function(TPMs, threshold=10) {
  TPMs_filt <- filter_if(TPMs, is.double, all_vars(. > threshold))
}

TPMs_filtered <- filterTPMs(TPMs,20)
cormat <- cor(select((TPMs_filtered), -Geneid))
experimentTypesPlus_mRNACounts <- merge(experimentTypes, total_mRNA_counts, by="Samples")
anno <- select(experimentTypesPlus_mRNACounts, -samps)
rownames(anno) <- experimentTypesPlus_mRNACounts$samps
TPM_cor_pheatmap <- pheatmap(cormat, display_numbers = T, color=colorRampPalette(brewer.pal(n=9, name="YlOrRd"))(100), annotation_col=select(anno, -Samples), treeheight_row = 10, treeheight_col = 10, show_colnames = F, labels_col = anno$Samples, show_rownames = F, fontsize=7)
TPM_cor_pheatmap

anno <- select(experimentTypesPlus_mRNACounts, -samps, -date, -mRNA_counts)
rownames(anno) <- experimentTypesPlus_mRNACounts$samps
TPM_cor_pheatmap_simple <- pheatmap(cormat, display_numbers = T, color=colorRampPalette(brewer.pal(n=9, name="YlOrRd"))(100), annotation_col=select(anno, -Samples), treeheight_row = 10, treeheight_col = 10, show_colnames = F, labels_col = anno$Samples, show_rownames = F, fontsize=7)
TPM_cor_pheatmap_simple

```



###Run RiboWaltz to evaluate 3-nt periodicity in the data sets. 
###Requires a directory of transcriptome-aligned BamFiles and a gtf
```{r bamToList, echo=FALSE}
#Try loading in all the files and see how it goes. After a more careful
#reading of the RiboWaltz vignette, it looks like it will process samples separately.

#make a named vector mapping bamFile names (no path or extension needed) to descriptiveNames
RP_samps <- filter(experimentTypes, exp=="RP")
RP_descriptiveNames <- RP_samps$Samples
names(RP_descriptiveNames) <- RP_samps$samps


annotation_dt <- create_annotation(gtfpath = gtf, dataSource = "GRCm38.95",organism="Mus musculus")
reads_list <- bamtolist(bamfolder = bamFiles, annotation = annotation_dt, transcript_align = FALSE, name_samples = RP_descriptiveNames) 
```

```{r lengthFilter, echo=FALSE}
#From last run, need to set length_filter_mode="custom" for "bad" or
#RNA-seq libraries to get scored. 
filtered_list <- length_filter(data = reads_list, length_filter_mode = "custom", length_filter_vector = 27:33)
```

```{r psite, echo=FALSE}
#perform p-site analysis for all samples, save data to a sub-folder of the transcriptome alignment.
dir.create(outDir)
psite_dir <- paste(outDir,'/psite', sep="")
dir.create(psite_dir)
psite <- psite(data=filtered_list, plot=T, plot_dir = psite_dir, plot_format = "pdf")
```

###Make plots that stratify by frame:
####At first, I plotted using reads_list. This makes plots across all reads lengths
####Plotting using the length-filtered reads (filtered_list) makes much nicer graphs.
```{r stratify, echo=FALSE}
reads_psite_list <- psite_info(filtered_list,psite)
frames_stratified <- frame_psite_length(reads_psite_list, region="all",cl=90)
strat_dir <- paste(outDir,"/frame_counts",sep="")
dir.create(strat_dir)
#setwd(strat_dir)
pdfName <- paste(strat_dir, "Frame_counts_all_samples.pdf", sep="/" )
frames_stratified$plot

pdf(pdfName)
frames_stratified$plot
dev.off()

#The plot looks squished if I try plotting all of them on the same plot. Try looping over the sampleNames and making individual plots for each one. Update: If I only include filtered reads, the graphs aren't too squished and are readable
sampleNames=unique(psite$sample)
for(sampleName in sampleNames) {
  frame_stratified=frame_psite_length(reads_psite_list,  sample=sampleName, region="all",cl=90)
  pdfName=paste(sampleName,".pdf",sep="")
  pdf(file=pdfName)
  plot(frame_stratified$plot)
  dev.off()
}
```

###P-site heatmaps (both directly from RiboWaltz, and by re-plotting the RiboWaltz data for a nicer-looking plot) for all samples in current set. 
###Currently using scale factors as 1/CDS_reads:
```{r psite_heatmap, echo=FALSE}
heat_dir <- paste(outDir,"/heatmaps",sep="")
dir.create(heat_dir)
#setwd(heat_dir)

#calculate scale factors as the inverse of CDS_counts
scale_factors <- filter(featureCounts_tidy, exp=="RP") %>%
  group_by(samps) %>%
  summarize(CDS_reads=sum(counts)) %>%
  mutate(scale_factors=1/CDS_reads) %>%
  select(scale_factors)
scale_factors <- c(t(scale_factors))
names(scale_factors) <- sampleNames

#Build up the list of sample names
sampNamesList <- list()
for (sampName in sampleNames) {
  sampNamesList[[sampName]] <- c(sampName)
}

psite_heatmap <- metaheatmap_psite(reads_psite_list, annotation=annotation_dt,log=F,utr5l=15, cdsl=30, utr3l=15, sample=sampNamesList, scale_factors=scale_factors)

psite_heatmap$plot

pdfName=paste(heat_dir, "psite_heatmap.pdf",sep="/")
pdf(file=pdfName, width=11, height=8.5)
plot(psite_heatmap$plot)
dev.off()


#Example plotting function that looks nicer than a heatmap:
psite_signal_plot <- ggplot(psite_heatmap$dt, aes(x=distance,y=reads,color=sample))+geom_line()+JF_theme+facet_wrap(~reg)
psite_signal_plot

psite_dt <- psite_heatmap$dt
psite_dt$reg <- plyr::revalue(psite_dt$reg, c("Distance from start (nt)"="from start (nt):","Distance from stop (nt)"="from stop (nt):"))

psite_signal_heat <- ggplot(psite_dt, aes(x=distance,fill=reads,y=sample))+geom_tile()+facet_wrap(~reg)+scale_fill_gradient(low="white", high="black")+JF_theme+xlab("distance from start or stop (nt)")+ylab("")+xlab("")
psite_signal_heat

#look at how scaling is done. Try row normalization of raw read counts to compute frequencies within the regions of interest

psite_heatmap_raw <- metaheatmap_psite(reads_psite_list, annotation=annotation_dt,log=F,utr5l=15, cdsl=30, utr3l=15, sample=sampNamesList)
psite_dt_raw <- psite_heatmap_raw$dt
psite_dt_raw$reg <- plyr::revalue(psite_dt_raw$reg, c("Distance from start (nt)"="from start (nt):","Distance from stop (nt)"="from stop (nt):"))

psite_region_sums <- psite_dt_raw %>%
  group_by(sample, reg) %>%
  summarize(region_sum=sum(reads))
psite_dt_raw <- merge(psite_dt_raw, psite_region_sums, by=c("sample","reg"))
psite_dt_raw <- psite_dt_raw %>%
  mutate(freq=reads/region_sum)

psite_freqHeat <- ggplot(psite_dt_raw, aes(x=distance,fill=freq,y=sample))+geom_tile()+facet_wrap(~reg)+scale_fill_gradient(low="white", high="black")+JF_theme+xlab("distance from start or stop (nt)")+ylab("")+xlab("")+theme(legend.position="bottom")
psite_freqHeat 

pdfName=paste(heat_dir, "psite_freq_heatmap.pdf",sep="/")
pdf(file=pdfName, width=11, height=8.5)
plot(psite_freqHeat)
dev.off()

```

###Next, calculate the percent of P-sites within each reading frame for each sample.
```{r framePsite, echo=FALSE}
frame_dir <- paste(outDir,"/frame_psite_percentage",sep="")
dir.create(frame_dir)
#setwd(frame_dir)

frame_psites <- frame_psite(reads_psite_list, region="all")

frame_psites$plot

pdfName <- paste(frame_dir, "Frame_percentages_all_samples.pdf", sep="/")
pdf(pdfName)
plot(frame_psites$plot)
dev.off()

sampleNames=unique(psite$sample)
for(sampleName in sampleNames) {
  aFramePsite=frame_psite(reads_psite_list, sample=sampleName, region="all")
  pdfName=paste(sampleName,".pdf",sep="")
  pdf(file=pdfName)
  plot(aFramePsite$plot)
  dev.off()
}
```

###And make P-site frequency plots for comparison across samples:
```{r psitePlotting, echo=FALSE}
plotFramePsites <- function(samples, frame_psites, sampNames=samples) {
  #function to take in a list of samples, and frame_psites Ribowaltz object    #Plots the percentage of reads over CDSs in each frame, grouped by samples.
  FPs <- frame_psites$dt
  FPsToPlot <- filter(FPs, sample %in% samples & region == "CDS" )
  FPplot <- ggplot(data=FPsToPlot, aes(x=sample,y=percentage,fill=as.factor(frame)))
  FPplot <- FPplot+geom_bar(stat="identity", position=position_dodge())+theme_minimal(base_size=18)+scale_fill_discrete(name="frame")+ylim(0,100)+scale_x_discrete(labels=sampNames)+JF_theme
  FPplot <- FPplot+theme(axis.text.x = element_text(angle = 45, hjust=1))+xlab("")+ylab("%P-sites in each frame")
                           
}

#setwd(frame_dir)
psiteFreq<- plotFramePsites(sampleNames,frame_psites)
pdfName=paste(frame_dir, "psite_freq.pdf",sep="/")
pdf(file=pdfName, width=11, height=8.5)
plot(psiteFreq)
dev.off()

psiteFreq

```

###Assemble plots into a "QC_fig" (showing region coverage, length distros, P-site heatmaps
###and frequency and "Counts_fig" (showing TPM_distro, #of genes above threshold, sample
###perason correlation and clustering). The QC_fig starts at Panel C because A and B are
###Gel images and browser shots to be added in Illustrator. 

```{r patchwork}
QC_patch <- (plot_spacer() | plot_spacer()) / (rpkm_byRegion_plot | CDS_riboProf) / allAnno_RiboProf_bySample / (psite_freqHeat | psiteFreq)+theme(text=element_text("ArialMT")) + plot_annotation(tag_levels = list(c("C","D","E","F","G"))) & theme(plot.tag=element_text(size=12, face="bold"))

pdfName=paste(outDir, "QC_fig.pdf", sep="/")
pdf(file=pdfName, width=6.5, height=9)
plot(QC_patch)
dev.off()

Counts_patch <- (TPMdensity | aboveTPM20_plot) / (as.ggplot(TPM_cor_pheatmap_simple)) +
  theme(text=element_text("ArialMT")) + plot_annotation(tag_levels = list(c("A","B","C"))) & theme(plot.tag=element_text(size=12, face="bold"))

pdfName=paste(outDir, "Counts_fig.pdf", sep="/")
pdf(file=pdfName, width=6.5, height=9)
plot(Counts_patch)
dev.off()

QC_patch
Counts_patch
```

